// @a10n/blueprint-typespec/dist/src/index.js
import { defineLinter } from "@typespec/compiler";
import { inlineRefsRule } from "./rules/inline-refs.rule.js";
import { featureInvariantCapRule } from "./rules/feature-invariant-cap.rule.js";
import { featureconfigFieldCapRule } from "./rules/featureconfig-field-cap.rule.js";
import { namespacePathRule } from "./rules/namespace-path.rule.js";
import { invariantMissingProofRule } from "./rules/invariant-missing-proof.rule.js";
import { systemProseRefsRule } from "./rules/system-prose-refs.rule.js";
import { entityFieldReachableRule } from "./rules/entity-field-reachable.rule.js";
import { proofMissingProjectsRule } from "./rules/proof-missing-projects.rule.js";
import { callsGraphRule } from "./rules/calls-graph.rule.js";
import { uiRouteAuthoringRule } from "./rules/ui-route-authoring.rule.js";
import { stateMachineAuthoringRule } from "./rules/state-machine-authoring.rule.js";
import { specReferencesRule } from "./rules/spec-references.rule.js";
import { operationReachabilityRule } from "./rules/operation-reachability.rule.js";
import { graphAuthoringRule } from "./rules/graph-authoring.rule.js";
import { eventCallGraphRule } from "./rules/event-call-graph.rule.js";
import { workerTypeSpecRule } from "./rules/worker-type-spec.rule.js";
import { dependencyAuthoringRule } from "./rules/dependency-authoring.rule.js";
import { componentBlockFormRule } from "./rules/component-block-form.rule.js";
import { singleBlockFormRule } from "./rules/single-block-form.rule.js";
import { opInvariantCapRule } from "./rules/op-invariant-cap.rule.js";
import { sdkOperationsNamespaceRule } from "./rules/sdk-operations-namespace.rule.js";
import { projectsTargetsInvariantRule } from "./rules/projects-targets-invariant.rule.js";
import { invariantProjectedOncePerOpRule } from "./rules/invariant-projected-once-per-op.rule.js";
import { noOrphanInvariantLayerRule } from "./rules/no-orphan-invariant-layer.rule.js";
import { entityPrimitiveConstrainedRule } from "./rules/entity-primitive-constrained.rule.js";
import { surfaceIsNamespaceRule } from "./rules/surface-is-namespace.rule.js";
import { datasetSlicesDistinctRule } from "./rules/dataset-slices-distinct.rule.js";
export { $lib } from "./lib.js";
// ── Component decorators ────────────────────────────────────────────────────
export { $component, $dependsOn, $testing, $flow, $uiRoute, 
// getters for emitters
getComponentConfig, getDependsOn, getTestingStrategies, isFlowOp, getUiRouteConfig, } from "./decorators/component.js";
// ── Call-graph decorators ────────────────────────────────────────────────────
export { $calls, $reads, $writes, getCalls, getReads, getWrites, } from "./decorators/call-graph.js";
// ── Tool decorators ─────────────────────────────────────────────────────────
export { $tool, $artifact, $interrupt, isInterruptOp, 
// getters for emitters
getToolConfig, } from "./decorators/tool.js";
// ── Graph decorators ────────────────────────────────────────────────────────
export { $graph, $node, $predict, 
// getters for emitters
isGraphNamespace, isNodeOp, getGraphNodeConfig, isPredictOp, } from "./decorators/graph.js";
// ── SDK decorators ──────────────────────────────────────────────────────────
export { $operation, 
// getters for emitters
getSdkOperationConfig, } from "./decorators/sdk.js";
// ── Product / feature decorators ────────────────────────────────────────────
export { $feature, $product, $assumption, 
// getters for emitters
isProduct, isFeature, } from "./decorators/product.js";
// ── Dependency decorators ───────────────────────────────────────────────────
export { $dependency, getDependencyConfig } from "./decorators/dependency.js";
// ── Eval / signal / scoring decorators ─────────────────────────────────────
export { $eval, $signal, $scoring, $configImpact, $featureConfig, $evalSuite, $dataset, getEvalConfig, getSignalsInEval, getFeatureConfigModels, getScoringOpInEval, getScoringConfig, getEvalSuiteConfig, getDatasetConfig, } from "./decorators/eval.js";
// ── File-store decorators ───────────────────────────────────────────────────
export { $storage, $storageGroup, $storageObject, getStorageConfig, getStorageGroupConfig, getStorageObjectConfig, resolveStoragePrefix, } from "./decorators/file-store.js";
// ── Domain layer decorators ─────────────────────────────────────────────────
export { $entity, 
// getters for emitters
isEntity, } from "./decorators/entity.js";
export { $invariant, 
// getters for emitters
getInvariantConfig, getInvariantsInFeature, } from "./decorators/invariant.js";
export { $projects, 
// getters for emitters
getProjectsTargets, getProjectsTargetFqns, hasProjects, } from "./decorators/projects.js";
// ── Linter ──────────────────────────────────────────────────────────────────
export const $linter = defineLinter({
    rules: [inlineRefsRule, featureInvariantCapRule, featureconfigFieldCapRule, namespacePathRule, invariantMissingProofRule, systemProseRefsRule, entityFieldReachableRule, proofMissingProjectsRule, callsGraphRule, uiRouteAuthoringRule, stateMachineAuthoringRule, specReferencesRule, operationReachabilityRule, graphAuthoringRule, eventCallGraphRule, workerTypeSpecRule, dependencyAuthoringRule, componentBlockFormRule, singleBlockFormRule, opInvariantCapRule, sdkOperationsNamespaceRule, projectsTargetsInvariantRule, invariantProjectedOncePerOpRule, noOrphanInvariantLayerRule, entityPrimitiveConstrainedRule, surfaceIsNamespaceRule, datasetSlicesDistinctRule],
    ruleSets: {
        all: {
            enable: {
                "@a10n/blueprint-typespec/inline-refs": true,
                "@a10n/blueprint-typespec/feature-invariant-cap": true,
                "@a10n/blueprint-typespec/featureconfig-field-cap": true,
                "@a10n/blueprint-typespec/namespace-path": true,
                "@a10n/blueprint-typespec/invariant-missing-proof": true,
                "@a10n/blueprint-typespec/system-prose-refs": true,
                "@a10n/blueprint-typespec/entity-field-reachable": true,
                "@a10n/blueprint-typespec/proof-missing-projects": true,
                "@a10n/blueprint-typespec/calls-graph": true,
                "@a10n/blueprint-typespec/ui-route-authoring": true,
                "@a10n/blueprint-typespec/state-machine-authoring": true,
                "@a10n/blueprint-typespec/spec-references": true,
                "@a10n/blueprint-typespec/operation-reachability": true,
                "@a10n/blueprint-typespec/graph-authoring": true,
                "@a10n/blueprint-typespec/event-call-graph": true,
                "@a10n/blueprint-typespec/worker-type-spec": true,
                "@a10n/blueprint-typespec/dependency-authoring": true,
                "@a10n/blueprint-typespec/component-block-form": true,
                "@a10n/blueprint-typespec/single-block-form": true,
                "@a10n/blueprint-typespec/op-invariant-cap": true,
                "@a10n/blueprint-typespec/sdk-operations-namespace": true,
                "@a10n/blueprint-typespec/projects-targets-invariant": true,
                "@a10n/blueprint-typespec/invariant-projected-once-per-op": true,
                "@a10n/blueprint-typespec/no-orphan-invariant-layer": true,
                "@a10n/blueprint-typespec/entity-primitive-constrained": true,
                "@a10n/blueprint-typespec/surface-is-namespace": true,
                "@a10n/blueprint-typespec/dataset-slices-distinct": true,
            },
        },
        recommended: {
            enable: {
                "@a10n/blueprint-typespec/inline-refs": true,
                "@a10n/blueprint-typespec/feature-invariant-cap": true,
                "@a10n/blueprint-typespec/featureconfig-field-cap": true,
                "@a10n/blueprint-typespec/namespace-path": true,
                "@a10n/blueprint-typespec/invariant-missing-proof": true,
                "@a10n/blueprint-typespec/system-prose-refs": true,
                "@a10n/blueprint-typespec/entity-field-reachable": true,
                "@a10n/blueprint-typespec/proof-missing-projects": true,
                "@a10n/blueprint-typespec/calls-graph": true,
                "@a10n/blueprint-typespec/ui-route-authoring": true,
                "@a10n/blueprint-typespec/state-machine-authoring": true,
                "@a10n/blueprint-typespec/spec-references": true,
                "@a10n/blueprint-typespec/operation-reachability": true,
                "@a10n/blueprint-typespec/graph-authoring": true,
                "@a10n/blueprint-typespec/event-call-graph": true,
                "@a10n/blueprint-typespec/worker-type-spec": true,
                "@a10n/blueprint-typespec/dependency-authoring": true,
                "@a10n/blueprint-typespec/component-block-form": true,
                "@a10n/blueprint-typespec/single-block-form": true,
                "@a10n/blueprint-typespec/op-invariant-cap": true,
                "@a10n/blueprint-typespec/sdk-operations-namespace": true,
                "@a10n/blueprint-typespec/projects-targets-invariant": true,
                "@a10n/blueprint-typespec/invariant-projected-once-per-op": true,
                "@a10n/blueprint-typespec/no-orphan-invariant-layer": true,
                "@a10n/blueprint-typespec/entity-primitive-constrained": true,
                "@a10n/blueprint-typespec/surface-is-namespace": true,
                "@a10n/blueprint-typespec/dataset-slices-distinct": true,
            },
        },
    },
});
export { specReferencesRule } from "./rules/spec-references.rule.js";
export { graphAuthoringRule } from "./rules/graph-authoring.rule.js";
export { eventCallGraphRule } from "./rules/event-call-graph.rule.js";
export { workerTypeSpecRule } from "./rules/worker-type-spec.rule.js";
export { componentBlockFormRule } from "./rules/component-block-form.rule.js";
export { singleBlockFormRule } from "./rules/single-block-form.rule.js";
export { dependencyAuthoringRule } from "./rules/dependency-authoring.rule.js";
// ── Worker decorators ───────────────────────────────────────────────────────
export { $publishes, $subscribes, $worker, getPublishes, isSubscribesOp, getWorkerConfig, isWorkerNamespace, } from "./decorators/worker.js";
