// @a10n/blueprint-typespec/dist/src/lib.js
import { createTypeSpecLibrary, paramMessage } from "@typespec/compiler";
export const $lib = createTypeSpecLibrary({
    name: "@a10n/blueprint-typespec",
    diagnostics: {
        "unresolved-inline-ref": {
            severity: "warning",
            messages: {
                default: paramMessage `Cannot resolve inline reference {@${"ref"}} — no type named "${"ref"}" found in the program. Check that the type is declared and imported, and that the name matches exactly (including namespace path).`,
                withSuggestion: paramMessage `Cannot resolve inline reference {@${"ref"}} — no type named "${"ref"}" found in the program. Did you mean {@${"suggestion"}}?`,
            },
        },
        "invalid-decorator-target": {
            severity: "error",
            messages: {
                default: paramMessage `@${"decorator"} can only be applied to ${"expected"}, but was applied to ${"actual"}. Move the decorator to the correct declaration kind.`,
            },
        },
        "namespace-block-form-required": {
            severity: "error",
            messages: {
                default: paramMessage `@${"decorator"} on namespace "${"fqn"}" requires block form \`namespace ${"fqn"} { }\` — decorators on semicolon-form namespaces are not indexed.`,
            },
        },
        "duplicate-decorator": {
            severity: "error",
            messages: {
                default: paramMessage `@${"decorator"} cannot be applied more than once to the same declaration. Remove the duplicate.`,
            },
        },
        "invalid-array-element-type": {
            severity: "error",
            messages: {
                default: paramMessage `Field "${"field"}[${"index"}]" must be ${"expected"}, but got ${"actual"}.`,
            },
        },
        "missing-required-fields": {
            severity: "error",
            messages: {
                default: paramMessage `@${"config"} is missing required fields: ${"fields"}. Add the missing fields to the decorator argument.`,
            },
        },
        "invalid-dependency-type": {
            severity: "error",
            messages: {
                default: paramMessage `@dependsOn argument[${"index"}] must be ${"expected"}, but got ${"actual"}. Pass a namespace reference, e.g. @dependsOn([Components.MyAgent]).`,
            },
        },
        "invalid-model-type": {
            severity: "error",
            messages: {
                default: paramMessage `Field "${"field"}" must reference a model, but got ${"actual"}. Pass a model type reference as the argument.`,
            },
        },
        "invalid-enum-value": {
            severity: "error",
            messages: {
                default: paramMessage `"${"value"}" is not a valid value for "${"field"}". Expected one of: ${"valid"}.`,
            },
        },
        "invalid-url-format": {
            severity: "error",
            messages: {
                default: paramMessage `"${"value"}" in field "${"field"}" is not a valid URL. Use a full URL starting with http:// or https://.`,
            },
        },
        "invalid-namespace-prefix": {
            severity: "error",
            messages: {
                default: paramMessage `@${"decorator"} must be on a namespace whose fully-qualified name starts with "${"expectedPrefix"}", but got "${"actualFqn"}". Rename the namespace to match the required prefix.`,
            },
        },
        "invalid-namespace-nesting": {
            severity: "error",
            messages: {
                default: paramMessage `@${"decorator"} must be nested inside a @${"requiredParent"} namespace, but "${"actualFqn"}" has no such ancestor. Declare this namespace inside a @${"requiredParent"} block.`,
            },
        },
        "invalid-decorator-context": {
            severity: "error",
            messages: {
                default: paramMessage `@${"decorator"} can only be applied to a namespace that also has @${"context"}, but "${"actualFqn"}" does not. Add @${"context"} to the same namespace first.`,
            },
        },
        "invalid-assumption-context": {
            severity: "error",
            messages: {
                default: paramMessage `@assumption can only be applied to a namespace that also has @product, @feature, or @dependency, but "${"actualFqn"}" has none of these.`,
            },
        },
        "invalid-namespace-name": {
            severity: "error",
            messages: {
                default: paramMessage `Namespace segment "${"segment"}" in "${"fqn"}" must be PascalCase (start with uppercase letter, letters and digits only).`,
            },
        },
        "component-name-suffix": {
            severity: "error",
            messages: {
                default: paramMessage `@component name "${"name"}" must end with one of [${"expected"}] for type ${"type"} (e.g. Components.MyReview${"example"}). Rename the namespace to match the component type.`,
            },
        },
        "invalid-eval-namespace": {
            severity: "error",
            messages: {
                default: paramMessage `@eval namespace must follow the pattern Evals.<FeatureName>.E<NNN> (e.g. Evals.PrReview.E001), but got "${"actual"}". The sub-namespace relative to the @component namespace must match exactly.`,
            },
        },
        "projects-field-must-target-entity-field": {
            severity: "error",
            messages: {
                default: paramMessage `@projects on a model field must target an entity field (ModelProperty of an @entity model), but ref "${"ref"}" is a ${"kind"}. Field-level @projects links a storage field to its domain entity field.`,
            },
        },
        "projects-op-must-target-invariant": {
            severity: "error",
            messages: {
                default: paramMessage `@projects on an operation must target an @invariant op, but ref "${"ref"}" is a ${"kind"}. Op-level @projects links a test case or signal to the invariant it proves.`,
            },
        },
        "projects-op-not-proof": {
            severity: "error",
            messages: {
                default: paramMessage `@projects on op "${"opName"}" is not allowed: it must be an op that enforces invariants — a @tool, @node, @predict, @interrupt, @operation, an op in an Http/Mcp/Grpc interface, an op in a @uiRoute or @worker namespace, or a @signal/@dataset op.`,
            },
        },
        "entity-field-unreachable": {
            severity: "warning",
            messages: {
                default: paramMessage `Field "${"field"}" of @entity "${"entity"}" is not referenced by any @invariant predicate or @projects target. Every entity field must be reachable from the domain layer.`,
            },
        },
        "invariant-missing-proof": {
            severity: "warning",
            messages: {
                default: paramMessage `@invariant "${"fqn"}" has no @testCase or @signal that declares @projects(...) targeting it. Add at least one test case that proves the invariant holds.`,
            },
        },
        "system-references-domain-entity": {
            severity: "warning",
            messages: {
                default: paramMessage `Field "${"field"}" in system-layer model "${"model"}" uses @entity type "${"entityFqn"}" as its field type. System-layer models must not structurally depend on domain entities — use @projects on the field instead (field-level projection) or extract the value type to a plain model or enum.`,
            },
        },
        "invariant-pred-system-ref": {
            severity: "warning",
            messages: {
                default: paramMessage `{@${"ref"}} in @invariant predicate references a system-layer construct. Invariant predicates must only reference domain types (entities, entity fields, enums, enum values), not system-layer constructs (components, component ops, projections, signals, test cases).`,
            },
        },
        "entity-primitive-unconstrained": {
            severity: "warning",
            messages: {
                default: paramMessage `@entity "${"entity"}" field "${"field"}" is a bare ${"typeName"} with no constraint decorator and no @invariant predicate reference. Add @pattern/@minLength/@maxValue/etc. or reference it in an @invariant predicate as {@${"fieldRef"}}.`,
            },
        },
        "featureconfig-field-cap": {
            severity: "warning",
            messages: {
                default: paramMessage `@featureConfig model "${"modelName"}" has ${"count"} fields; the maximum is 10. Split the feature config into two models or split the feature.`,
            },
        },
        "feature-invariant-cap": {
            severity: "warning",
            messages: {
                default: paramMessage `Feature "${"featureFqn"}" has ${"count"} invariants; the maximum is 20. Split into two feature files by capability.`,
            },
        },
        "proof-missing-projects": {
            severity: "warning",
            messages: {
                testCase: paramMessage `@testCase "${"opName"}" has no @projects. Add @projects(...) listing the invariant(s) it proves.`,
                signal: paramMessage `@signal "${"opName"}" has no @projects. Add @projects(...) listing the invariant(s) it proves.`,
                dataset: paramMessage `@dataset "${"opName"}" has no @projects. Add @projects(...) listing the invariant(s) it covers.`,
            },
        },
        "invariant-non-void-return": {
            severity: "error",
            messages: {
                default: paramMessage `@invariant op "${"opName"}" must return void, but returns ${"returnType"}. Change the return type to void.`,
            },
        },
        "invariant-id-shape": {
            severity: "warning",
            messages: {
                default: paramMessage `@invariant op name "${"opName"}" must be at most 3 snake_case words with no numeric suffix (e.g. task_has_id). Rename to a short verb phrase.`,
            },
        },
        "signal-id-shape": {
            severity: "warning",
            messages: {
                default: paramMessage `@signal op name "${"opName"}" must follow the pattern E{NNN}_{NN} (e.g. E001_01). Rename to match the pattern.`,
            },
        },
        "dataset-id-shape": {
            severity: "warning",
            messages: {
                default: paramMessage `@dataset op name "${"opName"}" must match [a-z][a-z0-9_-]+ (e.g. d001_pr_gate). Rename to match the pattern.`,
            },
        },
        "calls-invalid-target": {
            severity: "error",
            messages: {
                default: paramMessage `@calls target "${"ref"}" is not an operation — @calls can only list other operations as targets.`,
            },
        },
        "reads-invalid-target": {
            severity: "error",
            messages: {
                default: paramMessage `@reads target "${"ref"}" is not a model inside a @component(type: Store) — @reads can only list store models as targets.`,
            },
        },
        "writes-invalid-target": {
            severity: "error",
            messages: {
                default: paramMessage `@writes target "${"ref"}" is not a model inside a @component(type: Store) — @writes can only list store models as targets.`,
            },
        },
        "doc-omits-facts": {
            severity: "warning",
            messages: {
                default: paramMessage `@doc on "${"opName"}" restates call-graph targets already declared by @calls, @reads, or @writes — remove the restatement.`,
            },
        },
        "calls-route-target": {
            severity: "error",
            messages: {
                default: paramMessage `@calls target "${"ref"}" is a @uiRoute namespace — routes must be navigated to or nested, not invoked as calls.`,
            },
        },
        "calls-invalid-caller-callee": {
            severity: "error",
            messages: {
                default: paramMessage `@calls from "${"callerKind"}" to "${"calleeKind"}" ("${"ref"}") is not permitted. Allowed @calls targets per caller: @get/@post/@put/@patch/@delete/@head→(@operation|@node entry|Http/Mcp/Commands ops); @uiRoute action→(Http/Mcp/Commands ops only); @tool→(@operation|@node entry|Http/Mcp/Commands|@interrupt); @node→(@operation|@node entry|Http/Mcp/Commands|@tool|@predict|@interrupt); @subscribes worker→(@operation|@node entry|Http/Mcp/Commands); @evalSuite→(@scoring Score|@dataset|@predict|@node entry|Http/Mcp/Commands); @scoring Score→(@signal); probabilistic @signal→(@predict).`,
            },
        },
        "op-single-kind": {
            severity: "error",
            messages: {
                default: paramMessage `"${"decorator"}" conflicts with "${"existing"}" already on this op — an op may carry at most one op-kind decorator.`,
            },
        },
        "op-missing-kind": {
            severity: "error",
            messages: {
                default: `This op has no op-kind decorator — every op must carry exactly one of: @tool, @node, @predict, @interrupt, @operation, @signal, @dataset, @evalSuite, @scoring, @testCase, @invariant, @get/@post/@put/@patch/@delete/@head, @MCP.tool, @subscribes.`,
            },
        },
        "pop-names-parent": {
            severity: "warning",
            messages: {
                default: paramMessage `Pop action op "${"opName"}" references its parent route in @doc — Pop must not name the route it returns to.`,
            },
        },
        "no-mermaid-in-doc": {
            severity: "warning",
            messages: {
                default: paramMessage `@uiRoute namespace "${"ns"}" has a Mermaid diagram in @doc — Mermaid diagrams duplicate the call graph and must be removed.`,
            },
        },
        "doc-omits-call-list": {
            severity: "warning",
            messages: {
                default: paramMessage `@uiRoute namespace "${"ns"}" @doc enumerates backend calls — this is already declared via @calls on action ops.`,
            },
        },
        "test-case-reads-writes-in-steps": {
            severity: "warning",
            messages: {
                default: paramMessage `@testCase "${"opName"}" steps contain {@${"ref"}} which is a @reads/@writes target of a participant op — store models belong in setup (pre-existing state) or expects (observed outcomes), not in steps.`,
            },
        },
        "test-case-calls-in-setup": {
            severity: "warning",
            messages: {
                default: paramMessage `@testCase "${"opName"}" setup or cleanup contains {@${"ref"}} which is a @calls target of a participant op — called ops belong in steps or expects, not in setup or cleanup.`,
            },
        },
        "test-case-runtime-prose": {
            severity: "warning",
            messages: {
                default: paramMessage `@testCase "${"opName"}" describes runtime infrastructure (worker pools, dispatchers) — remove topology prose.`,
            },
        },
        "test-case-bare-name": {
            severity: "warning",
            messages: {
                default: paramMessage `@testCase "${"opName"}" references "${"name"}" as a bare name — use a {@...} typed reference instead.`,
            },
        },
        "state-machine-missing-invariant": {
            severity: "warning",
            messages: {
                default: paramMessage `Entity "${"entity"}" has a lifecycle field but no transition @invariant — declare an enum + @invariant to replace prose FSM descriptions.`,
            },
        },
        "path-on-modal": {
            severity: "warning",
            messages: {
                default: paramMessage `@uiRoute namespace "${"ns"}" carries a path field but has no URL-addressable context — remove the path from non-addressable surfaces.`,
            },
        },
        "no-inv-clause": {
            severity: "warning",
            messages: {
                default: paramMessage `@testCase "${"opName"}" prose contains an inline invariant satisfaction clause — express invariant linkage via @projects only.`,
            },
        },
        "no-nested-call-in-steps": {
            severity: "warning",
            messages: {
                default: paramMessage `@testCase "${"opName"}" steps narrate a nested @calls target — nested ops belong in expects, not steps.`,
            },
        },
        "expects-observable-only": {
            severity: "warning",
            messages: {
                default: paramMessage `@testCase "${"opName"}" expects field narrates internal call sequence — expects must assert observable end-state only.`,
            },
        },
        "spec-ref-bare-name": {
            severity: "warning",
            messages: {
                default: paramMessage `"${"name"}" in spec prose is a named construct used as a bare name — wrap it as {@${"name"}}.`,
            },
        },
        "spec-ref-missing-actor": {
            severity: "warning",
            messages: {
                default: paramMessage `"${"word"}" in spec prose is an unnamed external actor — use {@Actor} instead.`,
            },
        },
        "spec-ref-on-standard": {
            severity: "warning",
            messages: {
                default: paramMessage `{@${"ref"}} wraps a standard protocol term — standard vocabulary (HTTP methods, status codes, JSON, ISO dates) must not be wrapped in references.`,
            },
        },
        "spec-ref-on-generic": {
            severity: "warning",
            messages: {
                default: paramMessage `{@${"ref"}} wraps a generic prose noun with no declared construct — remove the reference wrapper.`,
            },
        },
        "spec-ref-missing-using": {
            severity: "warning",
            messages: {
                default: paramMessage `{@${"ref"}} references a namespace without a corresponding using declaration in this file.`,
            },
        },
        "spec-ref-unused-using": {
            severity: "warning",
            messages: {
                default: paramMessage `using "${"ns"}" is declared but no {@…} reference to any construct in this namespace appears in the file.`,
            },
        },
        "nav-not-return-type": {
            severity: "warning",
            messages: {
                default: paramMessage `Action op "${"opName"}" in @uiRoute namespace returns void — action ops must return UiCommand to declare their navigation intent.`,
            },
        },
        // ── call-graph proof ────────────────────────────────────────────────────
        "calls-unproven": {
            severity: "warning",
            messages: {
                default: paramMessage `@calls target "${"ref"}" on "${"opName"}" has no typed {@...} reference in any @testCase steps or expects — add a steps entry showing the invocation (e.g. "{@Components.X.${"ref"}} is called") or an expects entry asserting its observable outcome.`,
            },
        },
        "writes-unproven": {
            severity: "warning",
            messages: {
                default: paramMessage `@writes target "${"ref"}" on "${"opName"}" has no typed {@...} reference in any @testCase expects — add an expects entry asserting the mutation (e.g. "{@Components.Store.${"ref"}} row is created").`,
            },
        },
        "reads-unproven": {
            severity: "warning",
            messages: {
                default: paramMessage `@reads target "${"ref"}" on "${"opName"}" has no typed {@...} reference in any @testCase setup or expects — add a setup entry seeding the data (e.g. "{@Components.Store.${"ref"}} exists") or an expects entry asserting a value read from it.`,
            },
        },
        // ── operation-reachability ──────────────────────────────────────────────
        "op-missing-edge": {
            severity: "warning",
            messages: {
                default: paramMessage `op "${"opName"}" in a Service component has no @calls, @reads, or @writes — add @calls([Op]) listing every op it invokes, @reads([Model]) for every store model it queries, and @writes([Model]) for every store model it mutates.`,
            },
        },
        "ui-action-missing-edge": {
            severity: "warning",
            messages: {
                default: paramMessage `UI action op "${"opName"}" has no @calls, @reads, or @writes — add @calls([Op]) for backend ops it invokes, @reads([Model]) for store models it queries, or @writes([Model]) for store models it mutates.`,
            },
        },
        "tool-file-path": {
            severity: "warning",
            messages: {
                default: paramMessage `@tool op "${"opName"}" must be declared in a file at \`components/[agent]/tools/[name].tsp\`, but is in "${"actual"}".`,
            },
        },
        "tool-missing-sequential": {
            severity: "warning",
            messages: {
                default: paramMessage `@tool op "${"opName"}" does not declare \`sequential\`. Set \`sequential: true\` (cannot parallelize) or \`sequential: false\` (safe to parallelize) explicitly.`,
            },
        },
        "tool-missing-edge": {
            severity: "warning",
            messages: {
                default: paramMessage `@tool op "${"opName"}" has no @calls, @reads, or @writes — add @calls([Op]) for ops it invokes, @reads([Model]) for store models it queries, or @writes([Model]) for store models it mutates.`,
            },
        },
        "op-unreachable": {
            severity: "warning",
            messages: {
                default: paramMessage `op "${"opName"}" has no inbound @calls edge and is not referenced in any @testCase — either add @calls([Components.X.${"opName"}]) on a caller op, or add a @testCase with a steps or expects entry containing {@Components.X.${"opName"}}.`,
            },
        },
        "ui-action-unreachable": {
            severity: "warning",
            messages: {
                default: paramMessage `UI action op "${"opName"}" has no inbound @calls edge and is not referenced in any @testCase — add a @testCase with a steps or expects entry containing a typed {@...} reference to it.`,
            },
        },
        "tool-unreachable": {
            severity: "warning",
            messages: {
                default: paramMessage `@tool op "${"opName"}" has no inbound @calls edge and is not referenced in any @testCase — either add @calls([${"opName"}]) on a caller op, or add a @testCase with a steps or expects entry referencing it.`,
            },
        },
        "store-model-no-reader": {
            severity: "warning",
            messages: {
                default: paramMessage `Store model "${"modelName"}" has no @reads edge and no typed reference in any @testCase expects — add @reads([${"modelName"}]) on an op that queries it, or add a @testCase expects entry with {@Components.Store.${"modelName"}}.`,
            },
        },
        "store-model-no-writer": {
            severity: "warning",
            messages: {
                default: paramMessage `Store model "${"modelName"}" has no @writes edge and no typed reference in any @testCase setup — add @writes([${"modelName"}]) on an op that mutates it, or add a @testCase setup entry with {@Components.Store.${"modelName"}} if it is written by a human (e.g. a config file).`,
            },
        },
        "calls-not-a-scorer": {
            severity: "error",
            messages: {
                default: paramMessage `@evalSuite @calls target "${"ref"}" is not a Score op inside an @eval namespace — eval suites may only call Score ops.`,
            },
        },
        "evalSuite-missing-inputs": {
            severity: "warning",
            messages: {
                default: paramMessage `@evalSuite op "${"opName"}" must declare exactly one typed Inputs parameter (e.g. \`op Suite(inputs: SuiteInputs): void\`). Even a suite with no tunables must declare \`model EmptyInputs {}\` and pass it.`,
            },
        },
        "inputs-field-missing-example": {
            severity: "warning",
            messages: {
                default: paramMessage `Inputs field "${"field"}" on @evalSuite "${"opName"}" has no @example decorator. Add @example(<value>) so the eval emitter can use it as a default when no runtime value is supplied.`,
            },
        },
        "dataset-file-path": {
            severity: "warning",
            messages: {
                default: paramMessage `@dataset op "${"opName"}" must be declared in a file at \`components/[component]/datasets/[feature]/NNN_[purpose].tsp\`, but is in "${"actual"}".`,
            },
        },
        "evalSuite-file-path": {
            severity: "warning",
            messages: {
                default: paramMessage `@evalSuite op "${"opName"}" must be declared in a file at \`components/[component]/evals/[feature]/NNN_[purpose].tsp\`, but is in "${"actual"}".`,
            },
        },
        "dataset-missing-reads": {
            severity: "warning",
            messages: {
                default: paramMessage `@dataset op "${"opName"}" has no @reads decorator. Add @reads([...]) pointing at the store model(s) that supply the dataset items.`,
            },
        },
        "dataset-empty-title": {
            severity: "error",
            messages: {
                default: paramMessage `@dataset op "${"opName"}" has an empty title. Provide a non-empty title string describing the dataset's purpose.`,
            },
        },
        "scoring-op-returns-void": {
            severity: "error",
            messages: {
                default: paramMessage `Score op "${"opName"}" decorated with @scoring must not return void — it must return a numeric score type (e.g. float32).`,
            },
        },
        "storage-duplicate": {
            severity: "error",
            messages: {
                default: paramMessage `@storage is already declared in component "${"component"}". A Store component can have at most one @storage declaration.`,
            },
        },
        "storage-xor-group": {
            severity: "error",
            messages: {
                default: paramMessage `A namespace cannot carry both \`@storage\` and \`@storageGroup\`. Remove one of the decorators.`,
            },
        },
        "storage-path-collision": {
            severity: "error",
            messages: {
                default: paramMessage `Storage object "${"modelName"}" resolves to path "${"path"}" which is already used by "${"other"}". Each storage object in a component must have a unique address.`,
            },
        },
        "evalSuite-no-reads-writes": {
            severity: "error",
            messages: {
                reads: paramMessage `@evalSuite "${"opName"}" must not carry @reads — dataset inputs are declared via @calls.`,
                writes: paramMessage `@evalSuite "${"opName"}" must not carry @writes — eval suites are read-only measurement contracts.`,
            },
        },
        "eval-missing-score": {
            severity: "error",
            messages: {
                missing: paramMessage `@eval namespace "${"fqn"}" has no Score op carrying @scoring. Add exactly one \`op Score(): void\` decorated with @scoring and @calls.`,
                multiple: paramMessage `@eval namespace "${"fqn"}" has ${"count"} Score ops carrying @scoring. An @eval namespace must have exactly one.`,
            },
        },
        "dataset-unreachable": {
            severity: "warning",
            messages: {
                default: paramMessage `@dataset op "${"opName"}" is not called by any @evalSuite. Add @calls([${"opName"}]) on a suite that exercises it.`,
            },
        },
        "eval-unreachable": {
            severity: "warning",
            messages: {
                default: paramMessage `@eval namespace "${"fqn"}" is not called by any @evalSuite. Add an @evalSuite that carries @calls([${"fqn"}.Score, ...]).`,
            },
        },
        "evalSuite-no-projects": {
            severity: "error",
            messages: {
                default: paramMessage `@evalSuite op "${"opName"}" must not carry @projects — eval suites are runnable bundles, not proof carriers. Remove @projects.`,
            },
        },
        "evalSuite-no-calls": {
            severity: "warning",
            messages: {
                default: paramMessage `@evalSuite op "${"opName"}" has no @calls. Add @calls([SomeEval.Score, dataset_op]) listing at least one scorer and one dataset.`,
            },
        },
        "evalSuite-no-dataset-calls": {
            severity: "warning",
            messages: {
                default: paramMessage `@evalSuite op "${"opName"}" has no @dataset in @calls. Add @calls([..., dataset_op]) with at least one @dataset op.`,
            },
        },
        "evalSuite-scorer-ungated": {
            severity: "warning",
            messages: {
                default: paramMessage `@evalSuite op "${"opName"}" calls scorer "${"scorer"}" but does not reference it in expects. Add a \`{@${"scorer"}} >= <threshold>\` entry to expects.`,
            },
        },
        "dataset-missing-projects": {
            severity: "warning",
            messages: {
                default: paramMessage `@dataset op "${"opName"}" carries no @projects. Add @projects(...) listing the invariant(s) it exercises.`,
            },
        },
        "scoring-formula-external-signal": {
            severity: "warning",
            messages: {
                default: paramMessage `@scoring formula on "${"opName"}" references "${"signal"}" which is not a @signal op in this @eval namespace. Only local signals may appear in the formula.`,
            },
        },
        "signal-not-called-by-score": {
            severity: "warning",
            messages: {
                default: paramMessage `@signal "${"opName"}" in @eval "${"evalFqn"}" is not called by the Score op. Add it to @calls([..., ${"opName"}]) on the Score op.`,
            },
        },
        "evalSuite-no-signal": {
            severity: "warning",
            messages: {
                default: paramMessage `@evalSuite op "${"opName"}" has no @signal reachable via its Score op @calls. The gate measures nothing.`,
            },
        },
        "storage-bare-placeholder": {
            severity: "error",
            messages: {
                default: paramMessage `Bare placeholder "${"placeholder"}" in storage path must be a typed ref — use \`{@ModelName.fieldName}\` instead of \`{fieldName}\`.`,
            },
        },
        "storage-placeholder-unresolved": {
            severity: "warning",
            messages: {
                default: paramMessage `Placeholder "${"ref"}" in storage path/root/prefix does not resolve to a model in this component.`,
            },
        },
        "storage-body-has-path-field": {
            severity: "warning",
            messages: {
                default: paramMessage `@storageObject model "${"modelName"}" has field "${"field"}" which is also a path placeholder. Path fields must live on a separate path model, not on the body model.`,
            },
        },
        // ── @worker config validation ───────────────────────────────────────────
        "worker-one-per-key-missing-routing-key": {
            severity: "error",
            messages: {
                default: paramMessage `@worker namespace "${"ns"}" uses concurrency: OnePerKey but does not declare routing_key — add routing_key with the name of the field used to partition messages.`,
            },
        },
        // ── @graph entry / terminal ─────────────────────────────────────────────
        "graph-missing-entry": {
            severity: "error",
            messages: {
                default: paramMessage `@graph namespace "${"fqn"}" has no entry @node — exactly one @node must carry \`@graphNode(#{ is_entrypoint: true })\`.`,
            },
        },
        "graph-multiple-entries": {
            severity: "error",
            messages: {
                default: paramMessage `@graph namespace "${"fqn"}" has ${"count"} entry @node ops — exactly one may carry \`@graphNode(#{ is_entrypoint: true })\`.`,
            },
        },
        // ── suite-no-reads ──────────────────────────────────────────────────────
        "suite-no-reads": {
            severity: "error",
            messages: {
                default: paramMessage `@evalSuite "${"opName"}" must not carry @reads — dataset inputs are declared via @calls.`,
            },
        },
        // ── @predict diagnostics ────────────────────────────────────────────────
        "predict-untyped-output": {
            severity: "error",
            messages: {
                default: paramMessage `@predict op "${"opName"}" returns a free-form string — output must be a typed model. Declare a named model and use it as the return type.`,
            },
        },
        // ── pure-callee proof ───────────────────────────────────────────────────
        "pure-callee-unproven": {
            severity: "warning",
            messages: {
                default: paramMessage `op "${"ref"}" is a pure callee (no @reads or @writes) not referenced in any @testCase expects — add an expects entry asserting its observable outcome or add @reads/@writes to declare its effects.`,
            },
        },
        // ── dependency op reachability ──────────────────────────────────────────
        "dependency-op-unreached": {
            severity: "warning",
            messages: {
                default: paramMessage `op "${"opName"}" in a dependency component is not reached by any first-party @calls, @subscribes, or @publishes — remove it from the dependency spec or add a first-party caller.`,
            },
        },
        // ── @interrupt diagnostics ──────────────────────────────────────────────
        "interrupt-multi-arg-input": {
            severity: "error",
            messages: {
                default: paramMessage `@interrupt op "${"opName"}" declares multiple input args — an interrupt must have exactly one typed model parameter.`,
            },
        },
        "interrupt-missing-output": {
            severity: "error",
            messages: {
                default: paramMessage `@interrupt op "${"opName"}" returns void — an interrupt must declare a typed output model.`,
            },
        },
        "interrupt-declares-errors": {
            severity: "error",
            messages: {
                default: paramMessage `@interrupt op "${"opName"}" declares error return types — interrupts always resolve; move error handling to the caller.`,
            },
        },
        // ── @publishes / @subscribes / @worker diagnostics ──────────────────────────
        "publishes-event-model": {
            severity: "error",
            messages: {
                default: paramMessage `@publishes target "${"ref"}" is not a variant of an @events-decorated union — only event payload variants are valid @publishes targets.`,
            },
        },
        "publishes-op-only": {
            severity: "warning",
            messages: {
                default: paramMessage `@publishes can only appear on operations, but was found on a ${"kind"} — move it to an op declaration.`,
            },
        },
        "subscribes-op-only": {
            severity: "warning",
            messages: {
                default: paramMessage `@subscribes can only appear on ops inside a @worker namespace.`,
            },
        },
        "publishes-incomplete": {
            severity: "warning",
            messages: {
                default: paramMessage `op "${"opName"}" doc references event "${"ref"}" which is not declared in @publishes — add it to @publishes.`,
            },
        },
        "publishes-absent-pure": {
            severity: "warning",
            messages: {
                default: paramMessage `op "${"opName"}" carries @publishes but doc references no event emissions — remove @publishes or declare the event emission in @doc.`,
            },
        },
        "subscribes-incomplete": {
            severity: "warning",
            messages: {
                default: paramMessage `worker op "${"opName"}" doc references event "${"ref"}" not matched by its input type — update the input parameter to subscribe to all listed events.`,
            },
        },
        "publishes-no-subscriber": {
            severity: "warning",
            messages: {
                default: paramMessage `published event "${"ref"}" has no @subscribes worker op consuming it — add a worker that subscribes to it.`,
            },
        },
        "worker-input-variant": {
            severity: "warning",
            messages: {
                default: paramMessage `worker op "${"opName"}" input type "${"ref"}" is not a variant of an @events-decorated union — worker inputs must be event payload models.`,
            },
        },
        // ── worker-type-spec ────────────────────────────────────────────────────
        "worker-returns-void": {
            severity: "warning",
            messages: {
                default: paramMessage `@worker op "${"opName"}" returns non-void type "${"returnType"}" — worker ops are fire-and-forget and must return void.`,
            },
        },
        "worker-has-subscription": {
            severity: "warning",
            messages: {
                default: paramMessage `@worker op "${"opName"}" has no @subscribes — a worker op with no subscription has no trigger and is dead code.`,
            },
        },
        "worker-has-effect": {
            severity: "warning",
            messages: {
                default: paramMessage `@worker op "${"opName"}" has no @reads, @writes, or @calls — a no-op handler the spec cannot verify.`,
            },
        },
        "subscribes-live-event": {
            severity: "warning",
            messages: {
                default: paramMessage `@worker op "${"opName"}" subscribes to event "${"ref"}" which no op ever publishes — the subscription will never fire.`,
            },
        },
        "worker-test-excluded": {
            severity: "warning",
            messages: {
                default: paramMessage `@testCase "${"opName"}" ${"field"} contains a reference to worker op "${"ref"}" — worker ops are never invoked directly by tests.`,
            },
        },
        "pub-reads-setup": {
            severity: "warning",
            messages: {
                default: paramMessage `@testCase "${"opName"}" steps include op "${"stepOp"}" which publishes to worker chain; worker @reads target "${"ref"}" is missing from setup.`,
            },
        },
        "pub-writes-expects": {
            severity: "warning",
            messages: {
                default: paramMessage `@testCase "${"opName"}" steps include op "${"stepOp"}" which publishes to worker chain; worker @writes target "${"ref"}" is missing from expects.`,
            },
        },
        "actor-reads-setup": {
            severity: "warning",
            messages: {
                default: paramMessage `@testCase "${"opName"}" actor emits event "${"event"}" in steps; worker chain @reads target "${"ref"}" is missing from setup.`,
            },
        },
        "actor-writes-expects": {
            severity: "warning",
            messages: {
                default: paramMessage `@testCase "${"opName"}" actor emits event "${"event"}" in steps; worker chain @writes target "${"ref"}" is missing from expects.`,
            },
        },
        "worker-unreachable": {
            severity: "warning",
            messages: {
                default: paramMessage `@worker op "${"opName"}" is not reachable via any test — no op in any test steps publishes the event it subscribes to, and no test has {@Actor} emitting it.`,
            },
        },
        // ── test-case-peer-boundary ─────────────────────────────────────────────
        "steps-actor-only": {
            severity: "warning",
            messages: {
                default: paramMessage `@testCase "${"opName"}" step does not start with {@Actor}: "${"step"}" — all steps must describe an action taken by {@Actor}.`,
            },
        },
        "steps-runtime-kind": {
            severity: "warning",
            messages: {
                default: paramMessage `@testCase "${"opName"}" step references "${"ref"}" which is a ${"kind"} — steps may only reference Service/Ui ops or subject ops.`,
            },
        },
        "assert-direct-store": {
            severity: "warning",
            messages: {
                default: paramMessage `@testCase "${"opName"}" expects references "${"ref"}" via a peer operation for a subject-internal store — assert the store model directly instead.`,
            },
        },
        "assert-via-boundary": {
            severity: "warning",
            messages: {
                default: paramMessage `@testCase "${"opName"}" expects references store model "${"ref"}" of a peer Service/Ui component directly — assert via that peer's read operation instead.`,
            },
        },
        "seed-direct-store": {
            severity: "warning",
            messages: {
                default: paramMessage `@testCase "${"opName"}" setup references "${"ref"}" via an operation for a subject-internal store — seed the store model directly instead.`,
            },
        },
        "seed-via-boundary": {
            severity: "warning",
            messages: {
                default: paramMessage `@testCase "${"opName"}" setup references store model "${"ref"}" of a peer Service/Ui component directly — seed via that peer's write operation instead.`,
            },
        },
        "peer-assertable": {
            severity: "warning",
            messages: {
                default: paramMessage `Peer component "${"ref"}" reached from test "${"opName"}" has no typed read op and is not declared @testing(approach: Mock) — add a read op or declare the testing approach.`,
            },
        },
        "peer-seedable": {
            severity: "warning",
            messages: {
                default: paramMessage `Peer component "${"ref"}" reached from test "${"opName"}" has no typed write op and is not declared @testing(approach: Mock) — add a write op or declare the testing approach.`,
            },
        },
        "mock-invocation-asserted": {
            severity: "warning",
            messages: {
                default: paramMessage `@testCase "${"opName"}" exercises mocked peer "${"ref"}" but expects does not assert it was invoked — add an expects entry asserting the mock call.`,
            },
        },
        "testing-covers-reach": {
            severity: "warning",
            messages: {
                default: paramMessage `Component "${"ref"}" is transitively reached by test "${"opName"}" but has no @testing declaration on the subject's main file.`,
            },
        },
        "testing-targets-components": {
            severity: "warning",
            messages: {
                default: paramMessage `@testing target "${"ref"}" is a @dependency namespace — target the nested @component instead.`,
            },
        },
        // ── test-case store model validation ───────────────────────────────────
        "expects-undeclared-store": {
            severity: "warning",
            messages: {
                default: paramMessage `@testCase "${"opName"}" expects references store model "${"ref"}" which is not a @writes target of any transitively-reachable op — either add @writes([${"ref"}]) on a participant op or remove the reference.`,
            },
        },
        "setup-undeclared-store": {
            severity: "warning",
            messages: {
                default: paramMessage `@testCase "${"opName"}" setup references store model "${"ref"}" which is not a @reads target of any transitively-reachable op — either add @reads([${"ref"}]) on a participant op or remove the reference.`,
            },
        },
        // ── deep call-chain coverage ────────────────────────────────────────────
        "deep-reads-setup": {
            severity: "warning",
            messages: {
                default: paramMessage `op "${"opName"}" is transitively reachable via @calls from test participant but its @reads target "${"ref"}" is absent from setup — add a setup entry seeding "${"ref"}".`,
            },
        },
        "deep-writes-expects": {
            severity: "warning",
            messages: {
                default: paramMessage `op "${"opName"}" is transitively reachable via @calls from test participant but its @writes target "${"ref"}" is absent from expects — add an expects entry asserting "${"ref"}".`,
            },
        },
        "step-undeclared-op": {
            severity: "warning",
            messages: {
                default: paramMessage `@testCase "${"testCase"}" step references "${"ref"}" which is not reachable from any participant via @calls, @publishes, or @subscribes — add the missing call-graph edge or remove the reference.`,
            },
        },
        "steps-at-boundaries": {
            severity: "warning",
            messages: {
                default: paramMessage `@testCase "${"opName"}" step "${"step"}" contains no typed op or event model ref — every step must reference at least one operation or event model to describe a compile-checkable boundary crossing.`,
            },
        },
        "predict-setup-staged": {
            severity: "warning",
            messages: {
                default: paramMessage `@testCase "${"opName"}" has @predict op "${"ref"}" reachable via @calls within the subject component but no setup entry references it — add a setup entry staging the mock response.`,
            },
        },
        "interrupt-setup-staged": {
            severity: "warning",
            messages: {
                default: paramMessage `@testCase "${"opName"}" has @interrupt op "${"ref"}" reachable via @calls within the subject component but no setup entry references it — add a setup entry staging the mock decision.`,
            },
        },
        // ── dependency-authoring ────────────────────────────────────────────────
        "dep-component-kind": {
            severity: "warning",
            messages: {
                default: paramMessage `Component "${"ref"}" inside @dependency "${"dep"}" has type ${"type"} — only Sdk and Service components are permitted inside dependencies.`,
            },
        },
        "dep-uniform-kind": {
            severity: "warning",
            messages: {
                default: paramMessage `@dependency "${"dep"}" mixes Sdk and Service components — all components in a dependency must be the same kind.`,
            },
        },
        "dep-addresses-components": {
            severity: "warning",
            messages: {
                default: paramMessage `@dependsOn entry "${"ref"}" is a @dependency namespace — address a nested component instead (e.g. ${"ref"}.MyComponent).`,
            },
        },
        "dep-no-tests": {
            severity: "warning",
            messages: {
                default: paramMessage `@testSuite or @testCase found inside @dependency "${"dep"}" — dependencies are external surfaces and must not contain tests.`,
            },
        },
        "uiroute-not-ui-namespace": {
            severity: "warning",
            messages: {
                default: paramMessage `@uiRoute namespace "${"ns"}" is not nested inside a @component(type: Ui) — UI surfaces must live inside a Ui component namespace.`,
            },
        },
        "dataset-slices-identical": {
            severity: "warning",
            messages: {
                default: paramMessage `@dataset op "${"second"}" has the same @projects invariant set as "${"first"}" in the same file — datasets in the same file must target distinct invariant slices.`,
            },
        },
    },
});
export const { reportDiagnostic } = $lib;
