// @a10n/blueprint-typespec/dist/src/validators.js
import { getSourceLocation } from "@typespec/compiler";
import { reportDiagnostic } from "./lib.js";
/**
 * Convert a PascalCase segment to kebab-case.
 * Examples: A10nIndexMcp → a10n-index-mcp, ReviewBot → review-bot
 */
export function toKebab(s) {
    return s
        .replace(/([a-z0-9])([A-Z])/g, "$1-$2")
        .replace(/([A-Z]+)([A-Z][a-z])/g, "$1-$2")
        .toLowerCase();
}
/**
 * Get the source file path of a type node, normalized to forward slashes.
 * Returns undefined for synthetic nodes (e.g. test host, built-ins).
 */
export function getTypeFilePath(type) {
    if (!type.node)
        return undefined;
    const loc = getSourceLocation(type.node);
    if (!loc)
        return undefined;
    return loc.file.path.replace(/\\/g, "/");
}
/**
 * Returns true if the file path should be exempt from path enforcement.
 * The test host places the default code string at /test/main.tsp.
 * Files placed at structured paths under /test/ are intentional and NOT exempt.
 */
export function isExemptPath(filePath) {
    return filePath === "/test/main.tsp" || filePath === "Z:/test/main.tsp";
}
// (?<!\\) skips escaped refs: \{@Foo} is an intentional escape in spec examples
const INLINE_REF_REGEX = /(?<!\\)\{@([\w.]+)\}/g;
export function extractInlineRefs(str) {
    const refs = [];
    let match;
    INLINE_REF_REGEX.lastIndex = 0;
    while ((match = INLINE_REF_REGEX.exec(str)) !== null) {
        refs.push(match[1]);
    }
    return refs;
}
/**
 * Resolve an inline ref token to its TypeSpec Type.
 * Tries global FQN resolution first, then tries ancestor namespace prefixes
 * (simulating `using` scope for short refs like `WorkerEvents.my_variant`),
 * then falls back to walking the global namespace tree.
 * Also handles union variant access (e.g. "MyEvents.my_variant").
 *
 * Pass `contextNs` (the namespace of the op containing the inline ref) to
 * enable short-ref resolution via ancestor scope — all existing callers that
 * only pass program get global-only resolution as before.
 */
export function resolveInlineRef(program, ref, contextNs) {
    // 1. Try global FQN first (handles full FQNs and anything in global scope)
    const [t] = program.resolveTypeReference(ref);
    if (t !== undefined)
        return t;
    // 2. Try with each ancestor namespace as prefix (simulates `using` scope)
    //    e.g. ref="WorkerEvents.run_requested", contextNs=Components.WorkerService
    //    → tries "Components.WorkerService.WorkerEvents.run_requested" first
    if (contextNs) {
        const ancestors = [];
        let cur = contextNs;
        while (cur && cur.name) {
            ancestors.unshift(cur.name);
            cur = cur.namespace;
        }
        for (let i = ancestors.length; i > 0; i--) {
            const prefix = ancestors.slice(0, i).join(".");
            const candidate = `${prefix}.${ref}`;
            const [ct] = program.resolveTypeReference(candidate);
            if (ct !== undefined)
                return ct;
            const ns = resolveNsFromGlobal(program, prefix);
            if (ns) {
                const found = resolveInNs(ns, ref);
                if (found !== undefined)
                    return found;
            }
        }
    }
    // 3. Fallback: walk global namespace tree
    return resolveInNs(program.getGlobalNamespaceType(), ref);
}
function resolveNsFromGlobal(program, fqn) {
    const segs = fqn.split(".");
    let cur = program.getGlobalNamespaceType();
    for (const seg of segs) {
        const child = cur.namespaces.get(seg);
        if (!child)
            return undefined;
        cur = child;
    }
    return cur;
}
function resolveInNs(root, ref) {
    const segs = ref.split(".");
    let cur = root;
    for (let i = 0; i < segs.length; i++) {
        const seg = segs[i];
        const isLast = i === segs.length - 1;
        if (isLast) {
            const found = cur.namespaces.get(seg) ?? cur.operations.get(seg) ?? cur.models.get(seg) ?? cur.enums.get(seg);
            if (found !== undefined)
                return found;
            // Check union variants: e.g. "MyEvents.my_variant" where MyEvents is a union in cur
            for (const [, union] of cur.unions ?? new Map()) {
                if (String(union.name) === seg)
                    return union;
                const variant = union.variants?.get(seg);
                if (variant !== undefined)
                    return variant;
            }
            return undefined;
        }
        const child = cur.namespaces.get(seg);
        if (child) {
            cur = child;
            continue;
        }
        // Intermediate segment might be a union name — next seg is the variant
        for (const [, union] of cur.unions ?? new Map()) {
            if (String(union.name) === seg) {
                const nextSeg = segs[i + 1];
                if (nextSeg !== undefined) {
                    return union.variants?.get(nextSeg);
                }
                return union;
            }
        }
        return undefined;
    }
    return undefined;
}
function getNamespacePrefixes(ns) {
    const prefixes = [];
    let current = ns;
    const parts = [];
    while (current && current.name !== "") {
        parts.unshift(current.name);
        current = current.namespace;
    }
    for (let i = parts.length; i >= 0; i--) {
        prefixes.push(parts.slice(0, i).join("."));
    }
    return prefixes;
}
/** Extract the dotted name string from a using statement name node (IdentifierNode | MemberExpressionNode). */
function usingNodeToString(node) {
    if (node.sv !== undefined) {
        // IdentifierNode
        return node.sv;
    }
    // MemberExpressionNode: has .base and .id
    return `${usingNodeToString(node.base)}.${node.id.sv}`;
}
/** Collect all `using` namespace prefixes visible from a namespace.
 *  `using` statements live on the TypeSpecScriptNode (file root), which is found by
 *  walking up the AST from the namespace node until we find a node with `.usings`. */
function getUsingPrefixes(ns) {
    const seen = new Set();
    const prefixes = [];
    function collectUsingsFromNode(node) {
        if (!node)
            return;
        if (Array.isArray(node.usings)) {
            for (const u of node.usings) {
                const s = usingNodeToString(u.name);
                if (!seen.has(s)) {
                    seen.add(s);
                    prefixes.push(s);
                }
            }
        }
    }
    let current = ns;
    while (current) {
        // Walk up the AST from the namespace node to find the script/file node that has .usings
        let node = current.node;
        while (node) {
            collectUsingsFromNode(node);
            node = node.parent;
        }
        current = current.namespace;
    }
    return prefixes;
}
/** Normalize a ref for fuzzy matching: lowercase, strip all non-alphanumeric chars.
 *  "blueprint.task_context" → "blueprinttaskcontext"
 *  "blueprint_task_context" → same
 *  "My.Model" → "mymodel" */
function normalizeRef(ref) {
    return ref.toLowerCase().replace(/[^a-z0-9]/g, "");
}
const ALIAS_STATEMENT_KIND = 24;
/** Collect all inline-ref-addressable FQNs from the program for fuzzy suggestion.
 *  Walks the full global namespace tree and yields: namespace FQNs, type names
 *  (model/enum/scalar/union/interface/op), enum members, model properties, and alias names. */
function collectAllFqns(program) {
    const fqns = [];
    function walkNs(ns, prefix) {
        const nsName = prefix ? `${prefix}.${ns.name}` : ns.name;
        if (nsName)
            fqns.push(nsName);
        for (const [name, model] of ns.models) {
            const fqn = nsName ? `${nsName}.${name}` : name;
            fqns.push(fqn);
            for (const [propName] of model.properties) {
                fqns.push(`${fqn}.${propName}`);
            }
        }
        for (const [name, enumType] of ns.enums) {
            const fqn = nsName ? `${nsName}.${name}` : name;
            fqns.push(fqn);
            for (const [memberName] of enumType.members) {
                fqns.push(`${fqn}.${memberName}`);
            }
        }
        for (const kind of ["scalars", "unions", "interfaces", "operations"]) {
            for (const [name] of ns[kind]) {
                fqns.push(nsName ? `${nsName}.${name}` : name);
            }
        }
        // Aliases (e.g. `alias Entry = A | B`) are transparent after compilation and
        // don't appear in any namespace type map — collect them from the AST node.
        const stmts = ns.node?.statements;
        if (Array.isArray(stmts)) {
            for (const stmt of stmts) {
                if (stmt.kind === ALIAS_STATEMENT_KIND && stmt.id?.sv) {
                    fqns.push(nsName ? `${nsName}.${stmt.id.sv}` : stmt.id.sv);
                }
            }
        }
        for (const [, childNs] of ns.namespaces) {
            walkNs(childNs, nsName);
        }
    }
    walkNs(program.getGlobalNamespaceType(), "");
    return fqns;
}
/** Find a known FQN whose normalized form matches the normalized ref.
 *  Matches on the full FQN or any trailing suffix (e.g. the ref "CheckInputKind.blueprint_task_context"
 *  normalizes the same as the suffix "CheckInputKind.blueprint.task_context" of a fully-qualified name).
 *  Returns the shortest matching FQN, or undefined if nothing is close enough. */
export function findSuggestion(program, ref) {
    const normalized = normalizeRef(ref);
    if (!normalized)
        return undefined;
    let best;
    for (const fqn of collectAllFqns(program)) {
        // Check every dot-split suffix of the collected FQN (e.g. "Ns.Enum.member" → also try "Enum.member", "member")
        const parts = fqn.split(".");
        for (let i = 0; i < parts.length; i++) {
            const suffix = parts.slice(i).join(".");
            if (normalizeRef(suffix) === normalized) {
                // Prefer the shortest qualifying name that matches (fewest leading segments stripped)
                if (!best || fqn.length < best.length)
                    best = fqn;
                break;
            }
        }
    }
    return best;
}
/** Check if a resolved type is a Union where all model variants have the given property. */
function unionHasProperty(type, propName) {
    if (type?.kind !== "Union")
        return false;
    const variants = [...type.variants.values()];
    return variants.length > 0 && variants.every((v) => v.type?.kind === "Model" && v.type.properties.has(propName));
}
/** For a dotted ref like "ChecksYamlEntry.id" or "Ns.ChecksYamlEntry.id", try each
 *  split point where the prefix resolves to a Union (alias) type and the suffix is a
 *  property shared by all model variants. Applies the same namespace prefix and using
 *  expansion as resolveRef so that unqualified alias names resolve regardless of call site. */
function resolveUnionPropertyRef(program, ref, contextNamespace) {
    const segments = ref.split(".");
    for (let split = 1; split < segments.length; split++) {
        const prefix = segments.slice(0, split).join(".");
        const propName = segments[split];
        // Try bare prefix
        const [type] = program.resolveTypeReference(prefix);
        if (unionHasProperty(type, propName))
            return true;
        if (contextNamespace) {
            // Try ns-ancestor-qualified prefixes
            for (const nsPfx of getNamespacePrefixes(contextNamespace)) {
                const fullPrefix = nsPfx ? `${nsPfx}.${prefix}` : prefix;
                const [t] = program.resolveTypeReference(fullPrefix);
                if (unionHasProperty(t, propName))
                    return true;
            }
            // Try using-qualified prefixes
            for (const usingPath of getUsingPrefixes(contextNamespace)) {
                const [tA] = program.resolveTypeReference(`${usingPath}.${prefix}`);
                if (unionHasProperty(tA, propName))
                    return true;
                // alias expansion: ref starts with last segment of using path
                const usingLast = usingPath.split(".").at(-1);
                if (usingLast === prefix.split(".")[0]) {
                    const remainder = prefix.indexOf(".") !== -1 ? prefix.slice(prefix.indexOf(".") + 1) : "";
                    const expanded = remainder ? `${usingPath}.${remainder}` : usingPath;
                    const [tB] = program.resolveTypeReference(expanded);
                    if (unionHasProperty(tB, propName))
                        return true;
                }
            }
        }
    }
    return false;
}
/** Look up a type by FQN directly through the global namespace's namespace/model/enum/op maps.
 *  Handles template declarations that resolveTypeReference skips because they need type args.
 *  Also handles model property access (e.g. "Review.id") and enum member access. */
function findTypeInGlobalNamespace(program, fqn) {
    const segments = fqn.split(".");
    let currentNs = program.getGlobalNamespaceType();
    for (let i = 0; i < segments.length; i++) {
        const seg = segments[i];
        const isLast = i === segments.length - 1;
        if (isLast) {
            // Terminal segment — check all type maps in current namespace
            if (currentNs.models.has(seg))
                return true;
            if (currentNs.enums.has(seg))
                return true;
            if (currentNs.operations.has(seg))
                return true;
            if (currentNs.scalars.has(seg))
                return true;
            if (currentNs.unions.has(seg))
                return true;
            if (currentNs.interfaces.has(seg))
                return true;
            return false;
        }
        // Non-terminal segment — first try descending into a child namespace
        const childNs = currentNs.namespaces.get(seg);
        if (childNs) {
            currentNs = childNs;
            continue;
        }
        // Not a namespace — check if this segment is a model or enum, then look up the member
        const model = currentNs.models.get(seg);
        if (model) {
            // Remaining segments are property names — walk through them
            const remaining = segments.slice(i + 1);
            let props = model.properties;
            for (let j = 0; j < remaining.length; j++) {
                if (!props.has(remaining[j]))
                    return false;
                if (j < remaining.length - 1) {
                    // Intermediate property — descend into its model type if possible
                    const propType = props.get(remaining[j])?.type;
                    if (propType?.kind === "Model") {
                        props = propType.properties;
                    }
                    else {
                        return false;
                    }
                }
            }
            return true;
        }
        const enumType = currentNs.enums.get(seg);
        if (enumType) {
            // Remaining segments form the member name. A member may have a literal dot in
            // its name (declared with backticks, e.g. `blueprint.task_context`), so try
            // joining all remaining segments first, then fall back to a single segment.
            const remaining = segments.slice(i + 1);
            if (remaining.length === 0)
                return false;
            // Member names may contain a literal dot (backtick-declared, e.g. `blueprint.task_context`),
            // so try joining all remaining segments as one name.
            return enumType.members.has(remaining.join("."));
        }
        return false;
    }
    return false;
}
export function resolveRef(program, ref, contextNamespace) {
    // Try resolving the full ref globally (covers fully-qualified names like Components.ReviewApiService.Review)
    const [type] = program.resolveTypeReference(ref);
    if (type !== undefined) {
        return true;
    }
    // Fallback: direct namespace walk (handles template declarations and enum members)
    if (findTypeInGlobalNamespace(program, ref)) {
        return true;
    }
    // Fallback for alias union property refs (e.g. {@ChecksYamlEntry.id} where
    // ChecksYamlEntry is an alias union and .id is shared by all model variants).
    // Passes contextNamespace so using/ancestor prefixes are tried for unqualified alias names.
    if (resolveUnionPropertyRef(program, ref, contextNamespace)) {
        return true;
    }
    if (contextNamespace) {
        // 1. Qualify with ancestor namespace prefixes.
        //    ref="Review", ns="Components.ReviewApiService" → try "Components.ReviewApiService.Review", "Components.Review"
        const nsPrefixes = getNamespacePrefixes(contextNamespace);
        for (const prefix of nsPrefixes) {
            const fullRef = prefix ? `${prefix}.${ref}` : ref;
            const [resolvedType] = program.resolveTypeReference(fullRef);
            if (resolvedType !== undefined) {
                return true;
            }
            if (prefix && findTypeInGlobalNamespace(program, fullRef)) {
                return true;
            }
        }
        // 2. Apply `using` semantics: `using X.Y.Z` makes the last segment Z a local alias for X.Y.Z.
        //    Strategy A — qualify: ref="Review", using "Components.ReviewApiService" → try "Components.ReviewApiService.Review"
        //                 (covers models/enums/ops that are members of a used namespace)
        //    Strategy B — expand alias: ref="ReviewApi.listReviews", using "Components.ReviewApiService"
        //                 → last segment of using is "ReviewApi" == first segment of ref
        //                 → expand to "Components.ReviewApiService.listReviews"
        //                 (covers refs where the namespace name itself is part of the ref)
        const usingPrefixes = getUsingPrefixes(contextNamespace);
        const refFirstSegment = ref.split(".")[0];
        const refRemainder = ref.indexOf(".") !== -1 ? ref.slice(ref.indexOf(".") + 1) : "";
        for (const usingPath of usingPrefixes) {
            // Strategy A: qualify ref as a member of the used namespace
            const qualifiedRef = `${usingPath}.${ref}`;
            const [resolvedA] = program.resolveTypeReference(qualifiedRef);
            if (resolvedA !== undefined) {
                return true;
            }
            if (findTypeInGlobalNamespace(program, qualifiedRef)) {
                return true;
            }
            // Strategy B: alias expansion — ref starts with the last segment of the using path
            const usingLastSegment = usingPath.split(".").at(-1);
            if (usingLastSegment === refFirstSegment) {
                const expanded = refRemainder ? `${usingPath}.${refRemainder}` : usingPath;
                const [resolvedB] = program.resolveTypeReference(expanded);
                if (resolvedB !== undefined) {
                    return true;
                }
                if (findTypeInGlobalNamespace(program, expanded)) {
                    return true;
                }
            }
        }
    }
    return false;
}
/**
 * Validates that a decorator is applied to the correct target type.
 * Reports diagnostic if target kind doesn't match.
 */
/**
 * Returns true when a namespace node uses block form `namespace X { }`.
 * Returns false for semicolon form `namespace X;`.
 * Decorators on semicolon-form namespaces are not indexed by the symbol indexer.
 */
// a10n:blueprint:ignore
export function isNamespaceBlockForm(ns) {
    if (!ns.node)
        return true; // synthetic — no constraint
    const loc = getSourceLocation(ns.node);
    if (!loc)
        return true;
    const span = loc.file.text.slice(loc.pos, loc.end);
    return span.includes("{");
}
/**
 * Validates that the namespace uses block form. Reports a diagnostic and returns
 * false when semicolon form is detected. Call this early in decorators that must
 * appear on indexable namespaces.
 */
// a10n:blueprint:ignore
export function validateNamespaceBlockForm(ctx, target, decoratorName) {
    if (!isNamespaceBlockForm(target)) {
        const fqn = (() => {
            const parts = [];
            let cur = target;
            while (cur && cur.name) {
                parts.unshift(cur.name);
                cur = cur.namespace;
            }
            return parts.join(".");
        })();
        reportDiagnostic(ctx.program, {
            code: "namespace-block-form-required",
            format: { decorator: decoratorName, fqn },
            target,
        });
        return false;
    }
    return true;
}
export function validateTarget(ctx, target, decoratorName, expectedKind) {
    const expectedKinds = Array.isArray(expectedKind) ? expectedKind : [expectedKind];
    if (!expectedKinds.includes(target.kind)) {
        const kindStr = expectedKinds.length === 1 ? expectedKinds[0] : expectedKinds.join(" or ");
        reportDiagnostic(ctx.program, {
            code: "invalid-decorator-target",
            format: {
                decorator: decoratorName,
                expected: kindStr,
                actual: target.kind,
            },
            target: ctx.decoratorTarget,
        });
        return false;
    }
    return true;
}
/**
 * Validates that a decorator is not applied multiple times to the same entity.
 */
export function validateDecoratorNotDuplicated(ctx, target, decoratorName) {
    if (!("decorators" in target)) {
        return true;
    }
    const sameDecorators = target.decorators.filter((d) => d.definition?.name === decoratorName);
    if (sameDecorators.length > 1) {
        reportDiagnostic(ctx.program, {
            code: "duplicate-decorator",
            format: { decorator: decoratorName },
            target: ctx.decoratorTarget,
        });
        return false;
    }
    return true;
}
/**
 * Validates that all items in an array are strings.
 */
// a10n:blueprint:ignore
export function validateStringArray(ctx, items, fieldName) {
    for (let i = 0; i < items.length; i++) {
        if (typeof items[i] !== "string") {
            reportDiagnostic(ctx.program, {
                code: "invalid-array-element-type",
                format: {
                    field: fieldName,
                    index: i.toString(),
                    expected: "string",
                    actual: typeof items[i],
                },
                target: ctx.getArgumentTarget(0) ?? ctx.decoratorTarget,
            });
            return false;
        }
    }
    return true;
}
/**
 * Validates that all items are of a specific type.
 */
// a10n:blueprint:ignore
export function validateArrayElementType(ctx, items, fieldName, expectedType, validator) {
    for (let i = 0; i < items.length; i++) {
        if (!validator(items[i])) {
            reportDiagnostic(ctx.program, {
                code: "invalid-array-element-type",
                format: {
                    field: fieldName,
                    index: i.toString(),
                    expected: expectedType,
                    actual: typeof items[i],
                },
                target: ctx.getArgumentTarget(0) ?? ctx.decoratorTarget,
            });
            return false;
        }
    }
    return true;
}
/**
 * Validates required fields in a config object.
 */
// a10n:blueprint:ignore
export function validateRequiredFields(ctx, config, requiredFields, configName) {
    const missing = requiredFields.filter((field) => !(field in config));
    if (missing.length > 0) {
        reportDiagnostic(ctx.program, {
            code: "missing-required-fields",
            format: {
                config: configName,
                fields: missing.join(", "),
            },
            target: ctx.getArgumentTarget(0) ?? ctx.decoratorTarget,
        });
        return false;
    }
    return true;
}
/**
 * Validates that deps are all Namespace types.
 */
// a10n:blueprint:ignore
export function validateNamespaceDeps(ctx, deps) {
    for (let i = 0; i < deps.length; i++) {
        const item = deps[i];
        if (!item || item.kind !== "Namespace") {
            reportDiagnostic(ctx.program, {
                code: "invalid-dependency-type",
                format: {
                    index: i.toString(),
                    expected: "Namespace",
                    actual: item?.kind ?? typeof deps[i],
                },
                target: ctx.getArgumentTarget(0) ?? ctx.decoratorTarget,
            });
            return false;
        }
    }
    return true;
}
/**
 * Validates that a value is a Model type.
 */
// a10n:blueprint:ignore
export function validateModelType(ctx, value, fieldName) {
    const item = value;
    if (!item || item.kind !== "Model") {
        reportDiagnostic(ctx.program, {
            code: "invalid-model-type",
            format: {
                field: fieldName,
                actual: item?.kind ?? typeof value,
            },
            target: ctx.getArgumentTarget(1) ?? ctx.decoratorTarget,
        });
        return false;
    }
    return true;
}
/**
 * Validates valid enum values.
 */
// a10n:blueprint:ignore
export function validateEnumValue(ctx, value, fieldName, validValues) {
    if (!validValues.includes(value)) {
        reportDiagnostic(ctx.program, {
            code: "invalid-enum-value",
            format: {
                field: fieldName,
                value,
                valid: validValues.join(", "),
            },
            target: ctx.getArgumentTarget(0) ?? ctx.decoratorTarget,
        });
        return false;
    }
    return true;
}
/**
 * Validates that a string is a valid URL.
 */
// a10n:blueprint:ignore
export function validateUrl(ctx, url, fieldName) {
    try {
        // Simple regex-based URL validation for TypeScript compatibility
        const urlPattern = /^https?:\/\/.+/;
        if (!urlPattern.test(url)) {
            throw new Error("Invalid URL");
        }
        return true;
    }
    catch {
        reportDiagnostic(ctx.program, {
            code: "invalid-url-format",
            format: {
                field: fieldName,
                value: url,
            },
            target: ctx.getArgumentTarget(0) ?? ctx.decoratorTarget,
        });
        return false;
    }
}
/**
 * Builds fully-qualified namespace name by walking parent chain.
 */
export function getNamespaceFqn(ns) {
    const parts = [];
    let current = ns;
    while (current && current.name !== "") {
        parts.unshift(current.name);
        current = current.namespace;
    }
    return parts.join(".");
}
/**
 * Checks if a namespace has a specific decorator applied.
 */
export function hasDecorator(ns, decoratorName) {
    const item = ns;
    return (item &&
        item.decorators &&
        item.decorators.some((d) => d.definition?.name === decoratorName));
}
/**
 * Checks if any ancestor namespace has a specific decorator applied.
 */
export function hasAncestorWithDecorator(ns, decoratorName) {
    let current = ns.namespace;
    while (current && current.name !== "") {
        if (hasDecorator(current, decoratorName))
            return true;
        current = current.namespace;
    }
    return false;
}
/** Returns the portion of a namespace FQN after stripping a given ancestor FQN prefix.
 *  e.g. stripNsPrefix("Products.ReviewBot.Features.PrReview.Tests.T001", "Products.ReviewBot.Features.PrReview") → "Tests.T001" */
export function stripNsPrefix(fqn, prefix) {
    if (fqn.startsWith(prefix + "."))
        return fqn.slice(prefix.length + 1);
    return fqn;
}
/** Finds the nearest ancestor namespace that has a given decorator, and returns it. */
export function findAncestorWithDecorator(ns, decoratorName) {
    let current = ns.namespace;
    while (current && current.name !== "") {
        if (hasDecorator(current, decoratorName))
            return current;
        current = current.namespace;
    }
    return undefined;
}
/**
 * Validates inline references in a string (e.g., {@TypeName} patterns).
 * Reports diagnostics for unresolved references.
 */
export function validateInlineRefs(ctx, str, fieldName, contextNamespace) {
    const refs = extractInlineRefs(str);
    let allResolved = true;
    for (const ref of refs) {
        if (!resolveRef(ctx.program, ref, contextNamespace)) {
            const suggestion = findSuggestion(ctx.program, ref);
            if (suggestion && suggestion !== ref) {
                reportDiagnostic(ctx.program, {
                    code: "unresolved-inline-ref",
                    messageId: "withSuggestion",
                    format: { ref, suggestion },
                    target: ctx.decoratorTarget,
                });
            }
            else {
                reportDiagnostic(ctx.program, {
                    code: "unresolved-inline-ref",
                    messageId: "default",
                    format: { ref },
                    target: ctx.decoratorTarget,
                });
            }
            allResolved = false;
        }
    }
    return allResolved;
}
export const OP_KIND_DECORATORS = new Set([
    "@tool", "@node", "@predict", "@interrupt", "@operation",
    "@signal", "@dataset", "@evalSuite", "@scoring", "@invariant",
    "@get", "@post", "@put", "@patch", "@delete", "@head",
    "@MCP.tool", "@subscribes",
]);
/** Returns the op-kind decorators present on an op (by name). */
export function getOpKindDecorators(op) {
    return op.decorators
        .map((d) => d.definition?.name ?? "")
        .filter((n) => n && OP_KIND_DECORATORS.has(n));
}
/** Returns true if op is inside a @uiRoute namespace (exempt from kind-decorator requirement — namespace is the kind signal). */
export function isInsideKindExemptNs(op) {
    let ns = op.namespace;
    while (ns && ns.name) {
        if ((ns.decorators ?? []).some((d) => d.definition?.name === "@uiRoute"))
            return true;
        ns = ns.namespace;
    }
    return false;
}
