// typespec-mcp/dist/src/decorators.js
import { useStateMap, useStateSet } from "@typespec/compiler/utils";
import {} from "../generated-defs/MCP.js";
import { stateKeys } from "./lib.js";
function createMarkerDecorator(key, validate) {
    const [is, mark] = useStateSet(key);
    const decorator = (...args) => {
        if (validate && !validate(...args)) {
            return;
        }
        const [context, target] = args;
        mark(context.program, target);
    };
    return [is, mark, decorator];
}
export const [isTool, markTool, toolDecorator] = createMarkerDecorator(stateKeys.tool);
export const [isReadonly, markReadonly, readonlyDecorator] = createMarkerDecorator(stateKeys.readonly);
export const [isNondestructive, markNondestructive, nondestructiveDecorator] = createMarkerDecorator(stateKeys.nondestructive);
export const [isIdempotent, markIdempotent, idempotentDecorator] = createMarkerDecorator(stateKeys.idempotent);
export const [isClosedWorld, markClosedWorld, closedWorldDecorator] = createMarkerDecorator(stateKeys.closedWorld);
export const [getSerializeAsText, setSerializeAsText] = useStateMap(stateKeys.serializeAsText);
export function $serializeAsText(context, target, dataType) {
    setSerializeAsText(context.program, target, {
        dataType,
    });
}
export const [getMcpServer, setMcpServer] = useStateMap(stateKeys.mcpServer);
export const $mcpServer = (context, target, options = {}) => {
    setMcpServer(context.program, target, {
        ...options,
        container: target,
    });
};
export const [getResource, setResource] = useStateMap(stateKeys.resource);
export const $resource = (context, target, uri) => {
    setResource(context.program, target, {
        uri: uri ?? "",
    });
};
