// @a10n/blueprint-typespec/src/projecting-op.ts
import { Namespace, Operation } from "@typespec/compiler";

// ── Local helpers (not re-exported from validators to keep scope narrow) ──────

function hasDecorator(node: { decorators: readonly { definition?: { name: string } }[] }, name: string): boolean {
  return node.decorators.some((d) => d.definition?.name === name);
}

function hasAncestorNsNamed(op: Operation, nsName: string): boolean {
  let ns: Namespace | undefined = op.namespace;
  while (ns && ns.name) {
    if (ns.name === nsName) return true;
    ns = ns.namespace;
  }
  return false;
}

/** Returns the @component type string ("Sdk", "Agent", "Service", "Ui", …) for a namespace, or undefined. */
export function getComponentType(ns: Namespace): string | undefined {
  for (const dec of (ns as any).decorators ?? []) {
    if (dec.definition?.name !== "@component") continue;
    const arg = dec.args?.[0]?.value;
    if (!arg) continue;
    let typeVal: any;
    if (arg.valueKind === "ObjectValue") {
      typeVal = (arg.properties as Map<string, any>)?.get("type")?.value;
    } else if (arg.kind === "Object") {
      typeVal = (arg as any).properties?.type;
    }
    if (!typeVal) continue;
    return typeVal.valueKind === "EnumValue" ? typeVal.value?.name : typeVal.value?.name ?? String(typeVal);
  }
  return undefined;
}

/** Walks ancestor namespaces of op until a @component ns is found; returns it or undefined. */
export function findComponentAncestor(op: Operation): Namespace | undefined {
  let ns: Namespace | undefined = op.namespace;
  while (ns && ns.name) {
    if (hasDecorator(ns, "@component")) return ns;
    ns = ns.namespace;
  }
  return undefined;
}

/** Returns true if the op lives inside a namespace named `nsName` at any depth. */
export function hasAncestorNamespaceNamed(op: Operation, nsName: string): boolean {
  return hasAncestorNsNamed(op, nsName);
}

/**
 * Returns true for the 10 SpecConstructKinds that may carry @projects(invariant) and count as proof.
 *
 * Kinds (matching Go side's entryPointSpecLikeOnFQN / canonicalKindPatterns):
 *   http_op   — op inside an Http ns within a @component (FQN Components.%.Http.%)
 *   mcp_op    — op inside an Mcp ns within a @component (FQN Components.%.Mcp.%)
 *   grpc_op   — op inside a Grpc ns within a @component (FQN Components.%.Grpc.%)
 *   cli_op    — @operation op in a Commands ns inside a Service @component
 *   ui_action — op inside a @uiRoute namespace
 *   sdk_op    — @operation op inside an Sdk @component (lenient: regardless of Operations sub-ns)
 *   graph_node — @node op inside a @graph ns
 *   predict_op — @predict op inside a @graph ns
 *   worker_op  — op inside a @worker ns (the @worker decorator targets a Namespace)
 *   interrupt  — op with @interrupt
 *   tool       — @tool op inside a @component (agent tool-file-path is enforced by @tool)
 *
 * Boundary ops (http/mcp/grpc) are namespace-path classified because the
 * @typespec/http and typespec-mcp verb decorators do not surface in op.decorators.
 */
export function isProjectingOpKind(op: Operation): boolean {
  // interrupt — @interrupt on the op directly
  if (hasDecorator(op, "@interrupt")) return true;

  // tool — @tool op inside a @component. The @tool decorator already enforces an agent
  // tool-file-path (components/*-agent/tools/*), so we don't additionally gate on the
  // @component type being "Agent" — that would create a false negative for an agent
  // component whose type enum drifted. Mirrors the Go-side leniency (Components.%.Tools.%
  // OR Components.%Agent.%), both of which classify as tool.
  if (hasDecorator(op, "@tool") && findComponentAncestor(op)) return true;

  // graph_node — @node op inside a @graph ns
  if (hasDecorator(op, "@node")) {
    let ns: Namespace | undefined = op.namespace;
    while (ns && ns.name) {
      if (hasDecorator(ns, "@graph")) return true;
      ns = ns.namespace;
    }
  }

  // predict_op — @predict op inside a @graph ns
  if (hasDecorator(op, "@predict")) {
    let ns: Namespace | undefined = op.namespace;
    while (ns && ns.name) {
      if (hasDecorator(ns, "@graph")) return true;
      ns = ns.namespace;
    }
  }

  // worker_op — op inside a @worker namespace (the @worker decorator targets a Namespace,
  // so a worker op is one whose ancestor namespace carries @worker)
  {
    let ns: Namespace | undefined = op.namespace;
    while (ns && ns.name) {
      if (hasDecorator(ns, "@worker")) return true;
      ns = ns.namespace;
    }
  }

  // ui_action — op inside a @uiRoute namespace (at any ancestor depth)
  {
    let ns: Namespace | undefined = op.namespace;
    while (ns && ns.name) {
      if (hasDecorator(ns, "@uiRoute")) return true;
      ns = ns.namespace;
    }
  }

  // sdk_op — @operation op inside an Sdk @component (lenient: flat OR inside Operations ns)
  if (hasDecorator(op, "@operation")) {
    const comp = findComponentAncestor(op);
    if (comp && getComponentType(comp) === "Sdk") return true;
  }

  // http_op / mcp_op / grpc_op — boundary ops are classified by their namespace path,
  // not by their verb decorator. The @typespec/http (@get/@post/…) and typespec-mcp
  // (@MCP.tool) decorators are applied by external libraries and do NOT surface in the
  // linter-visible op.decorators array, so a decorator check would always miss them.
  // This mirrors the Go-side FQN patterns Components.%.Http.% / .Mcp.% / .Grpc.%:
  // an op inside an Http/Mcp/Grpc namespace within a @component is a boundary entry point.
  if (findComponentAncestor(op)) {
    let ns: Namespace | undefined = op.namespace;
    while (ns && ns.name) {
      if (ns.name === "Http" || ns.name === "Mcp" || ns.name === "Grpc") return true;
      ns = ns.namespace;
    }
  }

  // cli_op — @operation op in a Commands namespace inside a Service @component.
  // Mirrors the Go-side pattern `Components.%Service.Commands.%` and
  // operation-reachability's isInCommandsNs (compType === "Service").
  if (hasDecorator(op, "@operation") && hasAncestorNsNamed(op, "Commands")) {
    const comp = findComponentAncestor(op);
    if (comp && getComponentType(comp) === "Service") return true;
  }

  return false;
}
