// @a10n/blueprint-typespec/dist/src/rules/calls-graph.rule.js
import { createRule, getDoc, paramMessage } from "@typespec/compiler";
import { extractInlineRefs, resolveInlineRef } from "../validators.js";
import { getTestingStrategies } from "../decorators/component.js";
function hasDecorator(node, name) {
    return (node?.decorators ?? []).some((d) => d.definition?.name === name);
}
function isInComponent(op) {
    let ns = op.namespace;
    while (ns && ns.name) {
        if (hasDecorator(ns, "@component"))
            return true;
        ns = ns.namespace;
    }
    return false;
}
function nsToFqn(ns) {
    const parts = [];
    let cur = ns;
    while (cur && cur.name) {
        parts.unshift(cur.name);
        cur = cur.namespace;
    }
    return parts.join(".");
}
// Find the nearest @component ancestor of an Operation.
function findComponentAncestor(op) {
    let ns = op.namespace;
    while (ns && ns.name) {
        if (hasDecorator(ns, "@component"))
            return ns;
        ns = ns.namespace;
    }
    return undefined;
}
// Find the owning @component of a Model.
function findModelComponent(model) {
    let ns = model.namespace;
    while (ns && ns.name) {
        if (hasDecorator(ns, "@component"))
            return ns;
        ns = ns.namespace;
    }
    return undefined;
}
function getCallGraphTargetKeys(op) {
    const keys = new Set();
    for (const dec of op.decorators) {
        if (!["@calls", "@reads", "@writes"].includes(dec.definition?.name ?? ""))
            continue;
        for (const arg of dec.args) {
            const v = arg.value;
            const items = v?.valueKind === "ArrayValue" ? v.values : (v?.kind === "Tuple" ? v.values : [v]);
            for (const item of items ?? []) {
                const t = item;
                if (!t?.name)
                    continue;
                keys.add(t.name);
                if (t.namespace)
                    keys.add(`${nsToFqn(t.namespace)}.${t.name}`);
            }
        }
    }
    return keys;
}
// Collect all @calls / @reads / @writes targets (typed Operation or Model refs only).
function getCallGraphTargets(op) {
    const result = [];
    for (const dec of op.decorators) {
        if (!["@calls", "@reads", "@writes"].includes(dec.definition?.name ?? ""))
            continue;
        for (const arg of dec.args) {
            const v = arg.value;
            const items = v?.valueKind === "ArrayValue" ? v.values : (v?.kind === "Tuple" ? v.values : [v]);
            for (const item of items ?? []) {
                const t = item;
                if (t?.kind === "Operation")
                    result.push(t);
                else if (t?.kind === "Model")
                    result.push(t);
            }
        }
    }
    return result;
}
export const callsGraphRule = createRule({
    name: "calls-graph",
    severity: "warning",
    description: "@doc on an op must not contain {@...} references to its own call-graph targets.",
    messages: {
        default: paramMessage `@doc on "${"opName"}" contains {@${"ref"}} which is a declared @calls/@reads/@writes target — remove the restatement; the decorator already declares this relationship.`,
        missingTesting: paramMessage `Op "${"opName"}" declares a call-graph edge to "${"targetFqn"}" in foreign component "${"targetComponent"}", but the owning component "${"sourceComponent"}" has no @testing declaration for "${"targetComponent"}". Add @testing(${"targetComponent"}, #{...}) to the source component to declare whether the dep is mocked or run real.`,
    },
    // a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.rule_calls_graph
    create(context) {
        return {
            // a10n:index:call packages.blueprint-typespec.src.rules.calls-graph.rule.operation
            operation(op) {
                if (!isInComponent(op))
                    return;
                const targets = getCallGraphTargetKeys(op);
                // ── runtime_within_deps: every cross-component edge target's component ──
                // ── must be the subject of a @testing declaration on the source component ──
                // a10n:blueprint Products.A10nBlueprintEngine.Features.OperationCallGraph.runtime_within_deps
                const sourceComp = findComponentAncestor(op);
                if (sourceComp) {
                    const strategies = getTestingStrategies(context.program, sourceComp) ?? [];
                    // Match by FQN, not identity — TypeSpec may surface different Namespace
                    // instances for the same logical namespace across decorator args vs
                    // ancestor walks of a model's parents.
                    const declaredTestingDepFqns = new Set(strategies.map((s) => nsToFqn(s.dep)));
                    const reportedForComp = new Set();
                    for (const t of getCallGraphTargets(op)) {
                        const targetComp = t.kind === "Operation"
                            ? findComponentAncestor(t)
                            : findModelComponent(t);
                        if (!targetComp)
                            continue;
                        if (targetComp === sourceComp)
                            continue; // same-component edges are intra-component
                        const targetCompFqn = nsToFqn(targetComp);
                        if (declaredTestingDepFqns.has(targetCompFqn))
                            continue;
                        if (reportedForComp.has(targetCompFqn))
                            continue;
                        reportedForComp.add(targetCompFqn);
                        const targetName = t.name ?? "<anon>";
                        const targetFqn = t.namespace
                            ? `${nsToFqn(t.namespace)}.${targetName}`
                            : targetName;
                        context.reportDiagnostic({
                            messageId: "missingTesting",
                            format: {
                                opName: op.name,
                                targetFqn,
                                targetComponent: targetCompFqn,
                                sourceComponent: nsToFqn(sourceComp),
                            },
                            target: op,
                        });
                    }
                }
                if (targets.size === 0)
                    return;
                const docStr = getDoc(context.program, op);
                if (!docStr)
                    return;
                for (const ref of extractInlineRefs(docStr)) {
                    if (targets.has(ref)) {
                        // a10n:blueprint Products.A10nBlueprintEngine.Features.OperationCallGraph.doc_omits_facts
                        context.reportDiagnostic({ messageId: "default", format: { opName: op.name, ref }, target: op });
                        return;
                    }
                    const resolved = resolveInlineRef(context.program, ref, op.namespace);
                    if (resolved?.name) {
                        const fqn = resolved.namespace
                            ? `${nsToFqn(resolved.namespace)}.${resolved.name}`
                            : resolved.name;
                        if (targets.has(fqn) || targets.has(resolved.name)) {
                            // a10n:blueprint Products.A10nBlueprintEngine.Features.OperationCallGraph.doc_omits_facts
                            context.reportDiagnostic({ messageId: "default", format: { opName: op.name, ref }, target: op });
                            return;
                        }
                    }
                }
            },
        };
    },
});
