// @a10n/blueprint-typespec/dist/src/rules/dependency-authoring.rule.js
import { createRule, paramMessage } from "@typespec/compiler";
function hasDecorator(node, name) {
    return (node?.decorators ?? []).some((d) => d.definition?.name === name);
}
function nsToFqn(ns) {
    const parts = [];
    let cur = ns;
    while (cur && cur.name) {
        parts.unshift(cur.name);
        cur = cur.namespace;
    }
    return parts.join(".");
}
/** Return the @component type string ("Sdk", "Service", "Agent", etc.) or undefined. */
function getComponentType(ns) {
    for (const dec of ns.decorators ?? []) {
        if (dec.definition?.name !== "@component")
            continue;
        const arg = dec.args?.[0]?.value;
        if (!arg)
            continue;
        let typeVal;
        if (arg.valueKind === "ObjectValue") {
            typeVal = arg.properties?.get("type")?.value;
        }
        else if (arg.kind === "Object") {
            typeVal = arg.properties?.type;
        }
        if (!typeVal)
            continue;
        return typeVal.valueKind === "EnumValue" ? typeVal.value?.name : typeVal.value?.name ?? String(typeVal);
    }
    return undefined;
}
/** Walk all namespaces reachable from root, calling visitor for each (depth-first). */
function walkAllNamespaces(ns, visitor) {
    visitor(ns);
    for (const [, child] of ns.namespaces)
        walkAllNamespaces(child, visitor);
}
export const dependencyAuthoringRule = createRule({
    name: "dependency-authoring",
    severity: "warning",
    description: "Enforces structural rules for @dependency namespaces: component kinds and uniform kind.",
    messages: {
        depComponentKind: paramMessage `Component "${"ref"}" inside @dependency "${"dep"}" has type ${"type"} — only Sdk and Service components are permitted inside dependencies.`,
        depUniformKind: paramMessage `@dependency "${"dep"}" mixes Sdk and Service components — all components in a dependency must be the same kind.`,
    },
    // a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.rule_dependency_authoring
    // a10n:blueprint Products.A10nBlueprintEngine.Features.DependencyAuthoring.dep_component_kind
    // a10n:blueprint Products.A10nBlueprintEngine.Features.DependencyAuthoring.dep_uniform_kind
    create(context) {
        return {
            // a10n:index:call packages.blueprint-typespec.src.rules.dependency-authoring.rule.namespace
            namespace(ns) {
                // ── dep-component-kind + dep-uniform-kind + dep-no-tests ──────────────
                if (hasDecorator(ns, "@dependency")) {
                    const depFqn = nsToFqn(ns);
                    const componentKinds = [];
                    // Walk all nested namespaces (but not the dependency itself)
                    // a10n:blueprint Products.A10nBlueprintEngine.Features.DependencyAuthoring.dep_addresses_components
                    // a10n:blueprint Products.A10nBlueprintEngine.Features.DependencyAuthoring.dep_no_tests
                    for (const [, child] of ns.namespaces) {
                        walkAllNamespaces(child, (inner) => {
                            if (!hasDecorator(inner, "@component"))
                                return;
                            const kind = getComponentType(inner);
                            const innerFqn = nsToFqn(inner);
                            // dep-component-kind: only Sdk and Service allowed
                            if (kind !== "Sdk" && kind !== "Service") {
                                context.reportDiagnostic({
                                    messageId: "depComponentKind",
                                    format: { ref: innerFqn, dep: depFqn, type: kind ?? "unknown" },
                                    target: inner,
                                });
                            }
                            else {
                                componentKinds.push(kind);
                            }
                        });
                    }
                    // dep-uniform-kind: only Sdk or only Service, not both
                    const hasSdk = componentKinds.includes("Sdk");
                    const hasSvc = componentKinds.includes("Service");
                    if (hasSdk && hasSvc) {
                        context.reportDiagnostic({
                            messageId: "depUniformKind",
                            format: { dep: depFqn },
                            target: ns,
                        });
                    }
                }
            },
        };
    },
});
