// @a10n/blueprint-typespec/dist/src/rules/entity-field-reachable.rule.js
import { createRule, paramMessage, } from "@typespec/compiler";
const INLINE_REF_REGEX = /\{@([\w.]+)\}/g;
function extractInlineRefs(str) {
    const refs = [];
    let match;
    INLINE_REF_REGEX.lastIndex = 0;
    while ((match = INLINE_REF_REGEX.exec(str)) !== null) {
        refs.push(match[1]);
    }
    return refs;
}
function hasDecorator(node, name) {
    return node.decorators.some((d) => d.definition?.name === name);
}
function nsToFqn(ns) {
    const parts = [];
    let cur = ns;
    while (cur && cur.name) {
        parts.unshift(cur.name);
        cur = cur.namespace;
    }
    return parts.join(".");
}
function modelFqn(model) {
    const ns = model.namespace ? nsToFqn(model.namespace) : "";
    return ns ? `${ns}.${model.name}` : (model.name ?? "");
}
function propFqn(prop) {
    return `${modelFqn(prop.model)}.${prop.name}`;
}
/** Extract all string values from decorator args (strings, arrays, objects). */
function collectStringsFromValue(v, out) {
    if ("entityKind" in v) {
        const val = v;
        if (val.valueKind === "StringValue") {
            out.push(val.value);
            return;
        }
        if (val.valueKind === "ArrayValue") {
            for (const item of val.values)
                collectStringsFromValue(item, out);
            return;
        }
        if (val.valueKind === "ObjectValue") {
            for (const [, p] of val.properties)
                collectStringsFromValue(p.value, out);
            return;
        }
        return;
    }
    const t = v;
    if (t.kind === "String") {
        out.push(t.value);
        return;
    }
    if (t.kind === "Tuple") {
        for (const item of t.values)
            collectStringsFromValue(item, out);
    }
}
/**
 * Collect the set of entity field FQNs that are referenced in @invariant predicates.
 * A predicate token like "Task.id" or "Products.X.Entities.Task.Task.id" must be
 * matched against the property FQN — we collect every token that ends with ".<name>"
 * where <name> is a field name of some entity model, and cross-reference against
 * known entity field FQNs via suffix matching (same approach as mentions resolution).
 */
function collectReferencedFieldTokens(program) {
    const tokens = new Set();
    const global = program.getGlobalNamespaceType();
    function walkNs(ns) {
        for (const [, op] of ns.operations) {
            if (!hasDecorator(op, "@invariant"))
                continue;
            for (const dec of op.decorators) {
                if (dec.definition?.name !== "@invariant")
                    continue;
                const strings = [];
                for (const arg of dec.args)
                    collectStringsFromValue(arg.value, strings);
                for (const str of strings) {
                    for (const ref of extractInlineRefs(str)) {
                        tokens.add(ref);
                    }
                }
            }
        }
        for (const [, child] of ns.namespaces)
            walkNs(child);
    }
    walkNs(global);
    return tokens;
}
/**
 * Collect entity field FQNs referenced as @projects targets on ModelProperty nodes.
 * @projects(Entity.field) on a field → the arg is a ModelProperty type.
 */
function collectProjectsTargetFields(program) {
    const targeted = new Set();
    const global = program.getGlobalNamespaceType();
    function walkNs(ns) {
        for (const [, model] of ns.models) {
            for (const [, prop] of model.properties) {
                for (const dec of prop.decorators) {
                    if (dec.definition?.name !== "@projects")
                        continue;
                    for (const arg of dec.args) {
                        const v = arg.value;
                        // Each arg is a ModelProperty (entity field) target
                        if (v?.kind === "ModelProperty" && v.model) {
                            targeted.add(propFqn(v));
                        }
                    }
                }
            }
        }
        for (const [, child] of ns.namespaces)
            walkNs(child);
    }
    walkNs(global);
    return targeted;
}
/**
 * Returns true if the token (from an @invariant predicate) references a given
 * entity field FQN. Matches by suffix: token "Task.id" matches FQN
 * "Products.X.Entities.Task.Task.id" because the token is a suffix of the FQN
 * when split by ".".
 */
function tokenMatchesPropFqn(token, fqn) {
    if (token === fqn)
        return true;
    // Suffix match: FQN ends with "."+token
    return fqn.endsWith(`.${token}`);
}
export const entityFieldReachableRule = createRule({
    name: "entity-field-reachable",
    severity: "warning",
    description: "Every field of an @entity model must be referenced by at least one @invariant predicate or @projects target field.",
    messages: {
        default: paramMessage `Field "${"field"}" of @entity "${"entity"}" is not referenced by any @invariant predicate or @projects target. Every entity field must be reachable from the domain layer.`,
    },
    // a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.rule_entity_field_reachable
    // a10n:blueprint Products.A10nBlueprintEngine.Features.DomainAuthoring.entity_field_reachable
    // a10n:todo Products.A10nBlueprintEngine.Features.DomainAuthoring.entity_primitive_constrained — not yet implemented: no rule checks that bare primitive @entity fields have refinement decorators or invariant coverage
    // a10n:todo Products.A10nBlueprintEngine.Features.DomainAuthoring.field_refined — not yet implemented: no rule enforces that narrower-value-space fields use refined types rather than bare primitives
    // a10n:todo Products.A10nBlueprintEngine.Features.DomainAuthoring.refinement_implicit — not yet implemented: no rule detects that a refinement decorator is duplicated as a separate @invariant op
    // a10n:todo Products.A10nBlueprintEngine.Features.DomainAuthoring.refined_type_alias — not yet implemented: no rule enforces that repeated inline refinements across fields are extracted to a named alias
    // a10n:todo Products.A10nBlueprintEngine.Features.DomainAuthoring.enum_with_entity — not yet implemented: no rule verifies that enums used by @entity fields are declared in the same .tsp file
    create(context) {
        let invariantTokens = null;
        let projectsTargets = null;
        return {
            // a10n:index:call packages.blueprint-typespec.src.rules.entity-field-reachable.rule.model
            model(model) {
                if (!hasDecorator(model, "@entity"))
                    return;
                if (!model.name)
                    return;
                if (!invariantTokens)
                    invariantTokens = collectReferencedFieldTokens(context.program);
                if (!projectsTargets)
                    projectsTargets = collectProjectsTargetFields(context.program);
                const entityFqn = modelFqn(model);
                for (const [, prop] of model.properties) {
                    const fqn = `${entityFqn}.${prop.name}`;
                    // Check 1: referenced by an @invariant predicate token
                    const inInvariant = [...invariantTokens].some((token) => tokenMatchesPropFqn(token, fqn));
                    if (inInvariant)
                        continue;
                    // Check 2: targeted by @projects on a model field somewhere
                    if (projectsTargets.has(fqn))
                        continue;
                    context.reportDiagnostic({
                        messageId: "default",
                        format: { field: prop.name, entity: entityFqn },
                        target: prop,
                    });
                }
            },
        };
    },
});
