// @a10n/blueprint-typespec/dist/src/rules/entity-primitive-constrained.rule.js
import { createRule, getFormat, getMaxItems, getMaxLength, getMaxValue, getMaxValueExclusive, getMinItems, getMinLength, getMinValue, getMinValueExclusive, getPattern, isStdNamespace, paramMessage, } from "@typespec/compiler";
import { extractInlineRefs } from "../validators.js";
/**
 * The set of TypeSpec built-in scalar names that count as "bare primitives".
 *
 * Date/time scalars (utcDateTime, offsetDateTime, plainDate, plainTime, duration) are
 * intentionally excluded: value-refinement decorators (getMinValue/getMaxValue) return
 * undefined for date/time literals, so even a syntactically-valid @minValue on a
 * datetime field never registers as a constraint. Flagging datetime fields would produce
 * false positives with no actionable fix.
 */
const BARE_PRIMITIVE_NAMES = new Set([
    "string", "boolean", "numeric", "integer", "float",
    "int8", "int16", "int32", "int64",
    "uint8", "uint16", "uint32", "uint64", "safeint",
    "float32", "float64", "decimal", "decimal128",
    "bytes", "url",
]);
function hasDecorator(node, name) {
    return node.decorators.some((d) => d.definition?.name === name);
}
/**
 * Returns true if the given type is a bare TypeSpec intrinsic scalar (not an
 * enum, not a user-defined named alias/model, not another @entity).
 * The check walks the scalar's base chain — a user alias like `scalar MyId extends string`
 * is NOT considered bare (it already carries a name that adds domain meaning).
 */
function isBarePrimitiveType(type) {
    if (type.kind !== "Scalar")
        return false;
    const scalar = type;
    // The scalar is bare if it IS one of the built-in intrinsics (its namespace is the
    // TypeSpec built-in namespace checked via isStdNamespace) or its name is in our set.
    // We must also confirm it is not a user-defined named scalar (those live in non-std namespaces).
    if (!scalar.namespace) {
        // No namespace — only intrinsics can be here (e.g. string, boolean at root)
        return BARE_PRIMITIVE_NAMES.has(scalar.name);
    }
    if (isStdNamespace(scalar.namespace)) {
        return BARE_PRIMITIVE_NAMES.has(scalar.name);
    }
    // User-defined scalar alias — not bare, it has semantic meaning
    return false;
}
/** Returns true if the field carries at least one constraint decorator. */
function hasConstraintDecorator(program, prop) {
    // Check for well-known TypeSpec constraint decorators on the property itself
    if (getPattern(program, prop) !== undefined)
        return true;
    if (getFormat(program, prop) !== undefined)
        return true;
    if (getMinLength(program, prop) !== undefined)
        return true;
    if (getMaxLength(program, prop) !== undefined)
        return true;
    if (getMinValue(program, prop) !== undefined)
        return true;
    if (getMaxValue(program, prop) !== undefined)
        return true;
    if (getMinValueExclusive(program, prop) !== undefined)
        return true;
    if (getMaxValueExclusive(program, prop) !== undefined)
        return true;
    if (getMinItems(program, prop) !== undefined)
        return true;
    if (getMaxItems(program, prop) !== undefined)
        return true;
    return false;
}
/** Collect all string values from decorator args recursively. */
function collectStringsFromValue(v, out) {
    if ("entityKind" in v) {
        const val = v;
        if (val.valueKind === "StringValue") {
            out.push(val.value);
            return;
        }
        if (val.valueKind === "ArrayValue") {
            for (const item of val.values)
                collectStringsFromValue(item, out);
            return;
        }
        if (val.valueKind === "ObjectValue") {
            for (const [, p] of val.properties)
                collectStringsFromValue(p.value, out);
            return;
        }
        return;
    }
    const t = v;
    if (t.kind === "String") {
        out.push(t.value);
        return;
    }
    if (t.kind === "Tuple") {
        for (const item of t.values)
            collectStringsFromValue(item, out);
    }
}
/**
 * Collect the set of inline-ref tokens found inside @invariant predicate strings.
 * These are tokens like "Task.id" or "Task" that appear in {@...} refs.
 */
function collectInvariantRefTokens(program) {
    const tokens = new Set();
    const global = program.getGlobalNamespaceType();
    function walkNs(ns) {
        for (const [, op] of ns.operations) {
            if (!hasDecorator(op, "@invariant"))
                continue;
            for (const dec of op.decorators) {
                if (dec.definition?.name !== "@invariant")
                    continue;
                const strings = [];
                for (const arg of dec.args)
                    collectStringsFromValue(arg.value, strings);
                for (const str of strings) {
                    for (const ref of extractInlineRefs(str)) {
                        tokens.add(ref);
                    }
                }
            }
        }
        for (const [, child] of ns.namespaces)
            walkNs(child);
    }
    walkNs(global);
    return tokens;
}
function nsToFqn(ns) {
    const parts = [];
    let cur = ns;
    while (cur && cur.name) {
        parts.unshift(cur.name);
        cur = cur.namespace;
    }
    return parts.join(".");
}
function modelFqn(model) {
    const ns = model.namespace ? nsToFqn(model.namespace) : "";
    return ns ? `${ns}.${model.name}` : (model.name ?? "");
}
/**
 * Returns true if any @invariant predicate token references the given entity field.
 * Matches by suffix: token "Task.id" matches FQN "Products.X.Task.id".
 */
function isFieldCoveredByInvariant(tokens, entityFqn, fieldName) {
    const fqn = `${entityFqn}.${fieldName}`;
    for (const token of tokens) {
        if (token === fqn || fqn.endsWith(`.${token}`))
            return true;
    }
    return false;
}
export const entityPrimitiveConstrainedRule = createRule({
    name: "entity-primitive-constrained",
    severity: "warning",
    description: "Every bare-primitive field on an @entity must carry at least one constraint decorator " +
        "(@pattern, @minLength, @maxLength, @minValue, @maxValue, @format, etc.) " +
        "or be referenced by an @invariant predicate via {@Entity.field}.",
    messages: {
        default: paramMessage `@entity "${"entity"}" field "${"field"}" is a bare ${"typeName"} with no constraint decorator and no @invariant predicate reference. Add @pattern/@minLength/@maxValue/etc. or reference it in an @invariant predicate as {@${"fieldRef"}}.`,
    },
    // a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.rule_entity_primitive_constrained
    // a10n:blueprint Products.A10nBlueprintEngine.Features.DomainAuthoring.entity_primitive_constrained
    create(context) {
        let invariantTokens = null;
        return {
            // a10n:index:call packages.blueprint-typespec.src.rules.entity-primitive-constrained.rule.model
            model(model) {
                if (!hasDecorator(model, "@entity"))
                    return;
                if (!model.name)
                    return;
                if (!invariantTokens)
                    invariantTokens = collectInvariantRefTokens(context.program);
                const entityFqn = modelFqn(model);
                for (const [, prop] of model.properties) {
                    if (!isBarePrimitiveType(prop.type))
                        continue;
                    // Check 1: has a constraint decorator
                    if (hasConstraintDecorator(context.program, prop))
                        continue;
                    // Check 2: referenced by an @invariant predicate
                    if (isFieldCoveredByInvariant(invariantTokens, entityFqn, prop.name))
                        continue;
                    const typeName = prop.type.name;
                    // a10n:blueprint Products.A10nBlueprintEngine.Features.DomainAuthoring.entity_primitive_constrained
                    context.reportDiagnostic({
                        messageId: "default",
                        format: {
                            entity: entityFqn,
                            field: prop.name,
                            typeName,
                            fieldRef: `${model.name}.${prop.name}`,
                        },
                        target: prop,
                    });
                }
            },
        };
    },
});
