// @a10n/blueprint-typespec/dist/src/rules/feature-invariant-cap.rule.js
import { createRule, paramMessage } from "@typespec/compiler";
import { getNamespaceFqn } from "../validators.js";
/** Returns true if the namespace has @feature decorator applied directly. */
function hasFeatureDecorator(ns) {
    return ns.decorators.some((d) => d.definition?.name === "@feature");
}
/** Count ops with @invariant decorator in this namespace. */
function countInvariantOps(ns) {
    let count = 0;
    for (const op of ns.operations.values()) {
        if (op.decorators.some((d) => d.definition?.name === "@invariant")) {
            count++;
        }
    }
    return count;
}
export const featureInvariantCapRule = createRule({
    name: "feature-invariant-cap",
    severity: "warning",
    description: "A @feature namespace must not declare more than 20 @invariant operations. Split into two feature files by capability.",
    messages: {
        default: paramMessage `Feature "${"featureFqn"}" has ${"count"} invariants; the maximum is 20. Split into two feature files by capability.`,
    },
    // a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.rule_feature_invariant_cap
    create(context) {
        // a10n:index:call Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.feature_invariant_cap
        return {
            // a10n:index:call packages.blueprint-typespec.src.rules.feature-invariant-cap.rule.namespace
            namespace(ns) {
                if (!hasFeatureDecorator(ns)) {
                    return;
                }
                const count = countInvariantOps(ns);
                // a10n:blueprint Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.feature_invariant_cap
                if (count > 20) {
                    context.reportDiagnostic({
                        messageId: "default",
                        format: {
                            featureFqn: getNamespaceFqn(ns),
                            count: String(count),
                        },
                        target: ns,
                    });
                }
            },
        };
    },
});
