// @a10n/blueprint-typespec/dist/src/rules/invariant-missing-proof.rule.js
import { createRule, paramMessage, } from "@typespec/compiler";
import { isProjectingOpKind } from "../projecting-op.js";
function nsToFqn(ns) {
    const parts = [];
    let cur = ns;
    while (cur && cur.name) {
        parts.unshift(cur.name);
        cur = cur.namespace;
    }
    return parts.join(".");
}
function opFqn(op) {
    const ns = op.namespace ? nsToFqn(op.namespace) : "";
    return ns ? `${ns}.${op.name}` : op.name;
}
function hasDecorator(node, name) {
    return node.decorators.some((d) => d.definition?.name === name);
}
/**
 * Collect the set of invariant FQNs that are proven by:
 *   - any op of a isProjectingOpKind that @projects them, OR
 *   - any @signal op that @projects them.
 */
function collectProvenInvariants(program) {
    const proven = new Set();
    const global = program.getGlobalNamespaceType();
    function collectFromOp(op) {
        if (!isProjectingOpKind(op) && !hasDecorator(op, "@signal"))
            return;
        for (const dec of op.decorators) {
            if (dec.definition?.name !== "@projects")
                continue;
            for (const arg of dec.args) {
                const v = arg.value;
                if (v?.kind === "Operation") {
                    proven.add(opFqn(v));
                }
            }
        }
    }
    function walkNs(ns) {
        for (const [, op] of ns.operations)
            collectFromOp(op);
        // http_op / mcp_op live inside a TypeSpec interface, not directly in the namespace.
        for (const [, iface] of ns.interfaces) {
            for (const [, op] of iface.operations)
                collectFromOp(op);
        }
        for (const [, child] of ns.namespaces) {
            walkNs(child);
        }
    }
    walkNs(global);
    return proven;
}
export const invariantMissingProofRule = createRule({
    name: "invariant-missing-proof",
    severity: "warning",
    description: "Every @invariant must be targeted by at least one projecting-op-kind op (or @signal) that declares @projects(...) listing it.",
    messages: {
        default: paramMessage `@invariant "${"fqn"}" has no operation or @signal projecting onto it. Add at least one proof.`,
    },
    // a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.rule_invariant_missing_proof
    create(context) {
        let proven = null;
        // a10n:index:call Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.inv_has_proof
        return {
            // a10n:index:call packages.blueprint-typespec.src.rules.invariant-missing-proof.rule.operation
            operation(op) {
                if (!hasDecorator(op, "@invariant"))
                    return;
                if (!proven) {
                    proven = collectProvenInvariants(context.program);
                }
                const fqn = opFqn(op);
                // a10n:blueprint Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.inv_has_proof
                // a10n:blueprint Products.A10nBlueprintEngine.Features.SystemAuthoring.proof_targets_invariant
                if (!proven.has(fqn)) {
                    context.reportDiagnostic({
                        messageId: "default",
                        format: { fqn },
                        target: op,
                    });
                }
            },
        };
    },
});
