// @a10n/blueprint-typespec/dist/src/rules/no-orphan-invariant-layer.rule.js
import { createRule, paramMessage } from "@typespec/compiler";
function opFqn(op) {
    const parts = [];
    let ns = op.namespace;
    while (ns && ns.name) {
        parts.unshift(ns.name);
        ns = ns.namespace;
    }
    return parts.length ? `${parts.join(".")}.${op.name}` : op.name;
}
function hasDecorator(ns, name) {
    return (ns.decorators ?? []).some((d) => d.definition?.name === name);
}
/** Returns true if any ancestor namespace has the name "Components". */
function isUnderComponentsNamespace(op) {
    let ns = op.namespace;
    while (ns && ns.name) {
        if (ns.name === "Components")
            return true;
        ns = ns.namespace;
    }
    return false;
}
/** Returns true if there is a @feature-decorated ancestor namespace. */
function hasFeatureAncestor(op) {
    let ns = op.namespace;
    while (ns && ns.name) {
        if (hasDecorator(ns, "@feature"))
            return true;
        ns = ns.namespace;
    }
    return false;
}
// a10n:todo Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.two_layers_only — not yet implemented: no rule classifies constructs by SpecLayerKind/SpecConstructKind at compile time
// a10n:todo Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.system_self_sufficient — not yet implemented: no check that system-layer files compile without domain imports
// a10n:todo Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.invariant_unique — not yet implemented: no cross-file check for duplicate @invariant op names across feature namespaces
// a10n:todo Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.inv_conjunctive — not yet implemented: semantic meta-rule; not enforceable as a compile-time linter rule
// a10n:todo Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.new_topic_feature — not yet implemented: requires semantic subject-matching between invariants and feature namespaces
// a10n:todo Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.inv_has_negation — not yet implemented: negative-proof check not present in invariant-missing-proof rule
export const noOrphanInvariantLayerRule = createRule({
    name: "no-orphan-invariant-layer",
    severity: "warning",
    description: "An @invariant op must live under a @feature namespace in Products.* — invariants inside Components.* are invisible to the planner.",
    messages: {
        default: paramMessage `@invariant "${"fqn"}" must be declared under a @feature in Products.* — invariants inside Components.* are invisible to the planner.`,
    },
    // a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.rule_no_orphan_invariant_layer
    create(context) {
        // a10n:index:call Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.invariant_path
        // a10n:index:call Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.invariant_layer
        return {
            // a10n:index:call packages.blueprint-typespec.src.rules.no-orphan-invariant-layer.rule.operation
            operation(op) {
                if (!op.decorators.some((d) => d.definition?.name === "@invariant"))
                    return;
                // a10n:blueprint Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.invariant_path
                // a10n:blueprint Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.invariant_layer
                if (isUnderComponentsNamespace(op) || !hasFeatureAncestor(op)) {
                    context.reportDiagnostic({
                        messageId: "default",
                        format: { fqn: opFqn(op) },
                        target: op,
                    });
                }
            },
        };
    },
});
