// @a10n/blueprint-typespec/dist/src/rules/operation-reachability.rule.js
import { createRule, paramMessage } from "@typespec/compiler";
import { getOpKindDecorators, isInsideKindExemptNs } from "../validators.js";
function hasDecorator(node, name) {
    return (node?.decorators ?? []).some((d) => d.definition?.name === name);
}
function nsToFqn(ns) {
    const parts = [];
    let cur = ns;
    while (cur && cur.name) {
        parts.unshift(cur.name);
        cur = cur.namespace;
    }
    return parts.join(".");
}
function opFqn(op) {
    return op.namespace ? `${nsToFqn(op.namespace)}.${op.name}` : op.name;
}
function modelFqn(model) {
    return model.namespace ? `${nsToFqn(model.namespace)}.${model.name}` : model.name;
}
function getComponentType(ns) {
    for (const dec of ns.decorators ?? []) {
        if (dec.definition?.name !== "@component")
            continue;
        const arg = dec.args?.[0]?.value;
        if (!arg)
            continue;
        let typeVal;
        if (arg.valueKind === "ObjectValue") {
            typeVal = arg.properties?.get("type")?.value;
        }
        else if (arg.kind === "Object") {
            typeVal = arg.properties?.type;
        }
        if (!typeVal)
            continue;
        return typeVal.valueKind === "EnumValue" ? typeVal.value?.name : typeVal.value?.name ?? String(typeVal);
    }
    return undefined;
}
function findComponentAncestor(op) {
    let ns = op.namespace;
    while (ns && ns.name) {
        if (hasDecorator(ns, "@component"))
            return ns;
        ns = ns.namespace;
    }
    return undefined;
}
function isInsideUiRoute(op) {
    return !!op.namespace && hasDecorator(op.namespace, "@uiRoute");
}
// Terminal @node: routes to End (go_to is a Model named "End", not a template instantiation)
function isTerminalNodeOp(op) {
    const rt = op.returnType;
    if (!rt)
        return false;
    function checkType(t) {
        if (!t)
            return false;
        if (t.kind === "Union") {
            for (const v of t.variants?.values() ?? []) {
                if (checkType(v.type))
                    return true;
            }
            return false;
        }
        if (t.kind === "Model" && (t.indexer))
            return checkType(t.indexer?.value);
        if (t.kind !== "Model")
            return false;
        const templateName = t.templateMapper?.target?.name ?? t.name;
        if (templateName !== "Command")
            return false;
        const args = t.templateMapper?.args ?? [];
        const goTo = args[1];
        if (!goTo || goTo.kind !== "Model")
            return false;
        return (goTo.name === "End" && !goTo.templateMapper);
    }
    return checkType(rt);
}
function hasCallGraphEdge(op) {
    return op.decorators.some((d) => ["@calls", "@reads", "@writes"].includes(d.definition?.name ?? ""));
}
function normalizeItems(v) {
    if (v?.valueKind === "ArrayValue")
        return v.values ?? [];
    if (v?.kind === "Tuple")
        return v.values ?? [];
    return [v];
}
/** Get all @calls targets from an op's decorator AST (avoids circular deps). */
function getCallsFromAst(op) {
    const result = [];
    for (const dec of op.decorators) {
        if (dec.definition?.name !== "@calls")
            continue;
        for (const arg of dec.args ?? []) {
            for (const item of normalizeItems(arg.value)) {
                if (item?.kind === "Operation")
                    result.push(item);
            }
        }
    }
    return result;
}
function isInsideDependency(op) {
    let ns = op.namespace;
    while (ns && ns.name) {
        if (hasDecorator(ns, "@dependency"))
            return true;
        ns = ns.namespace;
    }
    return false;
}
function isNsInsideDependency(ns) {
    let cur = ns;
    while (cur && cur.name) {
        if (hasDecorator(cur, "@dependency"))
            return true;
        cur = cur.namespace;
    }
    return false;
}
function buildIndex(program) {
    const idx = {
        calledOpFqns: new Set(),
        readModelFqns: new Set(),
        writtenModelFqns: new Set(),
        hasSuites: false,
        calledByNonDepOpFqns: new Set(),
        callsGraph: new Map(),
        pubsubGraph: new Map(),
        allOps: new Map(),
        callsCycles: [],
        pubsubCycles: [],
    };
    function addOp(set, op) {
        set.add(opFqn(op));
        set.add(op.name);
    }
    function addModel(set, m) {
        set.add(modelFqn(m));
        set.add(m.name);
    }
    // Calls graph for transitive analysis: FQN -> Set<callee FQN>
    const rawCallsGraph = new Map();
    function walkNs(ns) {
        const nsIsInDep = isNsInsideDependency(ns);
        for (const [, op] of ns.operations) {
            const opIsInDep = nsIsInDep || isInsideDependency(op);
            const fqn = opFqn(op);
            idx.allOps.set(fqn, op);
            if (op.name !== fqn)
                idx.allOps.set(op.name, op);
            // @calls/@reads/@writes/@subscribes/@publishes targets
            for (const dec of op.decorators) {
                const dname = dec.definition?.name;
                if (!["@calls", "@reads", "@writes", "@subscribes", "@publishes"].includes(dname ?? ""))
                    continue;
                for (const arg of dec.args) {
                    for (const item of normalizeItems(arg.value)) {
                        if (!item)
                            continue;
                        if (dname === "@calls" && item.kind === "Operation") {
                            const callee = item;
                            addOp(idx.calledOpFqns, callee);
                            // Track non-dep callers for dependency_op_reached
                            if (!opIsInDep) {
                                addOp(idx.calledByNonDepOpFqns, callee);
                            }
                            // Build calls graph for transitive analysis
                            const calleeFqn = opFqn(callee);
                            if (!rawCallsGraph.has(fqn))
                                rawCallsGraph.set(fqn, new Set());
                            rawCallsGraph.get(fqn).add(calleeFqn);
                            if (callee.name !== calleeFqn)
                                rawCallsGraph.get(fqn).add(callee.name);
                        }
                        else if (dname === "@reads" && item.kind === "Model") {
                            addModel(idx.readModelFqns, item);
                        }
                        else if (dname === "@writes" && item.kind === "Model") {
                            addModel(idx.writtenModelFqns, item);
                        }
                        else if (dname === "@subscribes" && item.kind === "Operation") {
                            // @subscribes from a non-dep op reaches the subscribed op
                            if (!opIsInDep) {
                                addOp(idx.calledByNonDepOpFqns, item);
                            }
                        }
                    }
                }
            }
            // mark program as having suites
            if (hasDecorator(op, "@evalSuite"))
                idx.hasSuites = true;
        }
        for (const [, child] of ns.namespaces)
            walkNs(child);
    }
    walkNs(program.getGlobalNamespaceType());
    idx.callsGraph = rawCallsGraph;
    // ── Build pub/sub component graph ────────────────────────────────────────
    // Builds a component-level directed graph: publishing component → subscribing component.
    // A cycle means two components mutually depend on each other via events.
    // Key: event variant FQN → Set<component FQN>
    const eventPublisherComps = new Map();
    const eventSubscriberComps = new Map();
    function getEvtKey(item) {
        if (!item)
            return "";
        if (item.kind === "UnionVariant") {
            const variantName = String(item.name ?? "");
            const union = item.union;
            if (union?.name) {
                const unionFqn = union.namespace ? `${nsToFqn(union.namespace)}.${union.name}` : union.name;
                return `${unionFqn}.${variantName}`;
            }
            return variantName;
        }
        if (item.kind === "Model")
            return modelFqn(item);
        if (item.name)
            return item.namespace ? `${nsToFqn(item.namespace)}.${item.name}` : item.name;
        return "";
    }
    function getCompFqnOfOp(op) {
        let ns = op.namespace;
        while (ns && ns.name) {
            if (hasDecorator(ns, "@component"))
                return nsToFqn(ns);
            ns = ns.namespace;
        }
        return undefined;
    }
    function collectPubSubComps(ns) {
        for (const [, op] of ns.operations) {
            const compFqn = getCompFqnOfOp(op);
            if (!compFqn)
                continue;
            // @publishes
            for (const dec of op.decorators) {
                if (dec.definition?.name !== "@publishes")
                    continue;
                for (const arg of dec.args) {
                    for (const item of normalizeItems(arg.value)) {
                        const key = getEvtKey(item);
                        if (!key)
                            continue;
                        if (!eventPublisherComps.has(key))
                            eventPublisherComps.set(key, new Set());
                        eventPublisherComps.get(key).add(compFqn);
                    }
                }
            }
            // @subscribes: event inferred from first parameter type
            if (hasDecorator(op, "@subscribes")) {
                for (const [, param] of op.parameters.properties) {
                    const key = getEvtKey(param.type);
                    if (!key)
                        continue;
                    if (!eventSubscriberComps.has(key))
                        eventSubscriberComps.set(key, new Set());
                    eventSubscriberComps.get(key).add(compFqn);
                    break; // first param only
                }
            }
        }
        for (const [, child] of ns.namespaces)
            collectPubSubComps(child);
    }
    collectPubSubComps(program.getGlobalNamespaceType());
    // Link publisher component → subscriber component via shared event
    for (const [evtKey, pubComps] of eventPublisherComps) {
        const subComps = eventSubscriberComps.get(evtKey);
        if (!subComps)
            continue;
        for (const pub of pubComps) {
            if (!idx.pubsubGraph.has(pub))
                idx.pubsubGraph.set(pub, new Set());
            for (const sub of subComps) {
                if (sub !== pub)
                    idx.pubsubGraph.get(pub).add(sub); // skip self-pub/sub
            }
        }
    }
    // ── Cycle detection (DFS, three-color) ────────────────────────────────────
    function detectCycles(graph) {
        const cycles = [];
        const WHITE = 0, GREY = 1, BLACK = 2;
        const color = new Map();
        const path = [];
        function dfs(node) {
            color.set(node, GREY);
            path.push(node);
            for (const neighbor of graph.get(node) ?? []) {
                if (color.get(neighbor) === GREY) {
                    // Found cycle — extract just the cycle portion from path
                    const cycleStart = path.indexOf(neighbor);
                    const cyclePath = path.slice(cycleStart);
                    cycles.push([...cyclePath, neighbor].join(" → "));
                }
                else if ((color.get(neighbor) ?? WHITE) === WHITE) {
                    dfs(neighbor);
                }
            }
            path.pop();
            color.set(node, BLACK);
        }
        for (const node of graph.keys()) {
            if ((color.get(node) ?? WHITE) === WHITE)
                dfs(node);
        }
        return cycles;
    }
    idx.callsCycles = detectCycles(rawCallsGraph);
    idx.pubsubCycles = detectCycles(idx.pubsubGraph);
    return idx;
}
function isReached(i, op) {
    const fqn = opFqn(op);
    return i.calledOpFqns.has(fqn) || i.calledOpFqns.has(op.name);
}
/** Returns true if op is a @node with is_entrypoint: true. */
// ── Rule ──────────────────────────────────────────────────────────────────────
export const operationReachabilityRule = createRule({
    name: "operation-reachability",
    severity: "warning",
    description: "Enforces call-graph edge presence and reachability for all op kinds.",
    messages: {
        opMissingEdge: paramMessage `op "${"opName"}" in a Service component has no @calls, @reads, or @writes — every service op must declare its runtime effect.`,
        uiActionMissingEdge: paramMessage `UI action op "${"opName"}" has no @calls, @reads, or @writes — every UI action must declare its runtime effect.`,
        toolMissingEdge: paramMessage `@tool op "${"opName"}" has no @calls, @reads, or @writes — every agent tool must declare its runtime effect.`,
        toolUnreachable: paramMessage `@tool op "${"opName"}" is never the target of a @calls edge — remove it or add a caller.`,
        datasetUnreachable: paramMessage `@dataset op "${"opName"}" is not called by any @evalSuite. Add @calls([${"opName"}]) on a suite that exercises it.`,
        datasetMissingReads: paramMessage `@dataset op "${"opName"}" has no @reads decorator. Add @reads([...]) pointing at the store model(s) that supply the dataset items.`,
        evalMissingScore: paramMessage `@eval namespace "${"fqn"}" has no Score op carrying @scoring. Add exactly one \`op Score(): void\` decorated with @scoring.`,
        evalMultipleScores: paramMessage `@eval namespace "${"fqn"}" has ${"count"} Score ops carrying @scoring. An @eval namespace must have exactly one.`,
        signalNotCalledByScore: paramMessage `@signal "${"opName"}" in @eval "${"evalFqn"}" is not called by the Score op. Add it to @calls([..., ${"opName"}]) on the Score op.`,
        evalSuiteNoSignal: paramMessage `@evalSuite op "${"opName"}" has no @signal reachable via its Score op @calls. The gate measures nothing.`,
        evalUnreachable: paramMessage `@eval namespace "${"fqn"}" is not called by any @evalSuite. Add an @evalSuite that carries @calls([${"fqn"}.Score, ...]).`,
        dependencyOpUnreachable: paramMessage `op "${"opName"}" in a dependency component is not reached by any first-party @calls, @subscribes, or @publishes — remove it from the dependency spec or add a first-party caller.`,
        sdkCliOpMissingOperation: paramMessage `op "${"opName"}" in a ${"compType"} component is missing @operation — every op in an Sdk component or Commands sub-namespace must carry @operation to be part of the callable surface.`,
        opSingleKind: paramMessage `op "${"opName"}" carries conflicting kind decorators "${"a"}" and "${"b"}" — an op may carry at most one op-kind decorator.`,
        opMissingKind: paramMessage `op "${"opName"}" has no op-kind decorator — every op must carry exactly one of the recognized kind decorators.`,
        callsCycle: paramMessage `@calls graph contains a cycle: ${"cycle"} — mutual @calls dependencies have no valid task ordering.`,
        pubsubCycle: paramMessage `@publishes/@subscribes graph contains a cycle: ${"cycle"} — circular pub/sub chains have no valid task ordering.`,
    },
    // a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.rule_operation_reachability
    // a10n:blueprint Products.A10nBlueprintEngine.Features.SystemAuthoring.op_single_kind
    create(context) {
        let idx = null;
        let cyclesReported = false;
        function getIdx() {
            if (!idx)
                idx = buildIndex(context.program);
            return idx;
        }
        return {
            // a10n:index:call packages.blueprint-typespec.src.rules.operation-reachability.rule.operation
            // a10n:index:call packages.blueprint-typespec.src.rules.operation-reachability.rule.namespace
            operation(op) {
                // ── calls_acyclic + pubsub_acyclic: report once per program ──────────
                if (!cyclesReported) {
                    cyclesReported = true;
                    const i = getIdx();
                    for (const cycle of i.callsCycles) {
                        // a10n:blueprint Products.A10nBlueprintEngine.Features.OperationCallGraphProofs.calls_acyclic
                        context.reportDiagnostic({ messageId: "callsCycle", format: { cycle }, target: op });
                    }
                    for (const cycle of i.pubsubCycles) {
                        // a10n:blueprint Products.A10nBlueprintEngine.Features.OperationCallGraphProofs.pubsub_acyclic
                        context.reportDiagnostic({ messageId: "pubsubCycle", format: { cycle }, target: op });
                    }
                }
                // ── op_single_kind + op_missing_kind ─────────────────────────────────
                // Skip blueprint-reserved structure ops (invariants, test cases, signals checked separately)
                // and uiRoute ops (exempt — namespace is the kind signal).
                if (!isInsideKindExemptNs(op)) {
                    const kinds = getOpKindDecorators(op);
                    if (kinds.length > 1) {
                        context.reportDiagnostic({
                            messageId: "opSingleKind",
                            format: { opName: op.name, a: kinds[0], b: kinds[1] },
                            target: op,
                        });
                    }
                    else if (kinds.length === 0) {
                        context.reportDiagnostic({
                            messageId: "opMissingKind",
                            format: { opName: op.name },
                            target: op,
                        });
                    }
                }
                // Is this op inside a @dependency namespace?
                const isInDep = isInsideDependency(op);
                const comp = findComponentAncestor(op);
                // ── dependency_op_reached for ops without @component wrapper ──────────
                // Ops declared directly in a @dependency namespace (no @component ancestor)
                // must still be reached by non-dep callers.
                // a10n:blueprint Products.A10nBlueprintEngine.Features.OperationReachability.dependency_op_reached
                if (isInDep && !comp) {
                    // Skip blueprint-reserved ops
                    if (hasDecorator(op, "@invariant") || hasDecorator(op, "@signal"))
                        return;
                    const i = getIdx();
                    const fqn = opFqn(op);
                    if (!i.calledByNonDepOpFqns.has(fqn) && !i.calledByNonDepOpFqns.has(op.name)) {
                        context.reportDiagnostic({ messageId: "dependencyOpUnreachable", format: { opName: op.name }, target: op });
                    }
                    return;
                }
                if (!comp)
                    return;
                const compType = getComponentType(comp);
                const isTool = hasDecorator(op, "@tool");
                const isUiAction = isInsideUiRoute(op);
                // Worker ops have their own reachability check via worker-unreachable in worker-type-spec.rule.ts
                const isWorkerOp = !!(op.namespace && hasDecorator(op.namespace, "@worker"));
                // Service/Sdk reachability only applies to non-dependency, non-worker ops
                // sdk_op_exempt: Sdk ops are exempt from the edge-presence requirement (they are leaf nodes)
                // a10n:blueprint Products.A10nBlueprintEngine.Features.OperationReachability.sdk_op_exempt
                const isServiceOp = !isTool && !isUiAction && !isInDep && !isWorkerOp && compType === "Service";
                const isSdkOp = !isTool && !isUiAction && !isInDep && !isWorkerOp && compType === "Sdk";
                // Skip blueprint-reserved ops (invariants, test cases, signals)
                // @scoring (Score) ops are NOT skipped — eval_reached requires each Score be called by an @evalSuite.
                // @evalSuite and @dataset are NOT skipped — they have their own checks below.
                if (hasDecorator(op, "@invariant") || hasDecorator(op, "@signal"))
                    return;
                // ── sdk_commands_op_requires_operation ────────────────────────────────
                // Every op in an Sdk component or Commands sub-namespace must carry @operation.
                const isInCommandsNs = compType === "Service" && (() => {
                    let ns = op.namespace;
                    while (ns && ns !== comp && ns.name) {
                        if (ns.name === "Commands")
                            return true;
                        ns = ns.namespace;
                    }
                    return false;
                })();
                // a10n:blueprint Products.A10nBlueprintEngine.Features.InterfaceNamespacePlacement.op_requires_decorator
                if ((compType === "Sdk" || isInCommandsNs) && !hasDecorator(op, "@operation")) {
                    context.reportDiagnostic({
                        messageId: "sdkCliOpMissingOperation",
                        format: { opName: op.name, compType: compType ?? "unknown" },
                        target: op,
                    });
                }
                // ── eval_reached: @scoring Score op must be called by at least one @evalSuite ──
                // Only fires when at least one @evalSuite exists in the program — avoids false positives
                // in isolated eval-structure unit tests that don't declare suites.
                // Uses FQN-only matching to avoid false positives when multiple @eval namespaces each
                // have a "Score" op — short-name matching would mark all as reached if any is called.
                // a10n:blueprint Products.A10nBlueprintEngine.Features.OperationReachability.eval_reached
                if (hasDecorator(op, "@scoring")) {
                    const i = getIdx();
                    if (i.hasSuites) {
                        const fqn = opFqn(op);
                        if (!i.calledOpFqns.has(fqn)) {
                            context.reportDiagnostic({ messageId: "toolUnreachable", format: { opName: op.name }, target: op });
                        }
                    }
                    return;
                }
                // ── @node reachability / edge presence ───────────────────────────────
                const isNode = hasDecorator(op, "@node");
                const isPredict = hasDecorator(op, "@predict");
                const isInterrupt = hasDecorator(op, "@interrupt");
                if (isNode) {
                    // Terminal nodes (go_to is End) are exempt — they have no @calls by design.
                    // Entry nodes are the first in the graph — they may also have no @calls.
                    // Only non-terminal nodes that lack any call-graph edge are flagged.
                    // a10n:blueprint Products.A10nBlueprintEngine.Features.OperationReachability.node_has_edge
                    const isTerminal = isTerminalNodeOp(op);
                    if (!isTerminal && !hasCallGraphEdge(op)) {
                        context.reportDiagnostic({ messageId: "toolMissingEdge", format: { opName: op.name }, target: op });
                    }
                    return;
                }
                if (isPredict) {
                    // @predict reachability: must be @calls-ed by at least one @node
                    // (predict_edge_exempt: @predict ops are exempt from edge-presence requirement)
                    // a10n:blueprint Products.A10nBlueprintEngine.Features.OperationReachability.predict_edge_exempt
                    // a10n:blueprint Products.A10nBlueprintEngine.Features.OperationReachability.predict_op_reached
                    const i = getIdx();
                    if (!isReached(i, op)) {
                        context.reportDiagnostic({ messageId: "toolUnreachable", format: { opName: op.name }, target: op });
                    }
                    return;
                }
                if (isInterrupt) {
                    // @interrupt reachability: must be @calls-ed by at least one @tool or @node
                    // a10n:blueprint Products.A10nBlueprintEngine.Features.OpContractAuthoring.interrupt_reached
                    const i = getIdx();
                    if (!isReached(i, op)) {
                        context.reportDiagnostic({ messageId: "toolUnreachable", format: { opName: op.name }, target: op });
                    }
                    return;
                }
                // ── dataset_reached: @dataset op must be called by at least one @evalSuite ──
                // a10n:blueprint Products.A10nBlueprintEngine.Features.OperationReachability.dataset_reached
                if (hasDecorator(op, "@dataset")) {
                    const i = getIdx();
                    const fqn = opFqn(op);
                    if (!i.calledOpFqns.has(fqn) && !i.calledOpFqns.has(op.name)) {
                        context.reportDiagnostic({ messageId: "datasetUnreachable", format: { opName: op.name }, target: op });
                    }
                    // ── dataset_has_reads: @dataset op must carry @reads ──
                    if (!hasDecorator(op, "@reads")) {
                        context.reportDiagnostic({ messageId: "datasetMissingReads", format: { opName: op.name }, target: op });
                    }
                    return;
                }
                // ── suite_calls_signal: @evalSuite must have a @signal reachable via Score's @calls ──
                // a10n:blueprint Products.A10nBlueprintEngine.Features.OperationReachability.suite_calls_signal
                if (hasDecorator(op, "@evalSuite")) {
                    const suiteCalls = getCallsFromAst(op);
                    // Find all Score ops in the suite's @calls
                    const scoreOps = suiteCalls.filter((called) => called.name === "Score" && called.namespace && hasDecorator(called.namespace, "@eval"));
                    // For each Score op, check if it calls at least one @signal op
                    let hasSignal = false;
                    for (const scoreOp of scoreOps) {
                        const scoreCalls = getCallsFromAst(scoreOp);
                        if (scoreCalls.some((called) => hasDecorator(called, "@signal"))) {
                            hasSignal = true;
                            break;
                        }
                    }
                    if (scoreOps.length > 0 && !hasSignal) {
                        context.reportDiagnostic({ messageId: "evalSuiteNoSignal", format: { opName: op.name }, target: op });
                    }
                    return;
                }
                const i = getIdx();
                // ── edge presence ────────────────────────────────────────────────────
                // a10n:blueprint Products.A10nBlueprintEngine.Features.OperationReachability.service_op_edge
                if (isServiceOp && !hasCallGraphEdge(op)) {
                    context.reportDiagnostic({ messageId: "opMissingEdge", format: { opName: op.name }, target: op });
                }
                // a10n:blueprint Products.A10nBlueprintEngine.Features.OperationReachability.ui_action_edge
                if (isUiAction && !hasCallGraphEdge(op)) {
                    context.reportDiagnostic({ messageId: "uiActionMissingEdge", format: { opName: op.name }, target: op });
                }
                // a10n:blueprint Products.A10nBlueprintEngine.Features.OperationReachability.tool_has_edge
                if (isTool && !hasCallGraphEdge(op)) {
                    context.reportDiagnostic({ messageId: "toolMissingEdge", format: { opName: op.name }, target: op });
                }
                // ── reachability ─────────────────────────────────────────────────────
                // UI actions are entry points (invoked by the user via the UI), never the target
                // of a @calls edge — so they have no reachability obligation, like http/cli ops.
                // a10n:blueprint Products.A10nBlueprintEngine.Features.OperationReachability.tool_reached
                if (isTool && !isReached(i, op)) {
                    context.reportDiagnostic({ messageId: "toolUnreachable", format: { opName: op.name }, target: op });
                }
                // ── dependency_op_reached ─────────────────────────────────────────────
                // Ops inside a @dependency component must be reached by a non-dep first-party caller.
                // Test-case references don't count for dep ops (dep namespaces have no test suites).
                // a10n:blueprint Products.A10nBlueprintEngine.Features.OperationReachability.dependency_op_reached
                if (isInDep && (compType === "Service" || compType === "Sdk" || compType === "Store" || compType === "Agent" || compType === "Ui")) {
                    const fqn = opFqn(op);
                    if (!i.calledByNonDepOpFqns.has(fqn) && !i.calledByNonDepOpFqns.has(op.name)) {
                        context.reportDiagnostic({ messageId: "dependencyOpUnreachable", format: { opName: op.name }, target: op });
                    }
                }
            },
            namespace(ns) {
                // ── eval_has_score: @eval namespace must contain exactly one Score op with @scoring ──
                if (!hasDecorator(ns, "@eval"))
                    return;
                const fqn = nsToFqn(ns);
                // Count all ops with @scoring (only one should exist and it must be named "Score").
                // TypeSpec cannot have two ops with the same name in a namespace, so we must count
                // all @scoring ops to detect the case where a second scoring op has a different name.
                const scoringOps = [];
                let namedScoreOp;
                for (const [, op] of ns.operations) {
                    if (hasDecorator(op, "@scoring")) {
                        scoringOps.push(op);
                    }
                    if (op.name === "Score" && hasDecorator(op, "@scoring")) {
                        namedScoreOp = op;
                    }
                }
                // a10n:blueprint Products.A10nBlueprintEngine.Features.EvaluationAuthoring.eval_has_score
                if (scoringOps.length === 0) {
                    context.reportDiagnostic({ messageId: "evalMissingScore", format: { fqn }, target: ns });
                }
                else if (scoringOps.length > 1) {
                    context.reportDiagnostic({ messageId: "evalMultipleScores", format: { fqn, count: String(scoringOps.length) }, target: ns });
                }
                else if (!namedScoreOp) {
                    // exactly one @scoring op exists but it isn't named "Score"
                    context.reportDiagnostic({ messageId: "evalMissingScore", format: { fqn }, target: ns });
                }
                // ── score_calls_signals: every @signal in the @eval namespace must appear in Score's @calls ──
                // a10n:blueprint Products.A10nBlueprintEngine.Features.EvaluationAuthoring.score_calls_signals
                if (namedScoreOp) {
                    const scoreCalls = getCallsFromAst(namedScoreOp);
                    const calledOpNames = new Set(scoreCalls.map((op) => op.name));
                    for (const [, op] of ns.operations) {
                        if (!hasDecorator(op, "@signal"))
                            continue;
                        if (!calledOpNames.has(op.name)) {
                            context.reportDiagnostic({
                                messageId: "signalNotCalledByScore",
                                format: { opName: op.name, evalFqn: fqn },
                                target: op,
                            });
                        }
                    }
                }
                // ── eval_reached: @eval namespace must be called by at least one @evalSuite ──
                if (namedScoreOp) {
                    const i = getIdx();
                    const scoreFqn = opFqn(namedScoreOp);
                    if (!i.calledOpFqns.has(scoreFqn) && !i.calledOpFqns.has(namedScoreOp.name)) {
                        context.reportDiagnostic({ messageId: "evalUnreachable", format: { fqn }, target: ns });
                    }
                }
            },
        };
    },
});
