// @a10n/blueprint-typespec/dist/src/rules/proof-missing-projects.rule.js
import { createRule, paramMessage, } from "@typespec/compiler";
function hasDecorator(node, name) {
    return node.decorators.some((d) => d.definition?.name === name);
}
export const proofMissingProjectsRule = createRule({
    name: "proof-missing-projects",
    severity: "warning",
    description: "Every @signal and @dataset must declare @projects listing at least one invariant.",
    messages: {
        signal: paramMessage `@signal "${"opName"}" has no @projects. Add @projects(...) listing the invariant(s) it proves.`,
        dataset: paramMessage `@dataset "${"opName"}" has no @projects. Add @projects(...) listing the invariant(s) it covers.`,
    },
    // a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.rule_proof_missing_projects
    // a10n:blueprint Products.A10nBlueprintEngine.Features.SystemAuthoring.proof_targets_invariant
    // a10n:blueprint Products.A10nBlueprintEngine.Features.DatasetAuthoring.targets_invariant
    create(context) {
        return {
            // a10n:index:call packages.blueprint-typespec.src.rules.proof-missing-projects.rule.operation
            operation(op) {
                // @testCase no longer carries proofs in the invariant-mapping model — invariants are
                // proven by projecting-op-kind ops (@projects on the component op) or @signal. See
                // invariant-missing-proof. So @testCase is intentionally excluded here.
                const isSignal = hasDecorator(op, "@signal");
                const isDataset = hasDecorator(op, "@dataset");
                if (!isSignal && !isDataset)
                    return;
                const hasProjects = hasDecorator(op, "@projects");
                if (hasProjects)
                    return;
                const opName = op.name;
                if (isSignal) {
                    context.reportDiagnostic({
                        messageId: "signal",
                        format: { opName },
                        target: op,
                    });
                }
                else {
                    context.reportDiagnostic({
                        messageId: "dataset",
                        format: { opName },
                        target: op,
                    });
                }
            },
        };
    },
});
