// @a10n/blueprint-typespec/dist/src/rules/state-machine-authoring.rule.js
import { createRule, paramMessage } from "@typespec/compiler";
import { extractInlineRefs } from "../validators.js";
function hasDecorator(node, name) {
    return (node?.decorators ?? []).some((d) => d.definition?.name === name);
}
function isEntityModel(m) {
    return hasDecorator(m, "@entity");
}
function nsToFqn(ns) {
    const parts = [];
    let cur = ns;
    while (cur && cur.name) {
        parts.unshift(cur.name);
        cur = cur.namespace;
    }
    return parts.join(".");
}
function enumFqn(e) {
    const ns = e.namespace ? nsToFqn(e.namespace) : "";
    return ns ? `${ns}.${e.name}` : e.name;
}
/**
 * Returns true if any @invariant op in the program references the given enum
 * (by short name or FQN) in its predicate {@...} references.
 * Uses typed inline refs — predicate text is scanned for {@ tokens.
 */
function isEnumCoveredByInvariant(program, targetEnum) {
    const targetShort = targetEnum.name;
    const targetFqn = enumFqn(targetEnum);
    function walkNs(ns) {
        for (const [, op] of ns.operations) {
            if (!hasDecorator(op, "@invariant"))
                continue;
            for (const dec of op.decorators) {
                if (dec.definition?.name !== "@invariant")
                    continue;
                const arg = dec.args[0]?.value;
                if (!arg || arg.valueKind !== "ObjectValue")
                    continue;
                const pred = arg.properties?.get("predicate");
                const predText = pred?.value?.value ?? "";
                // Check for {@EnumName} or {@FQN} references in the predicate
                for (const ref of extractInlineRefs(predText)) {
                    if (ref === targetShort || ref === targetFqn || ref.endsWith(`.${targetShort}`) || ref.startsWith(`${targetShort}.`) || ref.startsWith(`${targetFqn}.`)) {
                        return true;
                    }
                }
            }
        }
        for (const [, child] of ns.namespaces) {
            if (walkNs(child))
                return true;
        }
        return false;
    }
    return walkNs(program.getGlobalNamespaceType());
}
export const stateMachineAuthoringRule = createRule({
    name: "state-machine-authoring",
    severity: "warning",
    description: "Every enum field on an @entity must be covered by at least one @invariant referencing that enum.",
    messages: {
        noInvariant: paramMessage `@entity "${"entity"}" has enum field "${"field"}" (${"enumName"}) with no @invariant referencing {@${"enumName"}} — add a transition @invariant to constrain legal values.`,
    },
    // a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.rule_state_machine_authoring
    // a10n:blueprint Products.A10nBlueprintEngine.Features.StateMachineAuthoring.state_via_invariants
    create(context) {
        return {
            // a10n:index:call packages.blueprint-typespec.src.rules.state-machine-authoring.rule.model
            model(model) {
                if (!isEntityModel(model))
                    return;
                for (const [, prop] of model.properties) {
                    const fieldType = prop.type;
                    if (fieldType?.kind !== "Enum")
                        continue;
                    const e = fieldType;
                    if (!isEnumCoveredByInvariant(context.program, e)) {
                        context.reportDiagnostic({
                            messageId: "noInvariant",
                            format: { entity: model.name, field: prop.name, enumName: e.name },
                            target: prop,
                        });
                    }
                }
            },
        };
    },
});
