// @a10n/blueprint-typespec/dist/src/rules/surface-is-namespace.rule.js
import { createRule, paramMessage } from "@typespec/compiler";
function hasDecorator(node, name) {
    return (node?.decorators ?? []).some((d) => d.definition?.name === name);
}
/** Return the @component type string ("Ui", "Service", etc.) or undefined. */
function getComponentType(ns) {
    for (const dec of ns.decorators ?? []) {
        if (dec.definition?.name !== "@component")
            continue;
        const arg = dec.args?.[0]?.value;
        if (!arg)
            continue;
        let typeVal;
        if (arg.valueKind === "ObjectValue") {
            typeVal = arg.properties?.get("type")?.value;
        }
        else if (arg.kind === "Object") {
            typeVal = arg.properties?.type;
        }
        if (!typeVal)
            continue;
        return typeVal.valueKind === "EnumValue"
            ? typeVal.value?.name
            : typeVal.value?.name ?? String(typeVal);
    }
    return undefined;
}
/** Walk up the ancestor chain to find the nearest @component namespace. */
function findEnclosingComponentNs(ns) {
    let cur = ns.namespace;
    while (cur && cur.name) {
        if (hasDecorator(cur, "@component"))
            return cur;
        cur = cur.namespace;
    }
    return undefined;
}
export const surfaceIsNamespaceRule = createRule({
    name: "surface-is-namespace",
    severity: "warning",
    description: "Every UI surface (@uiRoute) must be declared as a namespace nested inside a @component with type Ui.",
    messages: {
        noUiComponent: paramMessage `@uiRoute namespace "${"ns"}" is not nested inside a @component(type: Ui) — UI surfaces must live inside a Ui component namespace.`,
    },
    // a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.rule_surface_is_namespace
    // a10n:blueprint Products.A10nBlueprintEngine.Features.UiRouteAuthoring.surface_is_namespace
    create(context) {
        return {
            // a10n:index:call packages.blueprint-typespec.src.rules.surface-is-namespace.rule.namespace
            namespace(ns) {
                if (!hasDecorator(ns, "@uiRoute"))
                    return;
                // The $uiRoute decorator already enforces that @uiRoute must be nested inside a
                // @component or another @uiRoute (reporting invalid-namespace-nesting for the absent-
                // component case). This rule adds the complementary check: the enclosing @component
                // must be of type Ui (not Agent, Service, etc.).
                const componentNs = findEnclosingComponentNs(ns);
                if (!componentNs)
                    return; // decorator already reported invalid-namespace-nesting
                // a10n:blueprint Products.A10nBlueprintEngine.Features.UiRouteAuthoring.surface_is_namespace
                if (getComponentType(componentNs) !== "Ui") {
                    context.reportDiagnostic({
                        messageId: "noUiComponent",
                        format: { ns: ns.name },
                        target: ns,
                    });
                }
            },
        };
    },
});
