// @a10n/blueprint-typespec/dist/src/rules/system-prose-refs.rule.js
import { createRule, paramMessage, } from "@typespec/compiler";
import { extractInlineRefs } from "../validators.js";
// a10n:blueprint:ignore
function hasDecorator(node, name) {
    return node.decorators.some((d) => d.definition?.name === name);
}
/** Returns true if the type is an @entity model — prohibited as a prose ref target from system layer. */
// a10n:blueprint:ignore
function isEntityModel(type) {
    if (type.kind !== "Model")
        return false;
    return hasDecorator(type, "@entity");
}
/** Returns true if the type is a ModelProperty whose owning model is @entity — also prohibited. */
// a10n:blueprint:ignore
function isEntityField(type) {
    if (type.kind !== "ModelProperty")
        return false;
    const prop = type;
    return !!prop.model && isEntityModel(prop.model);
}
/** Walk the global namespace tree to resolve a dotted FQN to a Type (mirrors walkGlobalNs in mentions.ts). */
// a10n:blueprint:ignore
function walkGlobal(program, fqn) {
    const segs = fqn.split(".");
    let cur = program.getGlobalNamespaceType();
    for (let i = 0; i < segs.length; i++) {
        const seg = segs[i];
        const isLast = i === segs.length - 1;
        if (isLast) {
            if (cur.namespaces.has(seg))
                return cur.namespaces.get(seg);
            if (cur.models.has(seg))
                return cur.models.get(seg);
            if (cur.enums.has(seg))
                return cur.enums.get(seg);
            if (cur.operations.has(seg))
                return cur.operations.get(seg);
            return undefined;
        }
        const child = cur.namespaces.get(seg);
        if (child) {
            cur = child;
            continue;
        }
        // May be a model with property path
        const model = cur.models.get(seg);
        if (model) {
            const remaining = segs.slice(i + 1);
            let props = model.properties;
            for (let j = 0; j < remaining.length; j++) {
                const prop = props.get(remaining[j]);
                if (!prop)
                    return undefined;
                if (j === remaining.length - 1)
                    return prop;
                if (prop.type?.kind === "Model")
                    props = prop.type.properties;
                else
                    return undefined;
            }
        }
        const enumType = cur.enums.get(seg);
        if (enumType) {
            const remaining = segs.slice(i + 1);
            if (remaining.length === 1)
                return enumType.members.get(remaining[0]);
            return undefined;
        }
        return undefined;
    }
    return undefined;
}
// a10n:blueprint:ignore
function usingNodeToString(node) {
    if (node.sv !== undefined)
        return node.sv;
    return `${usingNodeToString(node.base)}.${node.id.sv}`;
}
// a10n:blueprint:ignore
function getUsingPrefixes(ns) {
    const seen = new Set();
    const prefixes = [];
    let cur = ns;
    while (cur) {
        let node = cur.node;
        while (node) {
            if (Array.isArray(node.usings)) {
                for (const u of node.usings) {
                    const s = usingNodeToString(u.name);
                    if (!seen.has(s)) {
                        seen.add(s);
                        prefixes.push(s);
                    }
                }
            }
            node = node.parent;
        }
        cur = cur.namespace;
    }
    return prefixes;
}
/**
 * Resolve a ref token to its Type using the same strategies as inline-refs rule
 * (global FQN, namespace ancestor prefixes, using-statement prefixes).
 */
// a10n:blueprint:ignore
function resolveToken(program, ref, contextNs) {
    // 1. Try global FQN directly
    const [direct] = program.resolveTypeReference(ref);
    if (direct !== undefined)
        return direct;
    const walked = walkGlobal(program, ref);
    if (walked)
        return walked;
    // 2. Ancestor namespace prefixes
    const parts = [];
    let c = contextNs;
    while (c && c.name) {
        parts.unshift(c.name);
        c = c.namespace;
    }
    for (let i = parts.length; i >= 0; i--) {
        const prefix = parts.slice(0, i).join(".");
        const full = prefix ? `${prefix}.${ref}` : ref;
        const [t] = program.resolveTypeReference(full);
        if (t !== undefined)
            return t;
        if (prefix) {
            const w = walkGlobal(program, full);
            if (w)
                return w;
        }
    }
    // 3. Using-statement prefixes
    const usingPaths = getUsingPrefixes(contextNs);
    const firstSeg = ref.split(".")[0];
    const remainder = ref.includes(".") ? ref.slice(ref.indexOf(".") + 1) : "";
    for (const up of usingPaths) {
        const [tA] = program.resolveTypeReference(`${up}.${ref}`);
        if (tA !== undefined)
            return tA;
        const wA = walkGlobal(program, `${up}.${ref}`);
        if (wA)
            return wA;
        const usingLast = up.split(".").at(-1);
        if (usingLast === firstSeg) {
            const expanded = remainder ? `${up}.${remainder}` : up;
            const [tB] = program.resolveTypeReference(expanded);
            if (tB !== undefined)
                return tB;
            const wB = walkGlobal(program, expanded);
            if (wB)
                return wB;
        }
    }
    return undefined;
}
/**
 * Returns true if the ref resolves to a forbidden domain entity type:
 * an @entity model or a field of an @entity model.
 * Enums (including those declared in entity files) are allowed.
 */
// a10n:blueprint:ignore
function isDomainEntityRef(program, ref, contextNs) {
    if (!contextNs)
        return false;
    const type = resolveToken(program, ref, contextNs);
    if (!type)
        return false;
    return isEntityModel(type) || isEntityField(type);
}
/**
 * Returns true if the resolved type is a system-layer operational construct
 * that must not appear in an @invariant predicate.
 *
 * Flagged (system layer for predicate purposes):
 *   - Operations declared inside a @component or @dependency namespace
 *     (these are @flow, @tool, @interrupt, @graph, @worker ops)
 *   - Namespace types that themselves carry @component or @dependency
 *     (i.e. the component namespace itself, or a sub-namespace like Tools)
 *
 * NOT flagged (plain data/structural types that happen to live in a component):
 *   - Model types (plain DTOs, input/output models)
 *   - ModelProperty types (fields of those models)
 *   - Enum types
 *   - EnumMember values
 *
 * The invariant is a domain-layer contract and may reference any data type;
 * it must not reference operational constructs (ops, component namespaces).
 */
// a10n:blueprint:ignore
function isSystemLayerType(type) {
    if (type.kind === "Operation") {
        // An op is system-layer if it is declared inside a @component namespace.
        return isInsideComponent(type.namespace);
    }
    if (type.kind === "Namespace") {
        // A namespace is system-layer if it IS a @component/@dependency namespace
        // or is nested inside one.
        return isInsideComponent(type);
    }
    // Plain models, enums, model properties, and enum members are data types —
    // they are allowed as predicate references even if they are co-located inside
    // a @component namespace alongside operational constructs.
    return false;
}
/**
 * Returns true if the ref resolves to a system-layer construct (inside @component).
 * Domain types (enums, entities, entity fields, feature namespaces) are NOT system-layer.
 */
// a10n:blueprint:ignore
function isSystemLayerRef(program, ref, contextNs) {
    if (!contextNs)
        return false;
    const type = resolveToken(program, ref, contextNs);
    if (!type)
        return false;
    return isSystemLayerType(type);
}
// a10n:blueprint:ignore
function checkStringForDomainRefs(str, program, contextNs, reportTarget, report) {
    for (const ref of extractInlineRefs(str)) {
        // a10n:index:call packages.blueprint-typespec.src.rules.system-prose-refs.rule.isDomainEntityRef
        if (isDomainEntityRef(program, ref, contextNs)) {
            report(ref, reportTarget);
        }
    }
}
// a10n:blueprint:ignore
function checkArgValue(entity, program, contextNs, reportTarget, report) {
    if ("entityKind" in entity) {
        const val = entity;
        if (val.valueKind === "StringValue") {
            checkStringForDomainRefs(val.value, program, contextNs, reportTarget, report);
        }
        else if (val.valueKind === "ArrayValue") {
            for (const item of val.values) {
                checkArgValue(item, program, contextNs, reportTarget, report);
            }
        }
        else if (val.valueKind === "ObjectValue") {
            for (const [, prop] of val.properties) {
                checkArgValue(prop.value, program, contextNs, reportTarget, report);
            }
        }
        return;
    }
    const t = entity;
    if (t.kind === "String") {
        checkStringForDomainRefs(t.value, program, contextNs, reportTarget, report);
    }
    else if (t.kind === "Tuple") {
        for (const item of t.values) {
            checkArgValue(item, program, contextNs, reportTarget, report);
        }
    }
}
// a10n:blueprint:ignore
function checkDecorators(decorators, program, contextNs, reportTarget, report) {
    for (const dec of decorators) {
        for (const arg of dec.args) {
            checkArgValue(arg.value, program, contextNs, reportTarget, report);
        }
    }
}
/** True if the namespace is inside a @component (system layer). */
// a10n:blueprint:ignore
function isInsideComponent(ns) {
    let cur = ns;
    while (cur && cur.name) {
        if (hasDecorator(cur, "@component"))
            return true;
        cur = cur.namespace;
    }
    return false;
}
export const systemProseRefsRule = createRule({
    name: "system-prose-refs",
    severity: "warning",
    description: "String fields in system-layer constructs must not reference domain entity models or entity fields via {@...}. Use enum values (allowed) or system-layer constructs instead.",
    messages: {
        default: paramMessage `{@${"ref"}} in a system-layer string field references a domain entity or entity field. System-layer prose must reference system constructs or domain enum values, not entities. Use a projection field or enum value instead.`,
        "invariant-pred-system-ref": paramMessage `{@${"ref"}} in @invariant predicate references a system-layer construct. Invariant predicates must only reference domain types (entities, entity fields, enums, enum values), not system-layer constructs (components, component ops, projections, signals, test cases).`,
    },
    // a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.rule_system_prose_refs
    // a10n:blueprint Products.A10nBlueprintEngine.Features.SystemAuthoring.prose_refs_system
    // a10n:blueprint Products.A10nBlueprintEngine.Features.DomainAuthoring.pred_no_sys
    // a10n:todo Products.A10nBlueprintEngine.Features.DomainAuthoring.business_logic_placement — not yet implemented: detecting behavioral statements in component @doc strings requires semantic analysis beyond structural TypeSpec linting
    // a10n:todo Products.A10nBlueprintEngine.Features.DomainAuthoring.entity_doc_definitional — not yet implemented: distinguishing definitional @doc from rule-stating @doc requires NLP semantic analysis
    // a10n:todo Products.A10nBlueprintEngine.Features.DomainAuthoring.feature_doc_capability — not yet implemented: detecting whether a @feature @doc states capability vs rule content requires NLP; no structural TypeSpec rule enforces this
    // a10n:todo Products.A10nBlueprintEngine.Features.DomainAuthoring.doc_no_restatement — not yet implemented: detecting field @doc restatements of field name or type requires string-matching heuristics not yet implemented
    create(context) {
        // a10n:index:call Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.domain_no_sys
        // a10n:blueprint Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.domain_no_sys
        function report(ref, target) {
            context.reportDiagnostic({
                messageId: "default",
                format: { ref },
                target,
            });
        }
        // a10n:blueprint:ignore
        function reportInvariantPredSysRef(ref, target) {
            context.reportDiagnostic({
                messageId: "invariant-pred-system-ref",
                format: { ref },
                target,
            });
        }
        return {
            // a10n:index:call packages.blueprint-typespec.src.rules.system-prose-refs.rule.model
            // a10n:index:call packages.blueprint-typespec.src.rules.system-prose-refs.rule.modelProperty
            // a10n:index:call packages.blueprint-typespec.src.rules.system-prose-refs.rule.operation
            // a10n:index:call packages.blueprint-typespec.src.rules.system-prose-refs.rule.namespace
            // @doc and decorator strings on models inside @component
            model(model) {
                if (!isInsideComponent(model.namespace))
                    return;
                checkDecorators(model.decorators, context.program, model.namespace, model, report);
            },
            modelProperty(prop) {
                if (!isInsideComponent(prop.model?.namespace))
                    return;
                checkDecorators(prop.decorators, context.program, prop.model?.namespace, prop, report);
            },
            // @flow, @tool, @signal, @scoring, @testing etc. on ops inside @component
            // AND @invariant predicate check (pred_no_sys invariant)
            operation(op) {
                if (isInsideComponent(op.namespace)) {
                    checkDecorators(op.decorators, context.program, op.namespace, op, report);
                }
                // Check @invariant predicates for system-layer refs (pred_no_sys)
                if (!hasDecorator(op, "@invariant"))
                    return;
                for (const dec of op.decorators) {
                    if (dec.definition?.name !== "@invariant")
                        continue;
                    const arg = dec.args[0]?.value;
                    if (!arg || arg.valueKind !== "ObjectValue")
                        continue;
                    const pred = arg.properties?.get("predicate");
                    const predText = pred?.value?.value ?? "";
                    for (const ref of extractInlineRefs(predText)) {
                        // a10n:index:call packages.blueprint-typespec.src.rules.system-prose-refs.rule.isSystemLayerRef
                        if (isSystemLayerRef(context.program, ref, op.namespace)) {
                            // a10n:index:call packages.blueprint-typespec.src.rules.system-prose-refs.rule.reportInvariantPredSysRef
                            reportInvariantPredSysRef(ref, op);
                        }
                    }
                }
            },
            // @component, @testing, @dependsOn, @testSuite etc. on namespaces
            namespace(ns) {
                if (!isInsideComponent(ns))
                    return;
                checkDecorators(ns.decorators, context.program, ns, ns, report);
            },
        };
    },
});
