// @a10n/blueprint-typespec/dist/src/rules/ui-route-authoring.rule.js
import { createRule, getDoc, paramMessage } from "@typespec/compiler";
import { extractInlineRefs, resolveInlineRef } from "../validators.js";
function hasDecorator(node, name) {
    return (node?.decorators ?? []).some((d) => d.definition?.name === name);
}
function isUiRouteNs(ns) {
    return hasDecorator(ns, "@uiRoute");
}
function getUiRoutePathArg(ns) {
    for (const dec of ns.decorators) {
        if (dec.definition?.name !== "@uiRoute")
            continue;
        for (const arg of dec.args) {
            const v = arg.value;
            if (v?.valueKind === "ObjectValue") {
                const prop = v.properties?.get("path");
                if (prop?.value?.value)
                    return prop.value.value;
            }
        }
    }
    return undefined;
}
function isNestedInsideRoute(ns) {
    let cur = ns.namespace;
    while (cur && cur.name) {
        if (isUiRouteNs(cur))
            return true;
        cur = cur.namespace;
    }
    return false;
}
function isActionOpInRoute(op) {
    return !!op.namespace && isUiRouteNs(op.namespace);
}
function nsToFqn(ns) {
    const parts = [];
    let cur = ns;
    while (cur && cur.name) {
        parts.unshift(cur.name);
        cur = cur.namespace;
    }
    return parts.join(".");
}
/**
 * Collect all @calls target operations across action ops in this @uiRoute namespace.
 * Returns a Set of FQNs of called operations.
 */
function getCallTargetFqns(ns) {
    const fqns = new Set();
    for (const [, op] of ns.operations) {
        for (const dec of op.decorators) {
            if (dec.definition?.name !== "@calls")
                continue;
            for (const arg of dec.args) {
                const v = arg.value;
                const items = v?.valueKind === "ArrayValue" ? v.values : (v?.kind === "Tuple" ? v.values : [v]);
                for (const item of items ?? []) {
                    const callee = item;
                    if (callee?.kind !== "Operation")
                        continue;
                    const calleeFqn = callee.namespace ? `${nsToFqn(callee.namespace)}.${callee.name}` : callee.name;
                    fqns.add(callee.name); // short name
                    fqns.add(calleeFqn); // fqn
                }
            }
        }
    }
    return fqns;
}
// a10n:todo Products.A10nBlueprintEngine.Features.UiRouteAuthoring.surface_is_namespace — not yet implemented: no check that @uiRoute is inside a @component(Ui) namespace
// a10n:todo Products.A10nBlueprintEngine.Features.UiRouteAuthoring.action_declares_calls — not yet implemented: no completeness check that all runtime backend invocations are declared via @calls
// a10n:todo Products.A10nBlueprintEngine.Features.UiRouteAuthoring.action_declares_reads — not yet implemented: no completeness check for undeclared @reads on @uiRoute action ops
// a10n:todo Products.A10nBlueprintEngine.Features.UiRouteAuthoring.action_declares_writes — not yet implemented: no completeness check for undeclared @writes on @uiRoute action ops
// a10n:todo Products.A10nBlueprintEngine.Features.UiRouteAuthoring.event_actor_only — not yet implemented: no check that event-driven ops use actor-typed test cases instead of event-handler decorators
// a10n:todo Products.A10nBlueprintEngine.Features.UiRouteAuthoring.subscription_via_calls — not yet implemented: no check that mount-time stream subscriptions are declared via @calls
export const uiRouteAuthoringRule = createRule({
    name: "ui-route-authoring",
    severity: "warning",
    description: "UI route authoring rules: no Mermaid in @doc, no call list in @doc, Pop must not name parent.",
    messages: {
        noMermaid: paramMessage `@uiRoute namespace "${"ns"}" has a Mermaid diagram in @doc — remove it.`,
        noCallList: paramMessage `@uiRoute namespace "${"ns"}" @doc contains a {@...} reference to a declared @calls target "${"ref"}" — call-graph edges belong on action ops, not in namespace prose.`,
        popNamesParent: paramMessage `Pop action op "${"opName"}" @doc contains a {@...} reference to @uiRoute namespace "${"ref"}" — Pop must not name the route it returns to.`,
        callsRouteTarget: paramMessage `@calls target "${"ref"}" is a @uiRoute namespace — routes are navigated to, not called.`,
        pathOnModal: paramMessage `@uiRoute namespace "${"ns"}" carries a path field but has no URL-addressable context — remove the path from non-addressable surfaces.`,
        navNotReturnType: paramMessage `Action op "${"opName"}" in @uiRoute namespace returns void — action ops must return UiCommand.`,
    },
    // a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.rule_ui_route_authoring
    create(context) {
        // a10n:index:call Products.A10nBlueprintEngine.Features.UiRouteAuthoring.path_discriminates_addressable
        // a10n:index:call Products.A10nBlueprintEngine.Features.UiRouteAuthoring.nav_return_type
        // a10n:index:call Products.A10nBlueprintEngine.Features.UiRouteAuthoring.route_not_callable
        // a10n:index:call Products.A10nBlueprintEngine.Features.UiRouteAuthoring.composition_via_nesting
        // a10n:index:call Products.A10nBlueprintEngine.Features.UiRouteAuthoring.pop_is_anonymous
        // a10n:index:call Products.A10nBlueprintEngine.Features.UiRouteAuthoring.no_mermaid_doc
        // a10n:index:call Products.A10nBlueprintEngine.Features.UiRouteAuthoring.no_call_list
        return {
            // a10n:index:call packages.blueprint-typespec.src.rules.ui-route-authoring.rule.namespace
            // a10n:index:call packages.blueprint-typespec.src.rules.ui-route-authoring.rule.operation
            namespace(ns) {
                if (!isUiRouteNs(ns))
                    return;
                // path-on-modal: nested @uiRoute must not carry a path
                // a10n:blueprint Products.A10nBlueprintEngine.Features.UiRouteAuthoring.path_discriminates_addressable
                if (isNestedInsideRoute(ns) && getUiRoutePathArg(ns) !== undefined) {
                    context.reportDiagnostic({ messageId: "pathOnModal", format: { ns: ns.name }, target: ns });
                }
                const doc = getDoc(context.program, ns);
                if (!doc)
                    return;
                // no-mermaid-in-doc: Mermaid blocks are declared with ```mermaid fences
                // a10n:blueprint Products.A10nBlueprintEngine.Features.UiRouteAuthoring.no_mermaid_doc
                if (/```mermaid/i.test(doc)) {
                    context.reportDiagnostic({ messageId: "noMermaid", format: { ns: ns.name }, target: ns });
                    return;
                }
                // no-call-list: @doc must not contain a {@...} reference to any op that is a
                // declared @calls target of an action op in this namespace.
                // Only {@...} refs are detectable — bare words are not validated.
                // a10n:blueprint Products.A10nBlueprintEngine.Features.UiRouteAuthoring.no_call_list
                const callTargetFqns = getCallTargetFqns(ns);
                if (callTargetFqns.size > 0) {
                    for (const ref of extractInlineRefs(doc)) {
                        if (callTargetFqns.has(ref)) {
                            context.reportDiagnostic({ messageId: "noCallList", format: { ns: ns.name, ref }, target: ns });
                            return;
                        }
                        // Also resolve the ref and check if the resolved op is a call target
                        const resolved = resolveInlineRef(context.program, ref, ns);
                        if (resolved?.kind === "Operation") {
                            const fqn = resolved.namespace
                                ? `${nsToFqn(resolved.namespace)}.${resolved.name}`
                                : resolved.name;
                            if (callTargetFqns.has(fqn) || callTargetFqns.has(resolved.name)) {
                                context.reportDiagnostic({ messageId: "noCallList", format: { ns: ns.name, ref }, target: ns });
                                return;
                            }
                        }
                    }
                }
            },
            operation(op) {
                if (!isActionOpInRoute(op))
                    return;
                const retType = op.returnType;
                const rtKind = retType?.kind;
                const rtName = (retType?.name ?? retType?.node?.id?.sv ?? "");
                // nav-not-return-type: action ops must not return void (@tool exempt)
                // a10n:blueprint Products.A10nBlueprintEngine.Features.UiRouteAuthoring.nav_return_type
                const isVoid = rtKind === "Intrinsic" || rtName === "void" || rtName === "";
                if (isVoid && !hasDecorator(op, "@tool")) {
                    context.reportDiagnostic({ messageId: "navNotReturnType", format: { opName: op.name }, target: op });
                }
                // pop-names-parent: if return type is Pop, @doc must not contain a {@...}
                // reference to any @uiRoute namespace (which would name a specific parent route)
                // a10n:blueprint Products.A10nBlueprintEngine.Features.UiRouteAuthoring.pop_is_anonymous
                if (rtName === "Pop") {
                    const doc = getDoc(context.program, op);
                    if (doc) {
                        for (const ref of extractInlineRefs(doc)) {
                            const resolved = resolveInlineRef(context.program, ref, op.namespace);
                            if (resolved?.kind === "Namespace" && isUiRouteNs(resolved)) {
                                context.reportDiagnostic({ messageId: "popNamesParent", format: { opName: op.name, ref }, target: op });
                                break;
                            }
                        }
                    }
                }
                // calls-route-target: @calls must not target @uiRoute namespaces
                // a10n:blueprint Products.A10nBlueprintEngine.Features.UiRouteAuthoring.route_not_callable
                // a10n:blueprint Products.A10nBlueprintEngine.Features.UiRouteAuthoring.composition_via_nesting
                for (const dec of op.decorators) {
                    if (dec.definition?.name !== "@calls")
                        continue;
                    for (const arg of dec.args) {
                        const v = arg.value;
                        const items = v?.valueKind === "ArrayValue" ? v.values : (v?.kind === "Tuple" ? v.values : [v]);
                        for (const item of items ?? []) {
                            if (item?.kind === "Namespace" && isUiRouteNs(item)) {
                                context.reportDiagnostic({ messageId: "callsRouteTarget", format: { ref: item.name }, target: op });
                            }
                        }
                    }
                }
            },
        };
    },
});
