// @a10n/blueprint-typespec/dist/src/rules/worker-type-spec.rule.js
import { createRule, paramMessage } from "@typespec/compiler";
// ── Internal helpers ──────────────────────────────────────────────────────────
function hasDecorator(node, name) {
    return (node?.decorators ?? []).some((d) => d.definition?.name === name);
}
function nsToFqn(ns) {
    const parts = [];
    let cur = ns;
    while (cur && cur.name) {
        parts.unshift(cur.name);
        cur = cur.namespace;
    }
    return parts.join(".");
}
function opFqn(op) {
    if (!op.namespace)
        return op.name;
    const nsFqn = nsToFqn(op.namespace);
    return nsFqn ? `${nsFqn}.${op.name}` : op.name;
}
function modelFqn(model) {
    if (!model.namespace)
        return model.name;
    const nsFqn = nsToFqn(model.namespace);
    return nsFqn ? `${nsFqn}.${model.name}` : model.name;
}
/** True if op is inside a @worker namespace (direct parent). */
function isWorkerOp(op) {
    return !!(op.namespace && hasDecorator(op.namespace, "@worker"));
}
/** True if op carries @publishes */
function hasPublishes(op) {
    return op.decorators.some(d => d.definition?.name === "@publishes");
}
/** True if op carries @subscribes */
function hasSubscribes(op) {
    return op.decorators.some(d => d.definition?.name === "@subscribes");
}
/** Normalize array-like value from decorator arg. */
function normalizeItems(v) {
    if (v?.valueKind === "ArrayValue")
        return v.values ?? [];
    if (v?.kind === "Tuple")
        return v.values ?? [];
    return [v];
}
/** Extract @publishes target models from an op's decorator AST. */
function getPublishesModels(op) {
    const result = [];
    for (const dec of op.decorators) {
        if (dec.definition?.name !== "@publishes")
            continue;
        for (const arg of dec.args ?? []) {
            for (const item of normalizeItems(arg.value)) {
                if (item?.kind === "Model")
                    result.push(item);
                // Also handle UnionVariant — the variant itself has a type
                if (item?.kind === "UnionVariant" && item.type?.kind === "Model")
                    result.push(item.type);
            }
        }
    }
    return result;
}
/** Extract @subscribes target model from an op's first parameter type. */
function getSubscribesModel(op) {
    // @subscribes means the op's first param carries an event type (Model)
    // Look at the first parameter's type
    for (const [, param] of op.parameters.properties) {
        const t = param.type;
        if (t?.kind === "Model")
            return t;
        if (t?.kind === "UnionVariant" && t.type?.kind === "Model")
            return t.type;
        break;
    }
    return undefined;
}
function buildWorkerIndex(program) {
    const idx = {
        workerOps: [],
        workerOpFqns: new Set(),
        publishedModelFqns: new Set(),
        subscriptionMap: new Map(),
    };
    function addWorkerOp(op) {
        idx.workerOps.push(op);
        idx.workerOpFqns.add(opFqn(op));
        idx.workerOpFqns.add(op.name);
    }
    function addPublishedModel(m) {
        idx.publishedModelFqns.add(modelFqn(m));
        idx.publishedModelFqns.add(m.name);
    }
    function addToSubscriptionMap(eventModel, workerOp) {
        const fqn = modelFqn(eventModel);
        const name = eventModel.name;
        for (const key of [fqn, name]) {
            if (!idx.subscriptionMap.has(key))
                idx.subscriptionMap.set(key, []);
            idx.subscriptionMap.get(key).push(workerOp);
        }
    }
    function walkNs(ns) {
        for (const [, op] of ns.operations) {
            // Collect all published models (from any op)
            if (hasPublishes(op)) {
                for (const m of getPublishesModels(op))
                    addPublishedModel(m);
            }
            // Collect worker ops
            if (isWorkerOp(op)) {
                addWorkerOp(op);
                // Build subscription map
                if (hasSubscribes(op)) {
                    const subscribed = getSubscribesModel(op);
                    if (subscribed)
                        addToSubscriptionMap(subscribed, op);
                }
            }
        }
        for (const [, child] of ns.namespaces)
            walkNs(child);
    }
    walkNs(program.getGlobalNamespaceType());
    return idx;
}
// ── Rule ──────────────────────────────────────────────────────────────────────
export const workerTypeSpecRule = createRule({
    name: "worker-type-spec",
    severity: "warning",
    description: "Enforces @worker op correctness: void return, @subscribes, effect, and live events.",
    messages: {
        workerReturnsVoid: paramMessage `@worker op "${"opName"}" returns non-void type "${"returnType"}" — worker ops are fire-and-forget and must return void.`,
        workerHasSubscription: paramMessage `@worker op "${"opName"}" has no @subscribes — a worker op with no subscription has no trigger and is dead code.`,
        workerHasEffect: paramMessage `@worker op "${"opName"}" has no @reads, @writes, or @calls — a no-op handler the spec cannot verify.`,
        subscribesLiveEvent: paramMessage `@worker op "${"opName"}" subscribes to event "${"ref"}" which no op ever publishes — the subscription will never fire.`,
        subscribesOpOnly: paramMessage `@subscribes on op "${"opName"}" is outside a @worker namespace — only ops inside @worker namespaces may carry @subscribes.`,
    },
    // a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.rule_worker_type_spec
    // a10n:blueprint Products.A10nBlueprintEngine.Features.WorkerTypeSpec.worker_returns_void
    // a10n:blueprint Products.A10nBlueprintEngine.Features.WorkerTypeSpec.worker_has_subscription
    // a10n:blueprint Products.A10nBlueprintEngine.Features.WorkerTypeSpec.worker_has_effect
    // a10n:blueprint Products.A10nBlueprintEngine.Features.WorkerTypeSpec.subscribes_live_event
    // a10n:blueprint Products.A10nBlueprintEngine.Features.WorkerTypeSpec.subscribes_op_only
    create(context) {
        let idx = null;
        function getIdx() {
            if (!idx)
                idx = buildWorkerIndex(context.program);
            return idx;
        }
        return {
            // a10n:index:call packages.blueprint-typespec.src.rules.worker-type-spec.rule.operation
            operation(op) {
                // ── 1. worker-returns-void ─────────────────────────────────────────
                // ── 2. worker-has-subscription ────────────────────────────────────
                // ── 3. worker-has-effect ──────────────────────────────────────────
                if (isWorkerOp(op)) {
                    // 1: must return void
                    const rt = op.returnType;
                    const isVoid = rt?.kind === "Intrinsic" && rt.name === "void";
                    if (!isVoid) {
                        const returnTypeName = rt ? (rt.name ?? rt.kind ?? "unknown") : "unknown";
                        // a10n:blueprint Products.A10nBlueprintEngine.Features.WorkerTypeSpec.worker_returns_void
                        context.reportDiagnostic({
                            messageId: "workerReturnsVoid",
                            format: { opName: op.name, returnType: returnTypeName },
                            target: op,
                        });
                    }
                    // 2: must have @subscribes
                    if (!hasSubscribes(op)) {
                        // a10n:blueprint Products.A10nBlueprintEngine.Features.WorkerTypeSpec.worker_has_subscription
                        context.reportDiagnostic({
                            messageId: "workerHasSubscription",
                            format: { opName: op.name },
                            target: op,
                        });
                    }
                    // 3: must have at least one @reads, @writes, or @calls
                    const hasEffect = op.decorators.some(d => ["@reads", "@writes", "@calls"].includes(d.definition?.name ?? ""));
                    if (!hasEffect) {
                        // a10n:blueprint Products.A10nBlueprintEngine.Features.WorkerTypeSpec.worker_has_effect
                        context.reportDiagnostic({
                            messageId: "workerHasEffect",
                            format: { opName: op.name },
                            target: op,
                        });
                    }
                    // ── 4. subscribes-live-event ────────────────────────────────────
                    if (hasSubscribes(op)) {
                        const i = getIdx();
                        const subscribedModel = getSubscribesModel(op);
                        if (subscribedModel) {
                            const fqn = modelFqn(subscribedModel);
                            const name = subscribedModel.name;
                            if (!i.publishedModelFqns.has(fqn) && !i.publishedModelFqns.has(name)) {
                                // a10n:blueprint Products.A10nBlueprintEngine.Features.WorkerTypeSpec.subscribes_live_event
                                context.reportDiagnostic({
                                    messageId: "subscribesLiveEvent",
                                    format: { opName: op.name, ref: fqn || name },
                                    target: op,
                                });
                            }
                        }
                    }
                    return; // worker ops handled — skip further checks below
                }
                // ── 5. subscribes-op-only ──────────────────────────────────────────
                // Non-worker op carrying @subscribes
                if (hasSubscribes(op)) {
                    // a10n:blueprint Products.A10nBlueprintEngine.Features.WorkerTypeSpec.subscribes_op_only
                    context.reportDiagnostic({
                        messageId: "subscribesOpOnly",
                        format: { opName: op.name },
                        target: op,
                    });
                }
            },
        };
    },
});
