// @a10n/blueprint-typespec/src/decorators/component.ts
import { DecoratorContext, EnumValue, Namespace, Operation, Program, StringValue } from "@typespec/compiler";
import { toKebab } from "../validators.js";

// ── Suffix ↔ ComponentType mapping ─────────────────────────────────────────
// Components.<Name> must end with the kebab suffix that matches its type.
// Checked by toKebab(segs[1]).endsWith(suffix).

const TYPE_SUFFIXES: Record<string, { allowed: string[]; example: string }> = {
  Agent:   { allowed: ["-agent"],              example: "Agent"   },
  Service: { allowed: ["-service", "-cli"],    example: "Service" },
  Store:   { allowed: ["-store"],              example: "Store"   },
  Sdk:     { allowed: ["-sdk"],                example: "Sdk"     },
  Ui:      { allowed: ["-ui"],                 example: "Ui"      },
};


function resolveStringValue(v: unknown): string {
  if (typeof v === "string") return v;
  const val = v as EnumValue | StringValue;
  if (val && typeof val === "object" && "valueKind" in val) {
    if (val.valueKind === "EnumValue") return (val as EnumValue).value.name;
    if (val.valueKind === "StringValue") return (val as StringValue).value;
  }
  return String(v);
}
import {
  validateTarget,
  validateDecoratorNotDuplicated,
  getNamespaceFqn,
  hasDecorator,
  hasAncestorWithDecorator,
  validateInlineRefs,
} from "../validators.js";
import { reportDiagnostic } from "../lib.js";

const COMPONENT_KEY = Symbol("blueprint.component");
const DEPENDS_ON_KEY = Symbol("blueprint.dependsOn");
const TESTING_KEY = Symbol.for("blueprint.testing");
const FLOW_KEY = Symbol("blueprint.flow");
const UI_ROUTE_KEY = Symbol("blueprint.uiRoute");

// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_component
export function $component(
  ctx: DecoratorContext,
  target: Namespace,
  config: { type: string; source?: string; language?: string },
): void {
  if (!validateTarget(ctx, target, "@component", "Namespace")) {
    return;
  }

  if (!validateDecoratorNotDuplicated(ctx, target, "@component")) {
    return;
  }

  // Validate namespace structure: @component must be one of:
  //   Components.<Name>                       — first-party component
  //   Dependencies.<Dep>.<Name>               — legacy stub nested inside a dependency
  //   Dependencies.<Dep>.Components.<Name>    — component nested inside a dependency,
  //                                             scoped under an explicit `Components` segment
  const fqn = getNamespaceFqn(target);
  const segs = fqn.split(".");
  const isFirstParty = segs.length === 2 && segs[0] === "Components";
  const isInDependencyLegacy = segs.length === 3 && segs[0] === "Dependencies";
  const isInDependencyScoped =
    segs.length === 4 && segs[0] === "Dependencies" && segs[2] === "Components";
  if (!isFirstParty && !isInDependencyLegacy && !isInDependencyScoped) {
    reportDiagnostic(ctx.program, {
      code: "invalid-namespace-prefix",
      format: {
        decorator: "component",
        expectedPrefix:
          "Components.<Name>, Dependencies.<Dep>.<Name>, or Dependencies.<Dep>.Components.<Name>",
        actualFqn: fqn,
      },
      target,
    });
    return;
  }

  // Validate name suffix matches the declared type.
  // Components.ReviewApiService (type: Service) → kebab "review-api-service" must end in "-service"
  // Dependency-nested components (Dependencies.<Dep>.<Name>) are exempt from the suffix rule.
  const componentName = segs[segs.length - 1];
  const typeStr = resolveStringValue(config.type);
  const suffixRule = TYPE_SUFFIXES[typeStr];
  if (suffixRule && isFirstParty) {
    const kebab = toKebab(componentName);
    if (!suffixRule.allowed.some((s) => kebab.endsWith(s))) {
      reportDiagnostic(ctx.program, {
        code: "component-name-suffix",
        format: {
          name: componentName,
          expected: suffixRule.allowed.join(", "),
          type: typeStr,
          example: suffixRule.example,
        },
        target,
      });
      return;
    }
  }

  if (config.source !== undefined) {
    const src = resolveStringValue(config.source);
    try {
      const u = new URL(src);
      if (u.protocol !== "http:" && u.protocol !== "https:" && u.protocol !== "file:") throw new Error();
    } catch {
      reportDiagnostic(ctx.program, {
        code: "invalid-url-format",
        format: { value: src, field: "source" },
        target,
      });
      return;
    }
  }

  if (config.language !== undefined) {
    const validLanguages = ["go", "typescript", "python"];
    const lang = resolveStringValue(config.language);
    if (!validLanguages.includes(lang)) {
      reportDiagnostic(ctx.program, {
        code: "invalid-enum-value",
        format: {
          value: lang,
          field: "language",
          valid: validLanguages.join(", "),
        },
        target,
      });
      return;
    }
  }

  ctx.program.stateMap(COMPONENT_KEY).set(target, config);
}

// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_depends_on
// a10n:blueprint Products.A10nBlueprintEngine.Features.SystemAuthoring.component_depends_on
export function $dependsOn(
  ctx: DecoratorContext,
  target: Namespace | Operation,
  deps: Namespace[],
): void {
  if (target.kind === "Namespace") {
    // On Namespace: must have @component, @testSuite, @product, or @feature on the same namespace
    if (!hasDecorator(target, "@component") && !hasDecorator(target, "@testSuite") && !hasDecorator(target, "@product") && !hasDecorator(target, "@feature")) {
      const fqn = getNamespaceFqn(target);
      reportDiagnostic(ctx.program, {
        code: "invalid-decorator-context",
        format: {
          decorator: "dependsOn",
          context: "component",
          actualFqn: fqn,
        },
        target,
      });
      return;
    }
  } else {
    // On Operation: must have @component or @testSuite ancestor
    const ns = target.namespace;
    const inComponent = ns && (hasDecorator(ns, "@component") || hasAncestorWithDecorator(ns, "@component"));
    const inTestSuite = ns && (hasDecorator(ns, "@testSuite") || hasAncestorWithDecorator(ns, "@testSuite"));
    if (!inComponent && !inTestSuite) {
      reportDiagnostic(ctx.program, {
        code: "invalid-namespace-nesting",
        format: {
          decorator: "dependsOn",
          requiredParent: "component",
          actualFqn: target.namespace ? getNamespaceFqn(target.namespace) : "<global>",
        },
        target,
      });
      return;
    }
  }
  // TypeSpec passes Namespace[] params as a Tuple type (not a plain JS array).
  // Tuple.values: Type[] holds the actual Namespace type objects.
  // Normalise to plain Namespace[] so the stateMap always holds a clean array.
  const raw = deps as unknown as Namespace[] | { kind?: string; values?: Namespace[] };
  const resolved: Namespace[] = Array.isArray(raw)
    ? raw
    : ((raw as { kind?: string; values?: Namespace[] }).values ?? []);
  ctx.program.stateMap(DEPENDS_ON_KEY).set(target, resolved);
}

// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_testing
export function $testing(
  ctx: DecoratorContext,
  target: Namespace,
  dep: Namespace,
  config: { approach: string; reasoning: string },
): void {
  if (!validateTarget(ctx, target, "@testing", "Namespace")) {
    return;
  }

  // Validate that @testing is applied to the same namespace as @component
  if (!hasDecorator(target, "@component")) {
    const fqn = getNamespaceFqn(target);
    reportDiagnostic(ctx.program, {
      code: "invalid-decorator-context",
      format: {
        decorator: "testing",
        context: "component",
        actualFqn: fqn,
      },
      target,
    });
    return;
  }

  // Validate inline refs in reasoning
  // Note: We continue even if validation fails so decorator continues to record the config
  validateInlineRefs(ctx, config.reasoning, "reasoning", target);

  const map = ctx.program.stateMap(TESTING_KEY);
  const existing: Array<{ dep: Namespace; approach: string; reasoning: string }> =
    map.get(target) ?? [];
  map.set(target, [...existing, { dep, ...config }]);
}

// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_flow
export function $flow(ctx: DecoratorContext, target: Operation): void {
  // must be inside a @component namespace
  if (!target.namespace || !hasAncestorWithDecorator(target.namespace, "@component") && !hasDecorator(target.namespace, "@component")) {
    reportDiagnostic(ctx.program, {
      code: "invalid-namespace-nesting",
      format: {
        decorator: "flow",
        requiredParent: "component",
        actualFqn: target.namespace ? getNamespaceFqn(target.namespace) : "<global>",
      },
      target,
    });
    return;
  }
  ctx.program.stateMap(FLOW_KEY).set(target, true);
}

// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_ui_route
export function $uiRoute(
  ctx: DecoratorContext,
  target: Namespace,
  config?: { path?: string; designs?: { url: string }[] },
): void {
  // must be inside a @component(type: Ui) namespace OR nested inside another @uiRoute namespace
  const hasUiComponentAncestor = hasAncestorWithDecorator(target, "@component") || hasDecorator(target, "@component");
  const hasUiRouteAncestor = hasAncestorWithDecorator(target, "@uiRoute");

  if (!hasUiComponentAncestor && !hasUiRouteAncestor) {
    reportDiagnostic(ctx.program, {
      code: "invalid-namespace-nesting",
      format: {
        decorator: "uiRoute",
        requiredParent: "component",
        actualFqn: getNamespaceFqn(target),
      },
      target,
    });
    return;
  }
  ctx.program.stateMap(UI_ROUTE_KEY).set(target, config ?? {});
}

// ── Getters (for emitters) ────────────────────────────────────────────────────

export function getComponentConfig(
  program: Program,
  target: Namespace,
): { type: string; source?: string; language?: string } | undefined {
  return program.stateMap(COMPONENT_KEY).get(target);
}

// a10n:blueprint:ignore
export function getDependsOn(
  program: Program,
  target: Namespace | Operation,
): Namespace[] | undefined {
  return program.stateMap(DEPENDS_ON_KEY).get(target);
}

// a10n:blueprint:ignore
export function getTestingStrategies(
  program: Program,
  target: Namespace,
): Array<{ dep: Namespace; approach: string; reasoning: string }> | undefined {
  return program.stateMap(TESTING_KEY).get(target);
}

// a10n:blueprint:ignore
export function isFlowOp(program: Program, target: Operation): boolean {
  return program.stateMap(FLOW_KEY).has(target);
}

// a10n:blueprint:ignore
export function getUiRouteConfig(
  program: Program,
  target: Namespace,
): { path?: string; designs?: { url: string }[] } | undefined {
  return program.stateMap(UI_ROUTE_KEY).get(target);
}
