// @a10n/blueprint-typespec/src/decorators/dependency.ts
import { DecoratorContext, Namespace, Program } from "@typespec/compiler";
import {
  validateTarget,
  validateDecoratorNotDuplicated,
  getNamespaceFqn,
  validateInlineRefs,
} from "../validators.js";
import { reportDiagnostic } from "../lib.js";

const DEPENDENCY_KEY = Symbol("blueprint.dependency");

// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_dependency
export function $dependency(
  ctx: DecoratorContext,
  target: Namespace,
  config: {
    title: string;
    type: string;
    status: string;
    sources: Array<{ url: string; description: string }>;
    why?: string[];
    how?: string[];
    gotchas?: string[];
  },
): void {
  if (!validateTarget(ctx, target, "@dependency", "Namespace")) {
    return;
  }

  if (!validateDecoratorNotDuplicated(ctx, target, "@dependency")) {
    return;
  }

  // Validate namespace structure: @dependency must be exactly Dependencies.<Name>
  const fqn = getNamespaceFqn(target);
  const segs = fqn.split(".");
  if (segs.length !== 2 || segs[0] !== "Dependencies") {
    reportDiagnostic(ctx.program, {
      code: "invalid-namespace-prefix",
      format: {
        decorator: "dependency",
        expectedPrefix: "Dependencies.<Name>",
        actualFqn: fqn,
      },
      target,
    });
    return;
  }

  // Validate inline refs in documentation fields
  for (const str of [...(config.why ?? []), ...(config.how ?? []), ...(config.gotchas ?? [])]) {
    validateInlineRefs(ctx, str, "documentation", target);
  }

  ctx.program.stateMap(DEPENDENCY_KEY).set(target, config);
}

// a10n:blueprint:ignore
export function getDependencyConfig(
  program: Program,
  target: Namespace,
): { title: string; type: string; status: string; sources: Array<{ url: string; description: string }>; why?: string[]; how?: string[]; gotchas?: string[] } | undefined {
  return program.stateMap(DEPENDENCY_KEY).get(target);
}
