// @a10n/blueprint-typespec/src/decorators/eval.ts
import { DecoratorContext, Model, Namespace, Operation, Program } from "@typespec/compiler";
import {
  validateTarget,
  validateDecoratorNotDuplicated,
  getNamespaceFqn,
  hasDecorator,
  hasAncestorWithDecorator,
  findAncestorWithDecorator,
  stripNsPrefix,
  validateInlineRefs,
  getTypeFilePath,
  isExemptPath,
} from "../validators.js";
import { reportDiagnostic } from "../lib.js";

/** Read resolved @calls targets from the operation's decorator AST (avoids circular import with call-graph.ts). */
function getCallsFromAst(op: Operation): Operation[] {
  const result: Operation[] = [];
  for (const dec of op.decorators) {
    if (dec.definition?.name !== "@calls") continue;
    for (const arg of dec.args ?? []) {
      const v = arg.value as any;
      const items: any[] = v?.valueKind === "ArrayValue" ? v.values : (v?.kind === "Tuple" ? v.values : [v]);
      for (const item of items) {
        if (item?.kind === "Operation") result.push(item as Operation);
      }
    }
  }
  return result;
}

/**
 * Check if a namespace is a @component with type === "Agent".
 * Walks the decorator AST the same way isStoreComponent does in call-graph.ts.
 */
// a10n:blueprint:ignore
function isAgentComponent(ns: Namespace): boolean {
  for (const dec of (ns as any).decorators ?? []) {
    if (dec.definition?.name !== "@component") continue;
    const arg = dec.args?.[0]?.value;
    if (!arg) continue;
    let typeVal: any;
    if (arg.valueKind === "ObjectValue") {
      const props = (arg as any).properties as Map<string, any>;
      typeVal = props.get("type")?.value;
    } else if (arg.kind === "Object") {
      typeVal = (arg as any).properties?.type;
    }
    if (!typeVal) continue;
    const typeStr = typeVal.valueKind === "EnumValue"
      ? typeVal.value?.name
      : typeVal.value?.name ?? String(typeVal);
    if (typeStr === "Agent") return true;
  }
  return false;
}

const EVAL_KEY = Symbol("blueprint.eval");
const SIGNAL_KEY = Symbol("blueprint.signal");
const SCORING_KEY = Symbol("blueprint.scoring");
const CONFIG_IMPACT_KEY = Symbol("blueprint.configImpact");
const FEATURE_CONFIG_KEY = Symbol("blueprint.featureConfig");
const EVAL_SUITE_KEY = Symbol("blueprint.evalSuite");
const DATASET_KEY = Symbol("blueprint.dataset");

// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_eval
// a10n:blueprint Products.A10nBlueprintEngine.Features.SystemAuthoring.proof_has_parent
export function $eval(
  ctx: DecoratorContext,
  target: Namespace,
  config: { title: string },
): void {
  if (!validateTarget(ctx, target, "@eval", "Namespace")) {
    return;
  }

  if (!validateDecoratorNotDuplicated(ctx, target, "@eval")) {
    return;
  }

  // Validate nesting: @eval must be nested inside a @component namespace
  if (!hasAncestorWithDecorator(target, "@component")) {
    const fqn = getNamespaceFqn(target);
    reportDiagnostic(ctx.program, {
      code: "invalid-namespace-nesting",
      format: {
        decorator: "eval",
        requiredParent: "component",
        actualFqn: fqn,
      },
      target,
    });
    return;
  }

  // Validate naming pattern: relative to @component ancestor must be Evals.<FeatureName>.E<NNN>
  const componentNs = findAncestorWithDecorator(target, "@component")!;
  const componentFqn = getNamespaceFqn(componentNs);
  const fqn = getNamespaceFqn(target);
  const relative = stripNsPrefix(fqn, componentFqn);
  if (!/^Evals\.[A-Z][A-Za-z0-9]*\.E\d{3}$/.test(relative)) {
    reportDiagnostic(ctx.program, {
      code: "invalid-eval-namespace",
      format: { actual: relative },
      target,
    });
    return;
  }

  ctx.program.stateMap(EVAL_KEY).set(target, config);
}

// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_eval_suite
// a10n:blueprint Products.A10nBlueprintEngine.Features.SystemAuthoring.eval_file_numbering
// a10n:blueprint Products.A10nBlueprintEngine.Features.SystemAuthoring.case_has_priority
export function $evalSuite(
  ctx: DecoratorContext,
  target: Operation,
  config: { title: string; priority: string; expects: unknown[] },
): void {
  // a10n:blueprint Products.A10nBlueprintEngine.Features.EvalSuiteAuthoring.under_component
  // must be inside a @component namespace
  if (!target.namespace || (!hasDecorator(target.namespace, "@component") && !hasAncestorWithDecorator(target.namespace, "@component"))) {
    reportDiagnostic(ctx.program, {
      code: "invalid-namespace-nesting",
      format: {
        decorator: "evalSuite",
        requiredParent: "component",
        actualFqn: target.namespace ? getNamespaceFqn(target.namespace) : "<global>",
      },
      target,
    });
    return;
  }
  // a10n:blueprint Products.A10nBlueprintEngine.Features.EvalSuiteAuthoring.suite_no_projects
    // suite_no_projects: @evalSuite must not carry @projects
  if (target.decorators.some((d: any) => d.definition?.name === "@projects")) {
    reportDiagnostic(ctx.program, {
      code: "evalSuite-no-projects",
      format: { opName: target.name },
      target,
    });
  }

  // a10n:blueprint Products.A10nBlueprintEngine.Features.EvalSuiteAuthoring.suite_returns_void
  // a10n:blueprint Products.A10nBlueprintEngine.Features.EvalSuiteAuthoring.suite_has_inputs
  // a10n:blueprint Products.A10nBlueprintEngine.Features.EvalSuiteAuthoring.inputs_model_local
    // Validate that the op has exactly one parameter referencing a Model (suite_has_inputs)
  const params = target.parameters.properties;
  const paramArray = Array.from(params.values());
  if (paramArray.length !== 1 || paramArray[0]!.type.kind !== "Model") {
    reportDiagnostic(ctx.program, {
      code: "evalSuite-missing-inputs",
      format: { opName: target.name },
      target,
    });
  } else {
    // inputs_example_required: every field of the Inputs model must carry @example
    // so the eval emitter has a default value when none is supplied at runtime.
    const inputModel = paramArray[0]!.type as import("@typespec/compiler").Model;
    for (const [fieldName, prop] of inputModel.properties) {
      if (prop.optional) continue; // optional fields have an implicit default of "omitted"
      const hasExample = (prop.decorators ?? []).some(
        (d: any) => d.definition?.name === "@example" || d.definition?.name === "example"
      );
      if (!hasExample) {
        reportDiagnostic(ctx.program, {
          code: "inputs-field-missing-example",
          format: { opName: target.name, field: fieldName },
          target: prop,
        });
      }
    }
  }
  // Validate inline refs in each expects string
  const expectsArr = Array.isArray(config.expects)
    ? config.expects
    : (config.expects as any)?.values ?? [];
  const expectsStrings: string[] = [];
  for (const str of expectsArr) {
    const s = typeof str === "string" ? str : (str as any)?.value ?? String(str);
    validateInlineRefs(ctx, s, "expects", target.namespace!);
    expectsStrings.push(s);
  }
  ctx.program.stateMap(EVAL_SUITE_KEY).set(target, config);

  // has_calls / has_dataset_calls / every_scorer_gated:
  // Read @calls targets from the op's AST — works regardless of decorator application order
  // because target.decorators always contains all applied decorators.
  const resolved = getCallsFromAst(target);

  // a10n:blueprint Products.A10nBlueprintEngine.Features.EvalSuiteAuthoring.has_calls
    // has_calls: suite must have at least one @calls target
  if (resolved.length === 0) {
    reportDiagnostic(ctx.program, {
      code: "evalSuite-no-calls",
      format: { opName: target.name },
      target,
    });
    return;
  }

  // a10n:blueprint Products.A10nBlueprintEngine.Features.EvalSuiteAuthoring.has_dataset_calls
    // has_dataset_calls: at least one resolved target must be a @dataset op
  const hasDataset = resolved.some((op) =>
    op.decorators.some((d: any) => d.definition?.name === "@dataset" || d.definition?.name === "dataset")
  );
  if (!hasDataset) {
    reportDiagnostic(ctx.program, {
      code: "evalSuite-no-dataset-calls",
      format: { opName: target.name },
      target,
    });
  }

  // a10n:blueprint Products.A10nBlueprintEngine.Features.EvalSuiteAuthoring.every_scorer_gated
  // a10n:blueprint Products.A10nBlueprintEngine.Features.EvalSuiteAuthoring.calls_scorers_only
  // a10n:blueprint Products.A10nBlueprintEngine.Features.EvalSuiteAuthoring.calls_datasets_ungated
  // a10n:blueprint Products.A10nBlueprintEngine.Features.EvalSuiteAuthoring.suite_has_expects
  // a10n:blueprint Products.A10nBlueprintEngine.Features.EvalSuiteAuthoring.expects_typed_refs
    // every_scorer_gated: each Score op in @calls must appear as a {@...} ref in expects
  const scoreOps = resolved.filter((op) =>
    op.name === "Score" && op.namespace && hasDecorator(op.namespace, "@eval")
  );
  for (const scoreOp of scoreOps) {
    // Build the namespace FQN for the Score op to produce a ref like "E001.Score" or full FQN
    const evalNsFqn = scoreOp.namespace ? getNamespaceFqn(scoreOp.namespace) : "";
    const scorerFqn = evalNsFqn ? `${evalNsFqn}.${scoreOp.name}` : scoreOp.name;
    // Check whether any expects string contains a {@...} reference to this scorer
    // The ref could be the short name or any suffix of the FQN
    const fqnParts = scorerFqn.split(".");
    const gated = expectsStrings.some((s) => {
      const refPattern = /\{@([\w.]+)\}/g;
      let m: RegExpExecArray | null;
      refPattern.lastIndex = 0;
      while ((m = refPattern.exec(s)) !== null) {
        const ref = m[1];
        for (let i = 0; i < fqnParts.length; i++) {
          if (ref === fqnParts.slice(i).join(".")) return true;
        }
      }
      return false;
    });
    if (!gated) {
      reportDiagnostic(ctx.program, {
        code: "evalSuite-scorer-ungated",
        format: { opName: target.name, scorer: scorerFqn },
        target,
      });
    }
  }

  // a10n:blueprint Products.A10nBlueprintEngine.Features.EvalSuiteAuthoring.suite_file_path
  // a10n:blueprint Products.A10nBlueprintEngine.Features.EvalSuiteAuthoring.suite_with_signals
  // a10n:blueprint Products.A10nBlueprintEngine.Features.EvalSuiteAuthoring.suite_namespace
  // a10n:blueprint Products.A10nBlueprintEngine.Features.EvalSuiteAuthoring.suite_has_priority
  // a10n:blueprint Products.A10nBlueprintEngine.Features.EvalSuiteAuthoring.suite_no_reads
  // a10n:blueprint Products.A10nBlueprintEngine.Features.EvalSuiteAuthoring.suite_no_writes
    // suite_file_path: @evalSuite must be in components/[component]/evals/[feature]/NNN_[purpose].tsp
  const suiteFilePath = getTypeFilePath(target) ?? "";
  if (suiteFilePath && !isExemptPath(suiteFilePath)) {
    const normalizedSuitePath = suiteFilePath.replace(/\\/g, "/");
    if (!/components\/[^/]+\/evals\/[^/]+\/\d{3}_[^/]+\.tsp$/.test(normalizedSuitePath)) {
      reportDiagnostic(ctx.program, {
        code: "evalSuite-file-path",
        format: { opName: target.name, actual: normalizedSuitePath },
        target,
      });
    }
  }
}

// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_dataset
// a10n:blueprint Products.A10nBlueprintEngine.Features.DatasetAuthoring.dataset_returns_void
// a10n:blueprint Products.A10nBlueprintEngine.Features.DatasetAuthoring.no_adequacy_strategy
// a10n:blueprint Products.A10nBlueprintEngine.Features.DatasetAuthoring.no_inline_items
// a10n:blueprint Products.A10nBlueprintEngine.Features.DatasetAuthoring.one_storage_per
export function $dataset(
  ctx: DecoratorContext,
  target: Operation,
  config: { title: string },
): void {
  // a10n:blueprint Products.A10nBlueprintEngine.Features.DatasetAuthoring.under_component
  // must be inside a @component namespace
  if (!target.namespace || (!hasDecorator(target.namespace, "@component") && !hasAncestorWithDecorator(target.namespace, "@component"))) {
    reportDiagnostic(ctx.program, {
      code: "invalid-namespace-nesting",
      format: {
        decorator: "dataset",
        requiredParent: "component",
        actualFqn: target.namespace ? getNamespaceFqn(target.namespace) : "<global>",
      },
      target,
    });
    return;
  }
  // a10n:blueprint Products.A10nBlueprintEngine.Features.DatasetAuthoring.id_shape
  // Validate op name: must match [a-z][a-z0-9_-]+
  if (!/^[a-z][a-z0-9_-]+$/.test(target.name)) {
    reportDiagnostic(ctx.program, {
      code: "dataset-id-shape",
      format: { opName: target.name },
      target,
    });
  }
  // a10n:blueprint Products.A10nBlueprintEngine.Features.DatasetAuthoring.file_path
  // a10n:blueprint Products.A10nBlueprintEngine.Features.DatasetAuthoring.file_numbering
  // Validate file path: must be in components/[component]/datasets/[feature]/NNN_[purpose].tsp
  const filePath = getTypeFilePath(target) ?? "";
  if (filePath && !isExemptPath(filePath)) {
    const normalizedPath = filePath.replace(/\\/g, "/");
    if (!/components\/[^/]+\/datasets\/[^/]+\/\d{3}_[^/]+\.tsp$/.test(normalizedPath)) {
      reportDiagnostic(ctx.program, {
        code: "dataset-file-path",
        format: { opName: target.name, actual: normalizedPath },
        target,
      });
    }
  }
  // title must be non-empty
  const titleStr = typeof config.title === "string" ? config.title : (config.title as any)?.value ?? "";
  if (!titleStr.trim()) {
    reportDiagnostic(ctx.program, {
      code: "dataset-empty-title",
      format: { opName: target.name },
      target,
    });
  }
  ctx.program.stateMap(DATASET_KEY).set(target, config);
}

// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_signal
// a10n:blueprint Products.A10nBlueprintEngine.Features.SystemAuthoring.signal_id_shape
// a10n:blueprint Products.A10nBlueprintEngine.Features.SystemAuthoring.signal_has_method
export function $signal(
  ctx: DecoratorContext,
  target: Operation,
  config: {
    title: string;
    method: string;
    criteria: string;
    examples: { value: unknown; description: string }[];
    projects?: unknown[];
  },
): void {
  // must be directly inside an @eval namespace
  if (!target.namespace || !hasDecorator(target.namespace, "@eval")) {
    reportDiagnostic(ctx.program, {
      code: "invalid-decorator-context",
      format: {
        decorator: "signal",
        context: "eval",
        actualFqn: target.namespace ? getNamespaceFqn(target.namespace) : "<global>",
      },
      target,
    });
    return;
  }

  // Op name must follow E{NNN}_{NN}
  if (!/^E\d{3}_\d{2}$/.test(target.name)) {
    reportDiagnostic(ctx.program, {
      code: "signal-id-shape",
      format: { opName: target.name },
      target,
    });
  }

  validateInlineRefs(ctx, config.criteria, "criteria", target.namespace);
  ctx.program.stateMap(SIGNAL_KEY).set(target, config);
}

// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_scoring
export function $scoring(
  ctx: DecoratorContext,
  target: Operation,
  config: { formula: string; examples: string[]; boundaryCases: string[] },
): void {
  // must be directly inside an @eval namespace
  if (!target.namespace || !hasDecorator(target.namespace, "@eval")) {
    reportDiagnostic(ctx.program, {
      code: "invalid-decorator-context",
      format: {
        decorator: "scoring",
        context: "eval",
        actualFqn: target.namespace ? getNamespaceFqn(target.namespace) : "<global>",
      },
      target,
    });
    return;
  }
  // Score op must not return void — it must return a numeric score value (float, numeric scalar, etc.)
  const rt = target.returnType;
  if (rt.kind === "Intrinsic" && (rt as any).name === "void") {
    reportDiagnostic(ctx.program, {
      code: "scoring-op-returns-void",
      format: { opName: target.name },
      target,
    });
  }

  for (const str of [config.formula, ...config.examples, ...config.boundaryCases]) {
    validateInlineRefs(ctx, str, "scoring", target.namespace);
  }

  // formula_local_signals: every signal ID in the formula (pattern \bE\d{3}_\d{2}\b) must
  // be an op in this @eval namespace.
  // a10n:blueprint Products.A10nBlueprintEngine.Features.EvaluationAuthoring.formula_local_signals
  const formulaSignalPattern = /\bE\d{3}_\d{2}\b/g;
  const evalNs = target.namespace;
  let m: RegExpExecArray | null;
  formulaSignalPattern.lastIndex = 0;
  while ((m = formulaSignalPattern.exec(config.formula)) !== null) {
    const signalId = m[0];
    if (!evalNs.operations.has(signalId)) {
      reportDiagnostic(ctx.program, {
        code: "scoring-formula-external-signal",
        format: { opName: target.name, signal: signalId },
        target,
      });
    }
  }

  ctx.program.stateMap(SCORING_KEY).set(target, config);
}

// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_config_impact
export function $configImpact(
  ctx: DecoratorContext,
  target: Operation,
  param: unknown,
  config: { relationship: string },
): void {
  // must be directly inside an @eval namespace
  if (!target.namespace || !hasDecorator(target.namespace, "@eval")) {
    reportDiagnostic(ctx.program, {
      code: "invalid-decorator-context",
      format: {
        decorator: "configImpact",
        context: "eval",
        actualFqn: target.namespace ? getNamespaceFqn(target.namespace) : "<global>",
      },
      target,
    });
    return;
  }
  const map = ctx.program.stateMap(CONFIG_IMPACT_KEY);
  const existing: Array<{ param: unknown; relationship: string }> =
    map.get(target) ?? [];
  map.set(target, [...existing, { param, relationship: config.relationship }]);
}

// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_feature_config
export function $featureConfig(ctx: DecoratorContext, target: Model): void {
  // must be inside a @feature namespace
  if (!target.namespace || !hasAncestorWithDecorator(target.namespace, "@feature") && !hasDecorator(target.namespace, "@feature")) {
    reportDiagnostic(ctx.program, {
      code: "invalid-namespace-nesting",
      format: {
        decorator: "featureConfig",
        requiredParent: "feature",
        actualFqn: target.namespace ? getNamespaceFqn(target.namespace) : "<global>",
      },
      target,
    });
    return;
  }
  ctx.program.stateMap(FEATURE_CONFIG_KEY).set(target, true);
}

// a10n:blueprint:ignore
export function getEvalConfig(
  program: Program,
  target: Namespace,
): { title: string } | undefined {
  return program.stateMap(EVAL_KEY).get(target);
}

// a10n:blueprint:ignore
export function getSignalsInEval(
  program: Program,
  ns: Namespace,
): Array<{ op: Operation; config: { title: string; method: string; criteria: string; examples: Array<{ value: unknown; description: string }>; projects?: unknown[] } }> {
  const result: Array<{ op: Operation; config: any }> = [];
  for (const op of ns.operations.values()) {
    const config = program.stateMap(SIGNAL_KEY).get(op);
    if (config) result.push({ op, config });
  }
  return result;
}

// a10n:blueprint:ignore
export function getScoringOpInEval(
  program: Program,
  ns: Namespace,
): Operation | undefined {
  for (const op of ns.operations.values()) {
    if (program.stateMap(SCORING_KEY).has(op)) return op;
  }
  return undefined;
}

// a10n:blueprint:ignore
export function getScoringConfig(
  program: Program,
  op: Operation,
): { formula: string; examples: string[]; boundaryCases: string[] } | undefined {
  return program.stateMap(SCORING_KEY).get(op);
}

// a10n:blueprint:ignore
export function getEvalSuiteConfig(
  program: Program,
  target: Operation,
): { title: string; priority: string; expects: unknown[] } | undefined {
  return program.stateMap(EVAL_SUITE_KEY).get(target);
}

// a10n:blueprint:ignore
export function getDatasetConfig(
  program: Program,
  target: Operation,
): { title: string } | undefined {
  return program.stateMap(DATASET_KEY).get(target);
}

// a10n:blueprint:ignore
export function getFeatureConfigModels(
  program: Program,
  ns: Namespace,
): Model[] {
  const result: Model[] = [];
  for (const model of ns.models.values()) {
    if (program.stateMap(FEATURE_CONFIG_KEY).has(model)) {
      result.push(model);
    }
  }
  return result;
}
