// @a10n/blueprint-typespec/src/decorators/file-store.ts
import { DecoratorContext, Model, Namespace, Program } from "@typespec/compiler";
import { getNamespaceFqn, hasDecorator, hasAncestorWithDecorator } from "../validators.js";
import { reportDiagnostic } from "../lib.js";

/**
 * Walk the namespace tree rooted at `ns` and collect all namespaces that have
 * STORAGE_KEY set in the stateMap.
 */
function collectStorageNamespaces(program: Program, ns: Namespace, key: symbol): Namespace[] {
  const result: Namespace[] = [];
  if (program.stateMap(key).has(ns)) result.push(ns);
  for (const [, child] of ns.namespaces) {
    result.push(...collectStorageNamespaces(program, child, key));
  }
  return result;
}

/**
 * Walk ancestor namespaces of `ns` (not including itself) to find the first
 * one that has `key` set in stateMap.
 */
function findAncestorWithKey(program: Program, ns: Namespace, key: symbol): Namespace | undefined {
  let cur = ns.namespace;
  while (cur && cur.name) {
    if (program.stateMap(key).has(cur)) return cur;
    cur = cur.namespace;
  }
  return undefined;
}

/**
 * Recursively collect all models in `ns` subtree that have STORAGE_OBJECT_KEY set.
 */
function collectStorageObjects(program: Program, ns: Namespace, key: symbol): Model[] {
  const result: Model[] = [];
  for (const [, model] of ns.models) {
    if (program.stateMap(key).has(model)) result.push(model);
  }
  for (const [, child] of ns.namespaces) {
    result.push(...collectStorageObjects(program, child, key));
  }
  return result;
}

/**
 * Check if a namespace is a @component with type === "Store".
 * Uses the same decorator-AST walk as isStoreComponent in call-graph.ts.
 */
function isStoreComponent(ns: Namespace): boolean {
  for (const dec of (ns as any).decorators ?? []) {
    if (dec.definition?.name !== "@component") continue;
    const arg = dec.args?.[0]?.value;
    if (!arg) continue;
    let typeVal: any;
    if (arg.valueKind === "ObjectValue") {
      const props = (arg as any).properties as Map<string, any>;
      typeVal = props.get("type")?.value;
    } else if (arg.kind === "Object") {
      typeVal = (arg as any).properties?.type;
    }
    if (!typeVal) continue;
    const typeStr = typeVal.valueKind === "EnumValue"
      ? typeVal.value?.name
      : typeVal.value?.name ?? String(typeVal);
    if (typeStr === "Store") return true;
  }
  return false;
}

/**
 * Find the nearest @component ancestor (or self) namespace.
 */
function findComponentAncestorNs(ns: Namespace): Namespace | undefined {
  let cur: Namespace | undefined = ns;
  while (cur && cur.name) {
    if (hasDecorator(cur, "@component")) return cur;
    cur = cur.namespace;
  }
  return undefined;
}

const STORAGE_KEY = Symbol("blueprint.storage");
const STORAGE_GROUP_KEY = Symbol("blueprint.storageGroup");
const STORAGE_OBJECT_KEY = Symbol("blueprint.storageObject");

/**
 * Extract placeholder model names from a path/root/prefix string.
 * Placeholders have the form {@ModelName.fieldName} or {@Ns.ModelName.fieldName}.
 * Returns an array of { ref: string, modelName: string } where modelName is all
 * segments except the last (the field name).
 */
function extractPathPlaceholders(str: string | undefined): Array<{ ref: string; modelName: string }> {
  if (!str) return [];
  const result: Array<{ ref: string; modelName: string }> = [];
  const pattern = /\{@([^}]+)\}/g;
  let m: RegExpExecArray | null;
  pattern.lastIndex = 0;
  while ((m = pattern.exec(str)) !== null) {
    const ref = m[1];
    const parts = ref.split(".");
    if (parts.length >= 2) {
      // All segments except the last form the model path; last is the field name
      const modelName = parts.slice(0, -1).join(".");
      result.push({ ref, modelName });
    }
  }
  return result;
}

/**
 * Try to find a model by its name (possibly qualified) in the component's namespace tree.
 * Accepts bare name (last segment), full FQN segments, or a path relative to any ancestor.
 */
function findModelInComponent(componentNs: Namespace, modelName: string): boolean {
  const nameParts = modelName.split(".");
  const lastName = nameParts[nameParts.length - 1];

  function walkNs(ns: Namespace): boolean {
    // Check if this namespace has a model matching the last segment of modelName
    if (ns.models.has(lastName)) {
      // If only one segment, any match is fine
      if (nameParts.length === 1) return true;
      // If multiple segments, verify the namespace path ends with the prefix
      const nsFqn = getNamespaceFqn(ns);
      const prefix = nameParts.slice(0, -1).join(".");
      if (nsFqn.endsWith(prefix) || nsFqn === prefix || nsFqn.includes(prefix)) return true;
    }
    for (const [, child] of ns.namespaces) {
      if (walkNs(child)) return true;
    }
    return false;
  }

  return walkNs(componentNs);
}

/**
 * Check path_refs_typed: every {@Model.field} placeholder in a path string must resolve
 * to a model in the component. Also reject bare {xxx} placeholders (not typed refs).
 */
function validatePathPlaceholders(
  program: Program,
  pathStr: string | undefined,
  componentNs: Namespace | undefined,
  target: Namespace | Model,
): void {
  if (!pathStr) return;
  // Reject bare {xxx} placeholders that are not typed {@...} refs
  const barePattern = /\{(?!@)([^}]+)\}/g;
  let bm: RegExpExecArray | null;
  barePattern.lastIndex = 0;
  while ((bm = barePattern.exec(pathStr)) !== null) {
    reportDiagnostic(program, {
      code: "storage-bare-placeholder",
      format: { placeholder: bm[0] },
      target,
    });
  }
  if (!componentNs) return;
  for (const { ref, modelName } of extractPathPlaceholders(pathStr)) {
    if (!findModelInComponent(componentNs, modelName)) {
      reportDiagnostic(program, {
        code: "storage-placeholder-unresolved",
        format: { ref },
        target,
      });
    }
  }
}

// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_storage
// a10n:blueprint Products.A10nBlueprintEngine.Features.FileStoreAuthoring.group_prefix_only
export function $storage(
  ctx: DecoratorContext,
  target: Namespace,
  config: { provider: unknown; root?: unknown },
): void {
  const program = ctx.program;

  // Must be inside a @component namespace
  // a10n:blueprint Products.A10nBlueprintEngine.Features.FileStoreAuthoring.storage_in_store
  if (!hasDecorator(target, "@component") && !hasAncestorWithDecorator(target, "@component")) {
    reportDiagnostic(program, {
      code: "invalid-namespace-nesting",
      format: {
        decorator: "storage",
        requiredParent: "component",
        actualFqn: getNamespaceFqn(target),
      },
      target,
    });
    return;
  }
  // The component must be a Store
  // a10n:blueprint Products.A10nBlueprintEngine.Features.FileStoreAuthoring.storage_in_store
  const componentNs = findComponentAncestorNs(target);
  if (componentNs && !isStoreComponent(componentNs)) {
    reportDiagnostic(program, {
      code: "invalid-namespace-nesting",
      format: {
        decorator: "storage",
        requiredParent: "component of type Store",
        actualFqn: getNamespaceFqn(target),
      },
      target,
    });
    return;
  }
  // one_per_store: walk the component namespace tree to find any existing @storage
  // a10n:blueprint Products.A10nBlueprintEngine.Features.FileStoreAuthoring.one_per_store
  if (componentNs) {
    const existing = collectStorageNamespaces(program, componentNs, STORAGE_KEY);
    if (existing.length > 0) {
      const compName = getNamespaceFqn(componentNs).split(".").pop() ?? getNamespaceFqn(componentNs);
      reportDiagnostic(program, {
        code: "storage-duplicate",
        format: { component: compName },
        target,
      });
      return;
    }
  }

  // path_refs_typed: validate {@Model.field} placeholders in root
  // a10n:blueprint Products.A10nBlueprintEngine.Features.FileStoreAuthoring.path_refs_typed
  const rootRawStr = (config.root as any)?.valueKind === "StringValue" ? (config.root as any).value : typeof config.root === "string" ? config.root : undefined;
  validatePathPlaceholders(program, rootRawStr, componentNs, target);

  // Normalize provider: EnumValue → string
  // a10n:blueprint Products.A10nBlueprintEngine.Features.FileStoreAuthoring.storage_has_provider
  const providerRaw = config.provider as any;
  const providerStr: string = providerRaw?.valueKind === "EnumValue"
    ? providerRaw.value?.name ?? String(providerRaw)
    : typeof providerRaw === "string" ? providerRaw : String(providerRaw);
  program.stateMap(STORAGE_KEY).set(target, { provider: providerStr, root: rootRawStr });
}

// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_storage_group
// a10n:blueprint Products.A10nBlueprintEngine.Features.FileStoreAuthoring.group_inside_storage
export function $storageGroup(
  ctx: DecoratorContext,
  target: Namespace,
  config: { prefix: string },
): void {
  const program = ctx.program;

  // Must be inside a @component namespace
  if (!hasDecorator(target, "@component") && !hasAncestorWithDecorator(target, "@component")) {
    reportDiagnostic(program, {
      code: "invalid-namespace-nesting",
      format: {
        decorator: "storageGroup",
        requiredParent: "component",
        actualFqn: getNamespaceFqn(target),
      },
      target,
    });
    return;
  }

  // storage_xor_group: target cannot also carry @storage (check AST decorators for sibling)
  // a10n:blueprint Products.A10nBlueprintEngine.Features.FileStoreAuthoring.storage_xor_group
  const hasStorageAst = (target as any).decorators?.some(
    (d: any) => d.definition?.name === "@storage",
  );
  if (hasStorageAst) {
    reportDiagnostic(program, {
      code: "storage-xor-group",
      format: {},
      target,
    });
    return;
  }

  // group_inside_storage: must have an ancestor namespace with @storage
  // a10n:blueprint Products.A10nBlueprintEngine.Features.FileStoreAuthoring.group_inside_storage
  const storageAncestor = findAncestorWithKey(program, target, STORAGE_KEY);
  if (!storageAncestor) {
    reportDiagnostic(program, {
      code: "invalid-namespace-nesting",
      format: {
        decorator: "storageGroup",
        requiredParent: "storage",
        actualFqn: getNamespaceFqn(target),
      },
      target,
    });
    return;
  }

  // path_refs_typed: validate {@Model.field} placeholders in prefix
  const componentNs = findComponentAncestorNs(target);
  validatePathPlaceholders(program, config.prefix, componentNs, target);

  program.stateMap(STORAGE_GROUP_KEY).set(target, config);
}

// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_storage_object
// a10n:blueprint Products.A10nBlueprintEngine.Features.FileStoreAuthoring.object_has_address
// a10n:blueprint Products.A10nBlueprintEngine.Features.FileStoreAuthoring.body_no_path
// a10n:blueprint Products.A10nBlueprintEngine.Features.FileStoreAuthoring.no_path_collision
export function $storageObject(
  ctx: DecoratorContext,
  target: Model,
  config: { path: string },
): void {
  const program = ctx.program;

  // Compute this model's resolved path: storagePrefix + "/" + config.path
  const ns = target.namespace;
  if (ns) {
    const { prefix } = resolveStoragePrefix(program, ns);
    const resolvedPath = [prefix, config.path].filter(Boolean).join("/");

    // no_path_collision: check all other @storageObject models in the same component
    // a10n:blueprint Products.A10nBlueprintEngine.Features.FileStoreAuthoring.no_path_collision
    // a10n:blueprint Products.A10nBlueprintEngine.Features.FileStoreAuthoring.object_has_address
    const componentNs = findComponentAncestorNs(ns);
    if (componentNs) {
      const siblings = collectStorageObjects(program, componentNs, STORAGE_OBJECT_KEY);
      for (const sibling of siblings) {
        if (sibling === target) continue;
        const siblingCfg = program.stateMap(STORAGE_OBJECT_KEY).get(sibling) as { path: string } | undefined;
        if (!siblingCfg) continue;
        const siblingNs = sibling.namespace;
        if (!siblingNs) continue;
        const { prefix: siblingPrefix } = resolveStoragePrefix(program, siblingNs);
        const siblingResolved = [siblingPrefix, siblingCfg.path].filter(Boolean).join("/");
        if (siblingResolved === resolvedPath) {
          reportDiagnostic(program, {
            code: "storage-path-collision",
            format: { modelName: target.name, path: resolvedPath, other: sibling.name },
            target,
          });
          // Still register so we don't double-report on subsequent siblings
          break;
        }
      }
    }

    // path_refs_typed: validate {@Model.field} placeholders in path
    validatePathPlaceholders(program, config.path, componentNs ?? undefined, target);

    // body_no_path: model must not have properties that are also used as path placeholders
    // a10n:blueprint Products.A10nBlueprintEngine.Features.FileStoreAuthoring.body_no_path
    // Extract field names used as path placeholders (last segment of each placeholder)
    const pathFieldNames = new Set<string>();
    const placeholderPattern = /\{@([^}]+)\}/g;
    let pm: RegExpExecArray | null;
    placeholderPattern.lastIndex = 0;
    while ((pm = placeholderPattern.exec(config.path)) !== null) {
      const parts = pm[1].split(".");
      if (parts.length >= 2) {
        pathFieldNames.add(parts[parts.length - 1]);
      }
    }
    for (const fieldName of pathFieldNames) {
      if (target.properties.has(fieldName)) {
        reportDiagnostic(program, {
          code: "storage-body-has-path-field",
          format: { modelName: target.name, field: fieldName },
          target,
        });
      }
    }
  }

  program.stateMap(STORAGE_OBJECT_KEY).set(target, config);
}

// a10n:blueprint:ignore
export function getStorageConfig(
  program: Program,
  target: Namespace,
): { provider: string; root?: string } | undefined {
  return program.stateMap(STORAGE_KEY).get(target);
}

// a10n:blueprint:ignore
export function getStorageGroupConfig(
  program: Program,
  target: Namespace,
): { prefix: string } | undefined {
  return program.stateMap(STORAGE_GROUP_KEY).get(target);
}

export function getStorageObjectConfig(
  program: Program,
  target: Model,
): { path: string } | undefined {
  return program.stateMap(STORAGE_OBJECT_KEY).get(target);
}

/**
 * Walk up a namespace's ancestors to find the nearest @storage ancestor and
 * collect all @storageGroup prefixes in between. Returns the composed path
 * prefix: `root + "/" + prefix1 + "/" + prefix2 + ...`  (empty string if no
 * @storage ancestor found).
 */
export function resolveStoragePrefix(program: Program, ns: Namespace): { found: boolean; prefix: string } {
  const groups: string[] = [];
  let cur: Namespace | undefined = ns;
  while (cur && cur.name) {
    const storageGroupCfg = program.stateMap(STORAGE_GROUP_KEY).get(cur);
    if (storageGroupCfg) {
      groups.unshift((storageGroupCfg as { prefix: string }).prefix);
    }
    const storageCfg = program.stateMap(STORAGE_KEY).get(cur);
    if (storageCfg) {
      const root = (storageCfg as { provider: string; root?: string }).root ?? "";
      const parts = [root, ...groups].filter(Boolean);
      return { found: true, prefix: parts.join("/") };
    }
    cur = cur.namespace;
  }
  return { found: false, prefix: "" };
}
