// @a10n/blueprint-typespec/src/decorators/graph.ts
import { DecoratorContext, Namespace, Operation, Program } from "@typespec/compiler";
import {
  getNamespaceFqn,
  hasDecorator,
  hasAncestorWithDecorator,
} from "../validators.js";
import { reportDiagnostic } from "../lib.js";

const GRAPH_KEY = Symbol("blueprint.graph");
const NODE_KEY = Symbol("blueprint.node");
const GRAPH_NODE_KEY = Symbol("blueprint.graphNode");
const PREDICT_KEY = Symbol("blueprint.predict");

// ── helpers ──────────────────────────────────────────────────────────────────

function isAgentComponent(ns: Namespace): boolean {
  for (const dec of (ns as any).decorators ?? []) {
    if (dec.definition?.name !== "@component") continue;
    const arg = dec.args?.[0]?.value;
    if (!arg) continue;
    let typeVal: any;
    if (arg.valueKind === "ObjectValue") {
      typeVal = (arg.properties as Map<string, any>)?.get("type")?.value;
    } else if (arg.kind === "Object") {
      typeVal = (arg as any).properties?.type;
    }
    if (!typeVal) continue;
    const name =
      typeVal.valueKind === "EnumValue"
        ? typeVal.value?.name
        : typeVal.value?.name ?? String(typeVal);
    if (name === "Agent") return true;
  }
  return false;
}

// ── @graph ───────────────────────────────────────────────────────────────────

// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_graph
export function $graph(ctx: DecoratorContext, target: Namespace): void {
  // Must be nested inside a @component(Agent) namespace
  const hasComponentAncestor =
    hasDecorator(target, "@component") ||
    hasAncestorWithDecorator(target, "@component");

  if (!hasComponentAncestor) {
    reportDiagnostic(ctx.program, {
      code: "invalid-namespace-nesting",
      format: {
        decorator: "graph",
        requiredParent: "component",
        actualFqn: getNamespaceFqn(target),
      },
      target,
    });
    return;
  }

  // Must be inside an Agent component
  // a10n:blueprint Products.A10nBlueprintEngine.Features.GraphAuthoring.graph_in_agent
  let compNs: Namespace | undefined = target.namespace;
  while (compNs && compNs.name) {
    if (hasDecorator(compNs, "@component")) break;
    compNs = compNs.namespace;
  }
  if (compNs && !isAgentComponent(compNs)) {
    reportDiagnostic(ctx.program, {
      code: "invalid-namespace-nesting",
      format: {
        decorator: "graph",
        requiredParent: "component of type Agent",
        actualFqn: getNamespaceFqn(target),
      },
      target,
    });
    return;
  }

  ctx.program.stateMap(GRAPH_KEY).set(target, true);
}

// ── @node ────────────────────────────────────────────────────────────────────

// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_node
export function $node(
  ctx: DecoratorContext,
  target: Operation,
  config?: { is_entrypoint?: boolean; is_join?: boolean },
): void {
  // Must be directly inside a @graph namespace
  // a10n:blueprint Products.A10nBlueprintEngine.Features.GraphAuthoring.node_in_graph
  if (!target.namespace || !hasDecorator(target.namespace, "@graph")) {
    reportDiagnostic(ctx.program, {
      code: "invalid-namespace-nesting",
      format: {
        decorator: "node",
        requiredParent: "graph",
        actualFqn: target.namespace ? getNamespaceFqn(target.namespace) : "<global>",
      },
      target,
    });
    return;
  }
  ctx.program.stateMap(NODE_KEY).set(target, config ?? {});
  if (config) ctx.program.stateMap(GRAPH_NODE_KEY).set(target, config);
}

// ── @predict ─────────────────────────────────────────────────────────────────

// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_predict
export function $predict(ctx: DecoratorContext, target: Operation): void {
  // Must be directly inside a @graph namespace
  if (!target.namespace || !hasDecorator(target.namespace, "@graph")) {
    reportDiagnostic(ctx.program, {
      code: "invalid-namespace-nesting",
      format: {
        decorator: "predict",
        requiredParent: "graph",
        actualFqn: target.namespace ? getNamespaceFqn(target.namespace) : "<global>",
      },
      target,
    });
    return;
  }

  // predict-untyped-output: output must not be plain string
  // a10n:blueprint Products.A10nBlueprintEngine.Features.PredictContextBoundary.predict_declares_output
  const rt = target.returnType;
  const isPlainString =
    rt.kind === "Scalar" && (rt as any).name === "string";
  if (isPlainString) {
    reportDiagnostic(ctx.program, {
      code: "predict-untyped-output",
      format: { opName: target.name },
      target,
    });
  }

  ctx.program.stateMap(PREDICT_KEY).set(target, true);
}

// ── getters ──────────────────────────────────────────────────────────────────

// a10n:blueprint:ignore
export function isGraphNamespace(program: Program, ns: Namespace): boolean {
  return program.stateMap(GRAPH_KEY).has(ns);
}

// a10n:blueprint:ignore
export function isNodeOp(program: Program, op: Operation): boolean {
  return program.stateMap(NODE_KEY).has(op);
}

// a10n:blueprint:ignore
export function getGraphNodeConfig(
  program: Program,
  op: Operation,
): { is_entrypoint?: boolean; is_join?: boolean } | undefined {
  return program.stateMap(GRAPH_NODE_KEY).get(op);
}

// a10n:blueprint:ignore
export function isPredictOp(program: Program, op: Operation): boolean {
  return program.stateMap(PREDICT_KEY).has(op);
}
