// @a10n/blueprint-typespec/src/decorators/invariant.ts
import { DecoratorContext, Operation, Program } from "@typespec/compiler";
import { validateTarget, validateDecoratorNotDuplicated, validateInlineRefs, getNamespaceFqn, hasAncestorWithDecorator, hasDecorator } from "../validators.js";
import { reportDiagnostic } from "../lib.js";

const INVARIANT_KEY = Symbol("blueprint.invariant");

// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_invariant
// a10n:blueprint Products.A10nBlueprintEngine.Features.DomainAuthoring.invariant_predicate_field
// a10n:blueprint Products.A10nBlueprintEngine.Features.DomainAuthoring.invariant_why_field
export function $invariant(
  ctx: DecoratorContext,
  target: Operation,
  config: { predicate: string; why: string },
): void {
  if (!validateTarget(ctx, target, "@invariant", "Operation")) {
    return;
  }

  if (!validateDecoratorNotDuplicated(ctx, target, "@invariant")) {
    return;
  }

  // must be inside a @feature namespace
  if (!target.namespace || (!hasDecorator(target.namespace, "@feature") && !hasAncestorWithDecorator(target.namespace, "@feature"))) {
    reportDiagnostic(ctx.program, {
      code: "invalid-namespace-nesting",
      format: {
        decorator: "invariant",
        requiredParent: "feature",
        actualFqn: target.namespace ? getNamespaceFqn(target.namespace) : "<global>",
      },
      target,
    });
    return;
  }

  // Op name must be ≤3 snake_case words, no pure-numeric suffix
  // a10n:blueprint Products.A10nBlueprintEngine.Features.DomainAuthoring.inv_id_opname
  const parts = target.name.split("_");
  if (parts.length > 3 || /^\d+$/.test(parts[parts.length - 1])) {
    reportDiagnostic(ctx.program, {
      code: "invariant-id-shape",
      format: { opName: target.name },
      target,
    });
  }

  // Return type must be void
  // a10n:blueprint Products.A10nBlueprintEngine.Features.DomainAuthoring.invariant_returns_void
  const rt = target.returnType;
  if (!(rt.kind === "Intrinsic" && (rt as any).name === "void")) {
    reportDiagnostic(ctx.program, {
      code: "invariant-non-void-return",
      format: { opName: target.name, returnType: rt.kind },
      target,
    });
  }

  // a10n:todo Products.A10nBlueprintEngine.Features.DomainAuthoring.predicate_one_sentence — not yet implemented: single-sentence check requires NLP heuristics not available in TypeSpec compiler callbacks
  // a10n:todo Products.A10nBlueprintEngine.Features.DomainAuthoring.invariant_is_atomic — not yet implemented: atomicity is a prose/style rule not enforceable via structural decorator validation
  // a10n:todo Products.A10nBlueprintEngine.Features.DomainAuthoring.predicate_references_domain — not yet implemented: checking that predicate contains at least one @entity or enum ref requires type-system inspection beyond current validateInlineRefs
  // a10n:todo Products.A10nBlueprintEngine.Features.DomainAuthoring.why_is_reason — not yet implemented: detecting that `why` restates the predicate requires semantic comparison not available in TypeSpec decorator validation
  validateInlineRefs(ctx, config.predicate, "predicate", target.namespace);
  validateInlineRefs(ctx, config.why, "why", target.namespace);

  ctx.program.stateMap(INVARIANT_KEY).set(target, config);
}

export function getInvariantConfig(
  program: Program,
  target: Operation,
): { predicate: string; why: string } | undefined {
  return program.stateMap(INVARIANT_KEY).get(target);
}

// a10n:blueprint:ignore
export function getInvariantsInFeature(
  program: Program,
  ns: import("@typespec/compiler").Namespace,
): Array<{ name: string; config: { predicate: string; why: string } }> {
  const results: Array<{ name: string; config: { predicate: string; why: string } }> = [];
  for (const [opName, op] of ns.operations) {
    const config = program.stateMap(INVARIANT_KEY).get(op);
    if (config) {
      results.push({ name: opName, config });
    }
  }
  return results;
}
