// @a10n/blueprint-typespec/src/decorators/projects.ts
import { DecoratorContext, Model, ModelProperty, Operation, Program } from "@typespec/compiler";
import { validateDecoratorNotDuplicated } from "../validators.js";
import { reportDiagnostic } from "../lib.js";
import { isEntity } from "./entity.js";
import { getInvariantConfig } from "./invariant.js";
import { isProjectingOpKind } from "../projecting-op.js";

const PROJECTS_KEY = Symbol("blueprint.projects");

type ProjectsTarget = ModelProperty | Operation;
type ProjectsRef = Model | ModelProperty | Operation;

// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_projects
// a10n:blueprint Products.A10nBlueprintEngine.Features.SystemAuthoring.field_declares_proj
// a10n:blueprint Products.A10nBlueprintEngine.Features.SystemAuthoring.projection_targets_domain
// a10n:blueprint Products.A10nBlueprintEngine.Features.SystemAuthoring.proj_source_system
// a10n:blueprint Products.A10nBlueprintEngine.Features.SystemAuthoring.proj_field_compat
// a10n:blueprint Products.A10nBlueprintEngine.Features.SystemAuthoring.field_projects_field
// a10n:blueprint Products.A10nBlueprintEngine.Features.SystemAuthoring.projects_crosses_features
// a10n:blueprint Products.A10nBlueprintEngine.Features.SystemAuthoring.proj_not_namespace
export function $projects(
  ctx: DecoratorContext,
  target: ProjectsTarget,
  ...refs: Array<ProjectsRef>
): void {
  if (!validateDecoratorNotDuplicated(ctx, target as any, "@projects")) {
    return;
  }

  // @projects on a plain Model is not allowed — use it on individual fields instead.
  if ((target as any).kind === "Model") {
    reportDiagnostic(ctx.program, {
      code: "invalid-decorator-target",
      format: {
        decorator: "projects",
        expected: "ModelProperty or Operation",
        actual: "Model",
      },
      target: ctx.decoratorTarget,
    });
    return;
  }

  const resolved: Array<ProjectsRef> = refs;

  // On an Operation: must be a @testCase, @signal, @dataset, or a projecting-op-kind op
  if ((target as any).kind === "Operation") {
    const op = target as Operation;
    const isProofOp = op.decorators.some(
      (d) =>
        d.definition?.name === "@testCase" ||
        d.definition?.name === "@signal" ||
        d.definition?.name === "@dataset",
    );
    if (!isProofOp && !isProjectingOpKind(op)) {
      reportDiagnostic(ctx.program, {
        code: "projects-op-not-proof",
        format: { opName: op.name },
        target: ctx.decoratorTarget,
      });
      return;
    }
  }

  // Validate cross-use contract:
  // - On a ModelProperty: every ref must be a ModelProperty of an @entity model
  // - On an Operation: every ref must be an @invariant Operation
  const isFieldTarget = (target as any).kind === "ModelProperty";
  let valid = true;
  for (let i = 0; i < resolved.length; i++) {
    const ref = resolved[i] as any;
    if (isFieldTarget) {
      // ref must be a ModelProperty whose owning model carries @entity
      if (ref.kind !== "ModelProperty" || !isEntity(ctx.program, ref.model)) {
        reportDiagnostic(ctx.program, {
          code: "projects-field-must-target-entity-field",
          format: {
            ref: ref.name ?? String(i),
            kind: ref.kind ?? "unknown",
          },
          target: ctx.decoratorTarget,
        });
        valid = false;
      }
    } else {
      // ref must be an Operation decorated with @invariant
      if (ref.kind !== "Operation" || !getInvariantConfig(ctx.program, ref)) {
        reportDiagnostic(ctx.program, {
          code: "projects-op-must-target-invariant",
          format: {
            ref: ref.name ?? String(i),
            kind: ref.kind ?? "unknown",
          },
          target: ctx.decoratorTarget,
        });
        valid = false;
      }
    }
  }
  if (!valid) return;

  ctx.program.stateMap(PROJECTS_KEY).set(target, resolved);
}

// a10n:blueprint:ignore
function refToFqn(t: ProjectsRef): string {
  if (t.kind === "ModelProperty") {
    const prop = t as ModelProperty;
    const modelName = prop.model?.name ?? "";
    const parts: string[] = [];
    let ns = prop.model?.namespace;
    while (ns && ns.name) { parts.unshift(ns.name); ns = ns.namespace; }
    const nsFqn = parts.join(".");
    const base = nsFqn ? `${nsFqn}.${modelName}` : modelName;
    return base ? `${base}.${prop.name}` : prop.name;
  }
  if (t.kind === "Model") {
    const m = t as Model;
    const parts: string[] = [];
    let ns = m.namespace;
    while (ns && ns.name) { parts.unshift(ns.name); ns = ns.namespace; }
    const nsFqn = parts.join(".");
    return nsFqn ? `${nsFqn}.${m.name}` : m.name;
  }
  if (t.kind === "Operation") {
    const op = t as Operation;
    const parts: string[] = [];
    let ns = op.namespace;
    while (ns && ns.name) { parts.unshift(ns.name); ns = ns.namespace; }
    const nsFqn = parts.join(".");
    return nsFqn ? `${nsFqn}.${op.name}` : op.name;
  }
  return (t as any).name ?? "unknown";
}

// a10n:blueprint:ignore
export function getProjectsTargets(
  program: Program,
  target: ProjectsTarget,
): Array<ProjectsRef> | undefined {
  return program.stateMap(PROJECTS_KEY).get(target);
}

// a10n:blueprint:ignore
export function getProjectsTargetFqns(
  program: Program,
  target: ProjectsTarget,
): string[] | undefined {
  const targets = getProjectsTargets(program, target);
  if (!targets) return undefined;
  return targets.map(refToFqn);
}

// a10n:blueprint:ignore
export function hasProjects(program: Program, target: ProjectsTarget): boolean {
  return program.stateMap(PROJECTS_KEY).has(target);
}
