// @a10n/blueprint-typespec/dist/src/decorators/call-graph.js
import { getNamespaceFqn, hasAncestorWithDecorator, hasDecorator } from "../validators.js";
import { reportDiagnostic } from "../lib.js";
import { getStorageObjectConfig } from "./file-store.js";
const CALLS_KEY = Symbol("blueprint.calls");
const READS_KEY = Symbol("blueprint.reads");
const WRITES_KEY = Symbol("blueprint.writes");
// Helper to find the @component namespace that owns a given Operation
function findComponentAncestor(op) {
    let ns = op.namespace;
    while (ns && ns.name) {
        if (hasDecorator(ns, "@component"))
            return ns;
        ns = ns.namespace;
    }
    return undefined;
}
// Helper to find the @component namespace that owns a given Namespace
function findComponentAncestorNs(ns) {
    let cur = ns;
    while (cur && cur.name) {
        if (hasDecorator(cur, "@component"))
            return cur;
        cur = cur.namespace;
    }
    return undefined;
}
// Get the @component that owns a model (walk namespace ancestors)
function findModelComponent(model) {
    return model.namespace ? findComponentAncestorNs(model.namespace) : undefined;
}
// Check if a component has ComponentType.Store
function isStoreComponent(ns) {
    // Check the @component decorator's config.type
    for (const dec of ns.decorators ?? []) {
        if (dec.definition?.name !== "@component")
            continue;
        const arg = dec.args?.[0]?.value;
        if (!arg)
            continue;
        // arg is an ObjectValue — look for type field
        let typeVal;
        if (arg.valueKind === "ObjectValue") {
            const props = arg.properties;
            typeVal = props.get("type")?.value;
        }
        else if (arg.kind === "Object") {
            typeVal = arg.properties?.type;
        }
        if (!typeVal)
            continue;
        const typeStr = typeVal.valueKind === "EnumValue"
            ? typeVal.value?.name
            : typeVal.value?.name ?? String(typeVal);
        if (typeStr === "Store")
            return true;
    }
    return false;
}
// Normalize raw TypeSpec tuple/array decorator arg to a plain array
function normalizeArray(raw) {
    if (Array.isArray(raw))
        return raw;
    const r = raw;
    if (r?.kind === "Tuple")
        return r.values ?? [];
    if (r?.valueKind === "ArrayValue")
        return r.values ?? [];
    return [];
}
function getComponentType(ns) {
    for (const dec of ns.decorators ?? []) {
        if (dec.definition?.name !== "@component")
            continue;
        const arg = dec.args?.[0]?.value;
        if (!arg)
            continue;
        let typeVal;
        if (arg.valueKind === "ObjectValue") {
            typeVal = arg.properties?.get("type")?.value;
        }
        else if (arg.kind === "Object") {
            typeVal = arg.properties?.type;
        }
        if (!typeVal)
            continue;
        return typeVal.valueKind === "EnumValue" ? typeVal.value?.name : typeVal.value?.name ?? String(typeVal);
    }
    return undefined;
}
function classifyCaller(op) {
    const decs = op.decorators.map((d) => d.definition?.name);
    if (decs.includes("@evalSuite"))
        return "eval_suite";
    if (decs.includes("@scoring"))
        return "score";
    if (decs.includes("@signal")) {
        // probabilistic signal check: method === Probabilistic
        for (const dec of op.decorators) {
            if (dec.definition?.name !== "@signal")
                continue;
            const arg = dec.args?.[0]?.value;
            if (arg?.valueKind === "ObjectValue") {
                const method = arg.properties?.get("method")?.value;
                const methodName = method?.valueKind === "EnumValue" ? method.value?.name : undefined;
                if (methodName === "Probabilistic")
                    return "probabilistic_signal";
            }
        }
        return "other"; // deterministic signal — no @calls allowed
    }
    if (decs.includes("@tool"))
        return "tool";
    if (decs.includes("@node"))
        return "graph_node";
    if (decs.includes("@worker") || (op.namespace && hasDecorator(op.namespace, "@worker")))
        return "worker";
    const comp = findComponentAncestor(op);
    if (!comp)
        return "other";
    const compType = getComponentType(comp);
    // ui_action: inside @uiRoute
    let ns = op.namespace;
    while (ns && ns.name) {
        if (hasDecorator(ns, "@uiRoute"))
            return "ui_action";
        ns = ns.namespace;
    }
    // sdk_op: @operation in Sdk component
    if (compType === "Sdk" && decs.includes("@operation"))
        return "sdk_op";
    // protocol_entry: Http/Mcp/Commands sub-namespace of Service
    if (compType === "Service") {
        const chain = [];
        let cur = op.namespace;
        while (cur && cur !== comp && cur.name) {
            chain.unshift(cur.name);
            cur = cur.namespace;
        }
        if (chain.length > 0 && (chain[0] === "Http" || chain[0] === "Mcp" || chain[0] === "Commands"))
            return "protocol_entry";
    }
    return "other";
}
function classifyCallee(op) {
    const decs = op.decorators.map((d) => d.definition?.name);
    if (decs.includes("@dataset"))
        return "dataset";
    if (decs.includes("@scoring"))
        return "score";
    if (decs.includes("@signal"))
        return "signal";
    if (decs.includes("@predict"))
        return "predict";
    if (decs.includes("@interrupt"))
        return "interrupt";
    if (decs.includes("@tool"))
        return "tool";
    if (decs.includes("@worker") || (op.namespace && hasDecorator(op.namespace, "@worker")))
        return "worker";
    // ui_action
    let ns = op.namespace;
    while (ns && ns.name) {
        if (hasDecorator(ns, "@uiRoute"))
            return "ui_action";
        ns = ns.namespace;
    }
    const comp = findComponentAncestor(op);
    if (!comp)
        return "other";
    const compType = getComponentType(comp);
    // graph_node entry
    if (decs.includes("@node")) {
        const isEntry = op.decorators.some((d) => {
            if (d.definition?.name !== "@node")
                return false;
            const arg = d.args?.[0]?.value;
            if (arg?.valueKind === "ObjectValue") {
                const prop = arg.properties?.get("is_entrypoint");
                return prop?.value?.value === true || prop?.value === true;
            }
            return false;
        });
        return isEntry ? "cross_boundary" : "other"; // non-entry nodes cannot be @calls targets
    }
    // sdk_op
    if (compType === "Sdk" && decs.includes("@operation"))
        return "cross_boundary";
    // protocol entry (Http/Mcp/Commands)
    if (compType === "Service") {
        const chain = [];
        let cur = op.namespace;
        while (cur && cur !== comp && cur.name) {
            chain.unshift(cur.name);
            cur = cur.namespace;
        }
        if (chain.length > 0 && (chain[0] === "Http" || chain[0] === "Mcp" || chain[0] === "Commands"))
            return "cross_boundary";
    }
    return "other";
}
// a10n:blueprint Products.A10nBlueprintEngine.Features.CallsWhitelist.protocol_entry_calls
// a10n:blueprint Products.A10nBlueprintEngine.Features.CallsWhitelist.ui_action_calls
// a10n:blueprint Products.A10nBlueprintEngine.Features.CallsWhitelist.tool_calls
// a10n:blueprint Products.A10nBlueprintEngine.Features.CallsWhitelist.node_calls
// a10n:blueprint Products.A10nBlueprintEngine.Features.CallsWhitelist.worker_calls
// a10n:blueprint Products.A10nBlueprintEngine.Features.CallsWhitelist.sdk_op_calls
// a10n:blueprint Products.A10nBlueprintEngine.Features.CallsWhitelist.eval_suite_calls
// a10n:blueprint Products.A10nBlueprintEngine.Features.CallsWhitelist.score_calls
// a10n:blueprint Products.A10nBlueprintEngine.Features.CallsWhitelist.probabilistic_signal_calls
// a10n:todo Products.A10nBlueprintEngine.Features.CallsWhitelist.ui_action_calls — no proof test yet; enforcement present
// a10n:todo Products.A10nBlueprintEngine.Features.CallsWhitelist.sdk_op_calls — no proof test yet; enforcement present
// a10n:todo Products.A10nBlueprintEngine.Features.CallsWhitelist.probabilistic_signal_calls — no proof test yet; enforcement present
function isCalleeAllowed(callerKind, calleeKind) {
    switch (callerKind) {
        case "protocol_entry": return calleeKind === "cross_boundary";
        case "ui_action": return calleeKind === "cross_boundary" && true; // but only protocol subset — checked separately
        case "tool": return calleeKind === "cross_boundary" || calleeKind === "interrupt";
        case "graph_node": return calleeKind === "cross_boundary" || calleeKind === "tool" || calleeKind === "predict" || calleeKind === "interrupt";
        case "worker": return calleeKind === "cross_boundary";
        case "sdk_op": return calleeKind === "cross_boundary";
        case "eval_suite": return calleeKind === "cross_boundary" || calleeKind === "dataset" || calleeKind === "predict" || calleeKind === "score";
        case "score": return calleeKind === "signal";
        case "probabilistic_signal": return calleeKind === "predict";
        default: return false;
    }
}
// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_calls
// Note: calls_lists_all / calls_no_prose / calls_absent_pure are projected onto
// Components.A10nBlueprintService.Commands.Mapping.check (the mapping-check stage)
// — the call-graph completeness check compares each declared @calls edge against
// the real impl call graph via the StageOperationCoverage gates.
export function $calls(ctx, target, ops) {
    const items = normalizeArray(ops);
    // Validate: source must be inside a @component
    // a10n:blueprint Products.A10nBlueprintEngine.Features.OperationCallGraph.calls_op_only
    if (!target.namespace || (!hasDecorator(target.namespace, "@component") && !hasAncestorWithDecorator(target.namespace, "@component"))) {
        reportDiagnostic(ctx.program, {
            code: "invalid-namespace-nesting",
            format: { decorator: "calls", requiredParent: "component", actualFqn: target.namespace ? getNamespaceFqn(target.namespace) : "<global>" },
            target,
        });
        return;
    }
    const callerKind = classifyCaller(target);
    const resolved = [];
    for (const item of items) {
        const it = item;
        if (it?.kind === "String" || it?.valueKind === "StringValue") {
            // a10n:blueprint Products.A10nBlueprintEngine.Features.OperationCallGraph.targets_typed
            reportDiagnostic(ctx.program, { code: "calls-invalid-target", format: { ref: it?.value ?? "<string>" }, target });
            continue;
        }
        if (!it || it.kind !== "Operation") {
            // a10n:blueprint Products.A10nBlueprintEngine.Features.OperationCallGraph.calls_target_op
            reportDiagnostic(ctx.program, { code: "calls-invalid-target", format: { ref: it?.name ?? String(it) }, target });
            continue;
        }
        const op = it;
        const calleeKind = classifyCallee(op);
        // ui_action has a tighter restriction: only protocol_entry callees (not sdk/graph)
        if (callerKind === "ui_action") {
            const comp = findComponentAncestor(op);
            const compType = comp ? getComponentType(comp) : undefined;
            const chain = [];
            let cur = op.namespace;
            while (cur && cur !== comp && cur.name) {
                chain.unshift(cur.name);
                cur = cur.namespace;
            }
            const isProtocolEntry = compType === "Service" && chain.length > 0 &&
                (chain[0] === "Http" || chain[0] === "Mcp" || chain[0] === "Commands");
            if (!isProtocolEntry) {
                reportDiagnostic(ctx.program, {
                    code: "calls-invalid-caller-callee",
                    format: { callerKind: "ui_action", calleeKind, ref: op.name },
                    target,
                });
                continue;
            }
            resolved.push(op);
            continue;
        }
        if (!isCalleeAllowed(callerKind, calleeKind)) {
            reportDiagnostic(ctx.program, {
                code: "calls-invalid-caller-callee",
                format: { callerKind, calleeKind, ref: op.name },
                target,
            });
            continue;
        }
        resolved.push(op);
    }
    ctx.program.stateMap(CALLS_KEY).set(target, resolved);
}
// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_reads
// Note: reads_lists_all / reads_absent_pure are projected onto
// Components.A10nBlueprintService.Commands.Mapping.check — the call-graph
// completeness check compares each declared @reads edge against the real impl
// call graph via the StageOperationCoverage gates (DriftReadsUnproven).
export function $reads(ctx, target, refs) {
    const items = normalizeArray(refs);
    // suite_no_reads: @evalSuite must not carry @reads — datasets are declared via @calls
    const bearingIsEvalSuite = target.decorators.some((d) => d.definition?.name === "@evalSuite");
    if (bearingIsEvalSuite) {
        reportDiagnostic(ctx.program, {
            code: "suite-no-reads",
            format: { opName: target.name },
            target,
        });
        return;
    }
    // a10n:blueprint Products.A10nBlueprintEngine.Features.OperationCallGraph.reads_op_only
    if (!target.namespace || (!hasDecorator(target.namespace, "@component") && !hasAncestorWithDecorator(target.namespace, "@component"))) {
        reportDiagnostic(ctx.program, {
            code: "invalid-namespace-nesting",
            format: { decorator: "reads", requiredParent: "component", actualFqn: target.namespace ? getNamespaceFqn(target.namespace) : "<global>" },
            target,
        });
        return;
    }
    // Check if source op is a @dataset — datasets read storage-object models (exactly one, carrying @storageObject)
    const targetIsDataset = target.decorators.some((d) => d.definition?.name === "@dataset");
    const resolved = [];
    for (const item of items) {
        const it = item;
        if (it?.kind === "String" || it?.valueKind === "StringValue") {
            reportDiagnostic(ctx.program, { code: "reads-invalid-target", format: { ref: it?.value ?? "<string>" }, target });
            continue;
        }
        if (!it || it.kind !== "Model") {
            reportDiagnostic(ctx.program, { code: "reads-invalid-target", format: { ref: it?.name ?? String(it) }, target });
            continue;
        }
        const model = it;
        if (targetIsDataset) {
            // a10n:blueprint Products.A10nBlueprintEngine.Features.DatasetAuthoring.reads_storage
            // dataset reads a single storage-object model (must carry @storageObject)
            const storageCfg = getStorageObjectConfig(ctx.program, model);
            if (!storageCfg) {
                reportDiagnostic(ctx.program, { code: "reads-invalid-target", format: { ref: model.name }, target });
                continue;
            }
            resolved.push(model);
        }
        else {
            // Regular reads: must be a model in a Store component
            // a10n:blueprint Products.A10nBlueprintEngine.Features.OperationCallGraph.reads_target_store
            const modelComp = findModelComponent(model);
            if (!modelComp || !isStoreComponent(modelComp)) {
                reportDiagnostic(ctx.program, { code: "reads-invalid-target", format: { ref: model.name }, target });
                continue;
            }
            resolved.push(model);
        }
    }
    // a10n:blueprint Products.A10nBlueprintEngine.Features.DatasetAuthoring.exactly_one_reads
    // For @dataset ops, enforce exactly one read target
    if (targetIsDataset && resolved.length > 1) {
        for (let i = 1; i < resolved.length; i++) {
            const extra = resolved[i];
            reportDiagnostic(ctx.program, { code: "reads-invalid-target", format: { ref: extra?.name ?? String(extra) }, target });
        }
        ctx.program.stateMap(READS_KEY).set(target, [resolved[0]]);
        return;
    }
    ctx.program.stateMap(READS_KEY).set(target, resolved);
}
// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_writes
// Note: writes_lists_all / writes_absent_pure are projected onto
// Components.A10nBlueprintService.Commands.Mapping.check — the call-graph
// completeness check compares each declared @writes edge against the real impl
// call graph via the StageOperationCoverage gates (DriftWritesUnproven).
// runtime_within_deps is repurposed to require @testing on cross-component edge
// targets; that rule arm + proof land in a follow-up PR (see calls-graph.rule.ts).
export function $writes(ctx, target, models) {
    const items = normalizeArray(models);
    // suite_no_writes: @evalSuite must not carry @writes
    const bearingIsEvalSuiteW = target.decorators.some((d) => d.definition?.name === "@evalSuite");
    if (bearingIsEvalSuiteW) {
        reportDiagnostic(ctx.program, {
            code: "evalSuite-no-reads-writes",
            messageId: "writes",
            format: { opName: target.name },
            target,
        });
        return;
    }
    // a10n:blueprint Products.A10nBlueprintEngine.Features.OperationCallGraph.writes_op_only
    if (!target.namespace || (!hasDecorator(target.namespace, "@component") && !hasAncestorWithDecorator(target.namespace, "@component"))) {
        reportDiagnostic(ctx.program, {
            code: "invalid-namespace-nesting",
            format: { decorator: "writes", requiredParent: "component", actualFqn: target.namespace ? getNamespaceFqn(target.namespace) : "<global>" },
            target,
        });
        return;
    }
    const resolved = [];
    for (const item of items) {
        const it = item;
        if (it?.kind === "String" || it?.valueKind === "StringValue") {
            reportDiagnostic(ctx.program, { code: "writes-invalid-target", format: { ref: it?.value ?? "<string>" }, target });
            continue;
        }
        if (!it || it.kind !== "Model") {
            reportDiagnostic(ctx.program, { code: "writes-invalid-target", format: { ref: it?.name ?? String(it) }, target });
            continue;
        }
        const model = it;
        // a10n:blueprint Products.A10nBlueprintEngine.Features.OperationCallGraph.writes_target_store
        const modelComp = findModelComponent(model);
        if (!modelComp || !isStoreComponent(modelComp)) {
            reportDiagnostic(ctx.program, { code: "writes-invalid-target", format: { ref: model.name }, target });
            continue;
        }
        resolved.push(model);
    }
    ctx.program.stateMap(WRITES_KEY).set(target, resolved);
}
// a10n:blueprint:ignore
export function getCalls(program, target) {
    return program.stateMap(CALLS_KEY).get(target);
}
// a10n:blueprint:ignore
export function getReads(program, target) {
    return program.stateMap(READS_KEY).get(target);
}
// a10n:blueprint:ignore
export function getWrites(program, target) {
    return program.stateMap(WRITES_KEY).get(target);
}
