// @a10n/blueprint-typespec/dist/src/decorators/entity.js
import { validateTarget, validateDecoratorNotDuplicated, getNamespaceFqn } from "../validators.js";
import { reportDiagnostic } from "../lib.js";
const ENTITY_KEY = Symbol("blueprint.entity");
// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_entity
export function $entity(ctx, target) {
    if (!validateTarget(ctx, target, "@entity", "Model")) {
        return;
    }
    if (!validateDecoratorNotDuplicated(ctx, target, "@entity")) {
        return;
    }
    // Validate namespace: @entity must be inside Products.<X>.Entities.<Y>
    const ns = target.namespace;
    if (!ns) {
        reportDiagnostic(ctx.program, {
            code: "invalid-namespace-prefix",
            format: {
                decorator: "entity",
                expectedPrefix: "Products.<Name>.Entities.<Name>",
                actualFqn: "<global>",
            },
            target,
        });
        return;
    }
    const fqn = getNamespaceFqn(ns);
    const segs = fqn.split(".");
    if (segs.length !== 4 || segs[0] !== "Products" || segs[2] !== "Entities") {
        reportDiagnostic(ctx.program, {
            code: "invalid-namespace-prefix",
            format: {
                decorator: "entity",
                expectedPrefix: "Products.<Name>.Entities.<Name>",
                actualFqn: fqn,
            },
            target,
        });
        return;
    }
    ctx.program.stateMap(ENTITY_KEY).set(target, true);
}
export function isEntity(program, target) {
    return program.stateMap(ENTITY_KEY).has(target);
}
