// @a10n/blueprint-typespec/dist/src/decorators/product.js
import { validateTarget, validateDecoratorNotDuplicated, getNamespaceFqn, hasDecorator, } from "../validators.js";
import { reportDiagnostic } from "../lib.js";
const FEATURE_KEY = Symbol("blueprint.feature");
const PRODUCT_KEY = Symbol("blueprint.product");
const ASSUMPTION_KEY = Symbol("blueprint.assumption");
// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_feature
export function $feature(ctx, target, config) {
    if (!validateTarget(ctx, target, "@feature", "Namespace")) {
        return;
    }
    if (!validateDecoratorNotDuplicated(ctx, target, "@feature")) {
        return;
    }
    // Validate namespace structure: @feature must be exactly Products.<Name>.Features.<Name>
    const fqn = getNamespaceFqn(target);
    const segs = fqn.split(".");
    if (segs.length !== 4 || segs[0] !== "Products" || segs[2] !== "Features") {
        reportDiagnostic(ctx.program, {
            code: "invalid-namespace-prefix",
            format: {
                decorator: "feature",
                expectedPrefix: "Products.<Name>.Features.<Name>",
                actualFqn: fqn,
            },
            target,
        });
        return;
    }
    ctx.program.stateMap(FEATURE_KEY).set(target, config ?? true);
}
// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_product
export function $product(ctx, target, config) {
    if (!validateTarget(ctx, target, "@product", "Namespace")) {
        return;
    }
    if (!validateDecoratorNotDuplicated(ctx, target, "@product")) {
        return;
    }
    // Validate namespace structure: @product must be exactly Products.<Name>
    const fqn = getNamespaceFqn(target);
    const segs = fqn.split(".");
    if (segs.length !== 2 || segs[0] !== "Products") {
        reportDiagnostic(ctx.program, {
            code: "invalid-namespace-prefix",
            format: {
                decorator: "product",
                expectedPrefix: "Products.<Name>",
                actualFqn: fqn,
            },
            target,
        });
        return;
    }
    ctx.program.stateMap(PRODUCT_KEY).set(target, config ?? true);
}
// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_assumption
export function $assumption(ctx, target, entry) {
    // must have @product, @feature, or @dependency on the same namespace
    if (!hasDecorator(target, "@product") && !hasDecorator(target, "@feature") && !hasDecorator(target, "@dependency")) {
        const fqn = getNamespaceFqn(target);
        reportDiagnostic(ctx.program, {
            code: "invalid-assumption-context",
            format: { actualFqn: fqn },
            target,
        });
        return;
    }
    const map = ctx.program.stateMap(ASSUMPTION_KEY);
    const existing = map.get(target) ?? [];
    map.set(target, [...existing, entry]);
}
// ── Getters (for emitters) ────────────────────────────────────────────────────
// a10n:blueprint:ignore
export function isProduct(program, target) {
    return program.stateMap(PRODUCT_KEY).has(target);
}
// a10n:blueprint:ignore
export function isFeature(program, target) {
    return program.stateMap(FEATURE_KEY).has(target);
}
