// @a10n/blueprint-typespec/dist/src/decorators/sdk.js
import { getNamespaceFqn, hasDecorator, hasAncestorWithDecorator, } from "../validators.js";
import { reportDiagnostic } from "../lib.js";
// a10n:blueprint:ignore
function getComponentType(ns) {
    for (const dec of ns.decorators ?? []) {
        if (dec.definition?.name !== "@component")
            continue;
        const arg = dec.args?.[0]?.value;
        if (!arg)
            continue;
        let typeVal;
        if (arg.valueKind === "ObjectValue") {
            typeVal = arg.properties?.get("type")?.value;
        }
        else if (arg.kind === "Object") {
            typeVal = arg.properties?.type;
        }
        if (!typeVal)
            continue;
        return typeVal.valueKind === "EnumValue" ? typeVal.value?.name : typeVal.value?.name ?? String(typeVal);
    }
    return undefined;
}
// a10n:blueprint:ignore
function findComponentAncestor(op) {
    let ns = op.namespace;
    while (ns && ns.name) {
        if (hasDecorator(ns, "@component"))
            return ns;
        ns = ns.namespace;
    }
    return undefined;
}
const OPERATION_KEY = Symbol("blueprint.operation");
// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_operation
export function $operation(ctx, target, config) {
    // must be inside a @component namespace
    if (!target.namespace || !hasAncestorWithDecorator(target.namespace, "@component") && !hasDecorator(target.namespace, "@component")) {
        reportDiagnostic(ctx.program, {
            code: "invalid-namespace-nesting",
            format: {
                decorator: "operation",
                requiredParent: "component",
                actualFqn: target.namespace ? getNamespaceFqn(target.namespace) : "<global>",
            },
            target,
        });
        return;
    }
    ctx.program.stateMap(OPERATION_KEY).set(target, config);
}
// a10n:blueprint:ignore
export function getSdkOperationConfig(program, target) {
    return program.stateMap(OPERATION_KEY).get(target);
}
