// @a10n/blueprint-typespec/dist/src/decorators/tool.js
import { getNamespaceFqn, hasDecorator, hasAncestorWithDecorator, getTypeFilePath, isExemptPath, } from "../validators.js";
import { reportDiagnostic } from "../lib.js";
const TOOL_KEY = Symbol("blueprint.tool");
const ARTIFACT_KEY = Symbol("blueprint.artifact");
const INTERRUPT_OP_KEY = Symbol("blueprint.interrupt.op");
// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_tool
// a10n:blueprint Products.A10nBlueprintEngine.Features.ToolAuthoring.tool_under_agent
export function $tool(ctx, target, config) {
    // must be inside a @component namespace
    // a10n:blueprint Products.A10nBlueprintEngine.Features.ToolAuthoring.tool_under_agent
    if (!target.namespace || !hasAncestorWithDecorator(target.namespace, "@component") && !hasDecorator(target.namespace, "@component")) {
        reportDiagnostic(ctx.program, {
            code: "invalid-namespace-nesting",
            format: {
                decorator: "tool",
                requiredParent: "component",
                actualFqn: target.namespace ? getNamespaceFqn(target.namespace) : "<global>",
            },
            target,
        });
        return;
    }
    // tool-file-path: must be in components/[agent]/tools/[name].tsp
    // a10n:blueprint Products.A10nBlueprintEngine.Features.ToolAuthoring.tool_file_path
    const filePath = getTypeFilePath(target) ?? "";
    if (filePath && !isExemptPath(filePath)) {
        const norm = filePath.replace(/\\/g, "/");
        if (!/components\/[^/]+-agent\/tools\/[^/]+\.tsp$/.test(norm)) {
            reportDiagnostic(ctx.program, {
                code: "tool-file-path",
                format: { opName: target.name, actual: norm },
                target,
            });
        }
    }
    // tool-missing-sequential: sequential must be explicitly declared
    // a10n:blueprint Products.A10nBlueprintEngine.Features.ToolAuthoring.tool_sequential
    if (!config || config.sequential === undefined) {
        reportDiagnostic(ctx.program, {
            code: "tool-missing-sequential",
            format: { opName: target.name },
            target,
        });
    }
    ctx.program.stateMap(TOOL_KEY).set(target, config ?? true);
}
// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_artifact
export function $artifact(ctx, target, model) {
    // must be inside a @component namespace
    if (!target.namespace || !hasAncestorWithDecorator(target.namespace, "@component") && !hasDecorator(target.namespace, "@component")) {
        reportDiagnostic(ctx.program, {
            code: "invalid-namespace-nesting",
            format: {
                decorator: "artifact",
                requiredParent: "component",
                actualFqn: target.namespace ? getNamespaceFqn(target.namespace) : "<global>",
            },
            target,
        });
        return;
    }
    const map = ctx.program.stateMap(ARTIFACT_KEY);
    const existing = map.get(target) ?? [];
    map.set(target, [...existing, model]);
}
/** @interrupt — no-arg marker decorator on a separate op called by @tool or @node. */
// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_interrupt
// a10n:blueprint Products.A10nBlueprintEngine.Features.OpContractAuthoring.interrupt_separate_op
export function $interrupt(ctx, target) {
    // a10n:todo Products.A10nBlueprintEngine.Features.OpContractAuthoring.interrupt_separate_op — same-file co-location with the @tool/@node caller is not yet checked; TypeSpec's `extern dec interrupt(target: Operation)` signature forces @interrupt to bind to a dedicated op (cannot be applied to the caller as a decorator), which is the structural half of the contract.
    // must be inside a @component namespace
    if (!target.namespace || !hasAncestorWithDecorator(target.namespace, "@component") && !hasDecorator(target.namespace, "@component")) {
        reportDiagnostic(ctx.program, {
            code: "invalid-namespace-nesting",
            format: {
                decorator: "interrupt",
                requiredParent: "component",
                actualFqn: target.namespace ? getNamespaceFqn(target.namespace) : "<global>",
            },
            target,
        });
        return;
    }
    // interrupt-multi-arg-input: must have exactly one parameter
    // a10n:blueprint Products.A10nBlueprintEngine.Features.OpContractAuthoring.interrupt_typed_io
    const params = Array.from(target.parameters.properties.values());
    if (params.length !== 1) {
        reportDiagnostic(ctx.program, {
            code: "interrupt-multi-arg-input",
            format: { opName: target.name },
            target,
        });
    }
    // interrupt-missing-output: must not return void
    const rt = target.returnType;
    const isVoid = rt.kind === "Intrinsic" && rt.name === "void";
    if (isVoid) {
        reportDiagnostic(ctx.program, {
            code: "interrupt-missing-output",
            format: { opName: target.name },
            target,
        });
    }
    // interrupt-declares-errors: must not be a union (which includes error types)
    // a10n:blueprint Products.A10nBlueprintEngine.Features.OpContractAuthoring.interrupt_no_errors
    if (rt.kind === "Union") {
        reportDiagnostic(ctx.program, {
            code: "interrupt-declares-errors",
            format: { opName: target.name },
            target,
        });
    }
    ctx.program.stateMap(INTERRUPT_OP_KEY).set(target, true);
}
// a10n:blueprint:ignore
export function isInterruptOp(program, op) {
    return program.stateMap(INTERRUPT_OP_KEY).has(op);
}
// a10n:blueprint:ignore
export function getToolConfig(program, target) {
    return program.stateMap(TOOL_KEY).get(target);
}
