// @a10n/blueprint-typespec/dist/src/decorators/worker.js
import { getNamespaceFqn, hasDecorator, hasAncestorWithDecorator } from "../validators.js";
import { reportDiagnostic } from "../lib.js";
const PUBLISHES_KEY = Symbol("blueprint.publishes");
const SUBSCRIBES_KEY = Symbol("blueprint.subscribes");
const WORKER_KEY = Symbol("blueprint.worker");
// ── helpers ───────────────────────────────────────────────────────────────────
/** Check if a union is decorated with @events (from @typespec/events). */
function isEventsUnion(union) {
    return union.decorators?.some((d) => d.definition?.name === "@events") ?? false;
}
/** Get the parent union for a UnionVariant, if any. */
function getParentUnion(variant) {
    return variant.union;
}
/** Check if a UnionVariant's parent union has the @events decorator. */
function variantIsInEventsUnion(variant) {
    const parent = getParentUnion(variant);
    if (!parent)
        return false;
    return isEventsUnion(parent);
}
/** Walk up namespace ancestors looking for a @worker decorator. */
function findWorkerAncestor(op) {
    let ns = op.namespace;
    while (ns && ns.name) {
        if (hasDecorator(ns, "@worker"))
            return ns;
        ns = ns.namespace;
    }
    return undefined;
}
// ── @publishes ────────────────────────────────────────────────────────────────
/**
 * @publishes(variant) — declares that this operation emits the given event variant.
 * May be applied multiple times (each call adds one entry to the list).
 */
// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_publishes
// a10n:blueprint Products.A10nBlueprintEngine.Features.EventCallGraph.publishes_event_model
// a10n:blueprint Products.A10nBlueprintEngine.Features.EventCallGraph.publishes_op_only
export function $publishes(ctx, target, variant) {
    const v = variant;
    // Validate: variant must be a UnionVariant
    // a10n:blueprint Products.A10nBlueprintEngine.Features.EventCallGraph.publishes_event_model
    // a10n:blueprint Products.A10nBlueprintEngine.Features.EventCallGraph.publishes_op_only
    if (!v || v.kind !== "UnionVariant") {
        reportDiagnostic(ctx.program, {
            code: "publishes-event-model",
            format: { ref: v?.name ?? String(variant) },
            target: ctx.decoratorTarget,
        });
        return;
    }
    const uv = v;
    // Validate: the variant's parent union must have @events
    if (!variantIsInEventsUnion(uv)) {
        reportDiagnostic(ctx.program, {
            code: "publishes-event-model",
            format: { ref: uv.name != null ? String(uv.name) : "<unnamed>" },
            target: ctx.decoratorTarget,
        });
        return;
    }
    // Accumulate — each call to @publishes adds one entry
    const existing = ctx.program.stateMap(PUBLISHES_KEY).get(target) ?? [];
    existing.push(uv);
    ctx.program.stateMap(PUBLISHES_KEY).set(target, existing);
}
// a10n:blueprint:ignore
export function getPublishes(program, op) {
    return program.stateMap(PUBLISHES_KEY).get(op) ?? [];
}
// ── @subscribes ───────────────────────────────────────────────────────────────
/**
 * @subscribes — marker decorator; input arg carries the event.
 * This decorator has NO arguments — it simply marks the op as a subscriber.
 * Must be placed inside a @worker namespace.
 */
// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_subscribes
export function $subscribes(ctx, target) {
    // Validate: must be inside a @worker namespace
    const workerNs = findWorkerAncestor(target);
    if (!workerNs) {
        reportDiagnostic(ctx.program, {
            code: "subscribes-op-only",
            format: {},
            target: ctx.decoratorTarget,
        });
        return;
    }
    ctx.program.stateMap(SUBSCRIBES_KEY).set(target, true);
}
// a10n:blueprint:ignore
export function isSubscribesOp(program, op) {
    return program.stateMap(SUBSCRIBES_KEY).get(op) === true;
}
// ── @worker ───────────────────────────────────────────────────────────────────
/**
 * @worker — marks a namespace as a worker component.
 * Must be nested inside a @component namespace.
 */
// a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.decorator_worker
export function $worker(ctx, target, config) {
    // Validate: must be inside a @component namespace
    const hasComponentAncestor = hasDecorator(target, "@component") ||
        hasAncestorWithDecorator(target, "@component");
    if (!hasComponentAncestor) {
        reportDiagnostic(ctx.program, {
            code: "invalid-namespace-nesting",
            format: {
                decorator: "worker",
                requiredParent: "component",
                actualFqn: getNamespaceFqn(target),
            },
            target,
        });
        return;
    }
    // Validate WorkerConfig fields when a config is provided
    if (config !== undefined && config !== null) {
        const cfg = config;
        // TypeSpec marshals `valueof NamedModel` as a plain JS object where enum
        // fields arrive as EnumValue objects. `ObjectValue` (with a .properties Map)
        // is used when reading decorator args from the decorator list rather than
        // from the $decorator parameter directly.
        //
        // Helper: resolve a raw value (EnumValue object OR plain string) → string
        function resolveEnumName(v) {
            if (!v)
                return undefined;
            if (typeof v === "string")
                return v;
            // EnumValue: { valueKind: "EnumValue", value: EnumMember }
            if (v?.valueKind === "EnumValue")
                return v.value?.name;
            // Fallback: ObjectProperty wrapper used when reading from decorator list
            if (v?.value?.valueKind === "EnumValue")
                return v.value?.value?.name;
            return undefined;
        }
        function resolveStringVal(v) {
            if (!v)
                return undefined;
            if (typeof v === "string")
                return v;
            if (v?.valueKind === "StringValue")
                return v.value;
            if (v?.value?.valueKind === "StringValue")
                return v.value?.value;
            return undefined;
        }
        const concurrency = resolveEnumName(cfg.concurrency);
        const ack = resolveEnumName(cfg.ack);
        // Validate: OnePerKey requires routing_key
        if (concurrency === "OnePerKey") {
            const routingKey = resolveStringVal(cfg.routing_key);
            if (!routingKey) {
                reportDiagnostic(ctx.program, {
                    code: "worker-one-per-key-missing-routing-key",
                    format: { ns: getNamespaceFqn(target) },
                    target,
                });
            }
        }
    }
    ctx.program.stateMap(WORKER_KEY).set(target, config ?? {});
}
// a10n:blueprint:ignore
export function getWorkerConfig(program, ns) {
    return program.stateMap(WORKER_KEY).get(ns);
}
// a10n:blueprint:ignore
export function isWorkerNamespace(program, ns) {
    return program.stateMap(WORKER_KEY).has(ns);
}
