// @a10n/typespec-dbml/dist/src/decorators.js
// ── State keys ────────────────────────────────────────────────────────────────
const SCHEMA_KEY = Symbol("typespec.dbml.schema");
const TABLE_KEY = Symbol("typespec.dbml.table");
const PRIMARY_KEY_KEY = Symbol("typespec.dbml.primaryKey");
const UNIQUE_KEY = Symbol("typespec.dbml.unique");
const REFERENCES_KEY = Symbol("typespec.dbml.references");
const INDEX_PROP_KEY = Symbol("typespec.dbml.index.property");
const INDEX_MODEL_KEY = Symbol("typespec.dbml.index.model");
const COLUMN_OPTION_KEY = Symbol("typespec.dbml.columnOption");
const DBML_ENUM_KEY = Symbol("typespec.dbml.dbmlEnum");
const COLUMN_COMMENT_KEY = Symbol("typespec.dbml.columnComment");
// ── Decorator implementations ─────────────────────────────────────────────────
/** @schema — groups models into a named database schema. */
// a10n:blueprint Components.A10nTypespecDbmlSdk.Operations.decorator_schema
export function $schema(ctx, target, name) {
    ctx.program.stateMap(SCHEMA_KEY).set(target, name ?? true);
}
/** @table — configures table-level DBML options. */
// a10n:blueprint Components.A10nTypespecDbmlSdk.Operations.decorator_table
export function $table(ctx, target, config) {
    ctx.program.stateMap(TABLE_KEY).set(target, config);
}
/** @primaryKey — marks a property as a primary key. */
// a10n:blueprint Components.A10nTypespecDbmlSdk.Operations.decorator_primary_key
export function $primaryKey(ctx, target) {
    ctx.program.stateMap(PRIMARY_KEY_KEY).set(target, true);
}
/** @unique — marks a property as having a UNIQUE constraint. */
// a10n:blueprint Components.A10nTypespecDbmlSdk.Operations.decorator_unique
export function $unique(ctx, target) {
    ctx.program.stateMap(UNIQUE_KEY).set(target, true);
}
/** @references — marks a property as a foreign key referencing the PK of the given model. */
// a10n:blueprint Components.A10nTypespecDbmlSdk.Operations.decorator_references
export function $references(ctx, target, model) {
    ctx.program.stateMap(REFERENCES_KEY).set(target, model);
}
/** @index — single-column index (on ModelProperty) or composite index (on Model). */
// a10n:blueprint Components.A10nTypespecDbmlSdk.Operations.decorator_index
export function $index(ctx, target, config) {
    if (target.kind === "ModelProperty") {
        ctx.program.stateMap(INDEX_PROP_KEY).set(target, config ?? {});
    }
    else {
        // Model: accumulate multiple composite index declarations
        const existing = ctx.program.stateMap(INDEX_MODEL_KEY).get(target) ?? [];
        existing.push(config ?? {});
        ctx.program.stateMap(INDEX_MODEL_KEY).set(target, existing);
    }
}
/** @columnOption — generic per-column metadata (e.g. key="default", value="`now()`"). */
// a10n:blueprint Components.A10nTypespecDbmlSdk.Operations.decorator_column_option
export function $columnOption(ctx, target, key, value) {
    const existing = ctx.program.stateMap(COLUMN_OPTION_KEY).get(target) ?? {};
    existing[key] = value;
    ctx.program.stateMap(COLUMN_OPTION_KEY).set(target, existing);
}
/** @columnComment — emits a comment line before this column. */
// a10n:blueprint Components.A10nTypespecDbmlSdk.Operations.decorator_column_comment
export function $columnComment(ctx, target, comment) {
    ctx.program.stateMap(COLUMN_COMMENT_KEY).set(target, comment);
}
/** @dbmlEnum — marks a TypeSpec enum for DBML Enum block emission. */
// a10n:blueprint Components.A10nTypespecDbmlSdk.Operations.decorator_dbml_enum
export function $dbmlEnum(ctx, target) {
    ctx.program.stateMap(DBML_ENUM_KEY).set(target, true);
}
// ── Getter functions ──────────────────────────────────────────────────────────
// a10n:blueprint:ignore
export function getSchema(program, target) {
    return program.stateMap(SCHEMA_KEY).get(target);
}
// a10n:blueprint:ignore
export function getTable(program, target) {
    return program.stateMap(TABLE_KEY).get(target);
}
// a10n:blueprint:ignore
export function getPrimaryKey(program, target) {
    return program.stateMap(PRIMARY_KEY_KEY).get(target) === true;
}
// a10n:blueprint:ignore
export function getUnique(program, target) {
    return program.stateMap(UNIQUE_KEY).get(target) === true;
}
// a10n:blueprint:ignore
export function getReferences(program, target) {
    return program.stateMap(REFERENCES_KEY).get(target);
}
// a10n:blueprint:ignore
export function getIndex(program, target) {
    return program.stateMap(INDEX_PROP_KEY).get(target);
}
// a10n:blueprint:ignore
export function getCompositeIndexes(program, target) {
    return program.stateMap(INDEX_MODEL_KEY).get(target) ?? [];
}
// a10n:blueprint:ignore
export function getColumnOptions(program, target) {
    return program.stateMap(COLUMN_OPTION_KEY).get(target) ?? {};
}
// a10n:blueprint:ignore
export function isDbmlEnum(program, target) {
    return program.stateMap(DBML_ENUM_KEY).get(target) === true;
}
// a10n:blueprint:ignore
export function getColumnComment(program, target) {
    return program.stateMap(COLUMN_COMMENT_KEY).get(target);
}
