// @a10n/typespec-dbml/src/decorators.ts
import { DecoratorContext, Enum, Model, ModelProperty, Namespace } from "@typespec/compiler";

// ── State keys ────────────────────────────────────────────────────────────────

const SCHEMA_KEY = Symbol("typespec.dbml.schema");
const TABLE_KEY = Symbol("typespec.dbml.table");
const PRIMARY_KEY_KEY = Symbol("typespec.dbml.primaryKey");
const UNIQUE_KEY = Symbol("typespec.dbml.unique");
const REFERENCES_KEY = Symbol("typespec.dbml.references");
const INDEX_PROP_KEY = Symbol("typespec.dbml.index.property");
const INDEX_MODEL_KEY = Symbol("typespec.dbml.index.model");
const COLUMN_OPTION_KEY = Symbol("typespec.dbml.columnOption");
const DBML_ENUM_KEY = Symbol("typespec.dbml.dbmlEnum");
const COLUMN_COMMENT_KEY = Symbol("typespec.dbml.columnComment");

// ── Decorator implementations ─────────────────────────────────────────────────

/** @schema — groups models into a named database schema. */
// a10n:blueprint Components.A10nTypespecDbmlSdk.Operations.decorator_schema
export function $schema(
  ctx: DecoratorContext,
  target: Namespace | Model,
  name?: string,
): void {
  ctx.program.stateMap(SCHEMA_KEY).set(target, name ?? true);
}

/** @table — configures table-level DBML options. */
// a10n:blueprint Components.A10nTypespecDbmlSdk.Operations.decorator_table
export function $table(
  ctx: DecoratorContext,
  target: Model,
  config: { name?: string; note?: string },
): void {
  ctx.program.stateMap(TABLE_KEY).set(target, config);
}

/** @primaryKey — marks a property as a primary key. */
// a10n:blueprint Components.A10nTypespecDbmlSdk.Operations.decorator_primary_key
export function $primaryKey(
  ctx: DecoratorContext,
  target: ModelProperty,
): void {
  ctx.program.stateMap(PRIMARY_KEY_KEY).set(target, true);
}

/** @unique — marks a property as having a UNIQUE constraint. */
// a10n:blueprint Components.A10nTypespecDbmlSdk.Operations.decorator_unique
export function $unique(
  ctx: DecoratorContext,
  target: ModelProperty,
): void {
  ctx.program.stateMap(UNIQUE_KEY).set(target, true);
}

/** @references — marks a property as a foreign key referencing the PK of the given model. */
// a10n:blueprint Components.A10nTypespecDbmlSdk.Operations.decorator_references
export function $references(
  ctx: DecoratorContext,
  target: ModelProperty,
  model: Model,
): void {
  ctx.program.stateMap(REFERENCES_KEY).set(target, model);
}

/** @index — single-column index (on ModelProperty) or composite index (on Model). */
// a10n:blueprint Components.A10nTypespecDbmlSdk.Operations.decorator_index
export function $index(
  ctx: DecoratorContext,
  target: ModelProperty | Model,
  config?: { columns?: string[]; name?: string; unique?: boolean; note?: string },
): void {
  if (target.kind === "ModelProperty") {
    ctx.program.stateMap(INDEX_PROP_KEY).set(target, config ?? {});
  } else {
    // Model: accumulate multiple composite index declarations
    const existing = (ctx.program.stateMap(INDEX_MODEL_KEY).get(target) as any[] | undefined) ?? [];
    existing.push(config ?? {});
    ctx.program.stateMap(INDEX_MODEL_KEY).set(target, existing);
  }
}

/** @columnOption — generic per-column metadata (e.g. key="default", value="`now()`"). */
// a10n:blueprint Components.A10nTypespecDbmlSdk.Operations.decorator_column_option
export function $columnOption(
  ctx: DecoratorContext,
  target: ModelProperty,
  key: string,
  value: string,
): void {
  const existing = (ctx.program.stateMap(COLUMN_OPTION_KEY).get(target) as Record<string, string> | undefined) ?? {};
  existing[key] = value;
  ctx.program.stateMap(COLUMN_OPTION_KEY).set(target, existing);
}

/** @columnComment — emits a comment line before this column. */
// a10n:blueprint Components.A10nTypespecDbmlSdk.Operations.decorator_column_comment
export function $columnComment(
  ctx: DecoratorContext,
  target: ModelProperty,
  comment: string,
): void {
  ctx.program.stateMap(COLUMN_COMMENT_KEY).set(target, comment);
}

/** @dbmlEnum — marks a TypeSpec enum for DBML Enum block emission. */
// a10n:blueprint Components.A10nTypespecDbmlSdk.Operations.decorator_dbml_enum
export function $dbmlEnum(
  ctx: DecoratorContext,
  target: Enum,
): void {
  ctx.program.stateMap(DBML_ENUM_KEY).set(target, true);
}

// ── Getter functions ──────────────────────────────────────────────────────────

// a10n:blueprint:ignore
export function getSchema(
  program: { stateMap(key: symbol): Map<unknown, unknown> },
  target: Namespace | Model,
): string | true | undefined {
  return program.stateMap(SCHEMA_KEY).get(target) as string | true | undefined;
}

// a10n:blueprint:ignore
export function getTable(
  program: { stateMap(key: symbol): Map<unknown, unknown> },
  target: Model,
): { name?: string; note?: string } | undefined {
  return program.stateMap(TABLE_KEY).get(target) as { name?: string; note?: string } | undefined;
}

// a10n:blueprint:ignore
export function getPrimaryKey(
  program: { stateMap(key: symbol): Map<unknown, unknown> },
  target: ModelProperty,
): boolean {
  return program.stateMap(PRIMARY_KEY_KEY).get(target) === true;
}

// a10n:blueprint:ignore
export function getUnique(
  program: { stateMap(key: symbol): Map<unknown, unknown> },
  target: ModelProperty,
): boolean {
  return program.stateMap(UNIQUE_KEY).get(target) === true;
}

// a10n:blueprint:ignore
export function getReferences(
  program: { stateMap(key: symbol): Map<unknown, unknown> },
  target: ModelProperty,
): Model | undefined {
  return program.stateMap(REFERENCES_KEY).get(target) as Model | undefined;
}

// a10n:blueprint:ignore
export function getIndex(
  program: { stateMap(key: symbol): Map<unknown, unknown> },
  target: ModelProperty,
): { name?: string; note?: string } | undefined {
  return program.stateMap(INDEX_PROP_KEY).get(target) as { name?: string; note?: string } | undefined;
}

// a10n:blueprint:ignore
export function getCompositeIndexes(
  program: { stateMap(key: symbol): Map<unknown, unknown> },
  target: Model,
): Array<{ columns?: string[]; name?: string; unique?: boolean; note?: string }> {
  return (program.stateMap(INDEX_MODEL_KEY).get(target) as any[] | undefined) ?? [];
}

// a10n:blueprint:ignore
export function getColumnOptions(
  program: { stateMap(key: symbol): Map<unknown, unknown> },
  target: ModelProperty,
): Record<string, string> {
  return (program.stateMap(COLUMN_OPTION_KEY).get(target) as Record<string, string> | undefined) ?? {};
}

// a10n:blueprint:ignore
export function isDbmlEnum(
  program: { stateMap(key: symbol): Map<unknown, unknown> },
  target: Enum,
): boolean {
  return program.stateMap(DBML_ENUM_KEY).get(target) === true;
}

// a10n:blueprint:ignore
export function getColumnComment(
  program: { stateMap(key: symbol): Map<unknown, unknown> },
  target: ModelProperty,
): string | undefined {
  return program.stateMap(COLUMN_COMMENT_KEY).get(target) as string | undefined;
}
