// @a10n/blueprint-typespec/src/rules/component-block-form.rule.ts
import {
  createRule,
  getLocationContext,
  getSourceLocation,
  paramMessage,
  Program,
} from "@typespec/compiler";

// SyntaxKind.NamespaceStatement = 8
const NS_STATEMENT_KIND = 8;

// Build FQN by walking parent NamespaceStatementNodes in the AST.
function fqnFromAst(node: any): string | undefined {
  const parts: string[] = [];
  // Dotted leaf id is just the last segment; the wrapping segments are intermediate parents.
  const leafId: string | undefined = node.id?.sv;
  if (leafId) parts.unshift(leafId);
  let cur: any = node.parent;
  while (cur && cur.kind === NS_STATEMENT_KIND) {
    const idText: string | undefined = cur.id?.sv;
    if (idText) parts.unshift(idText);
    cur = cur.parent;
  }
  return parts.join(".") || undefined;
}

// Returns true when this AST node is a leaf (not a dotted intermediate segment).
// Intermediate segments have statements as a single NamespaceStatementNode.
function isLeafNamespaceNode(node: any): boolean {
  if (!node || node.kind !== NS_STATEMENT_KIND) return false;
  const stmts = node.statements;
  if (stmts && !Array.isArray(stmts) && stmts.kind === NS_STATEMENT_KIND) return false;
  return true;
}

// Returns true when the namespace AST node uses semicolon form: `namespace X;`
// Block form `namespace X { ... }` contains `{` after the `namespace` keyword.
function isSemicolonFormNode(node: any): boolean {
  if (!node || node.kind !== NS_STATEMENT_KIND) return false;
  // Block form: parser sets `statements` to an array (possibly empty).
  // Semicolon form: `statements` is undefined.
  // Dotted intermediate segment (`A` in `namespace A.B { ... }`):
  //   `statements` is the inner NamespaceStatementNode — not an array.
  //   These are filtered out by `isLeafNamespaceNode` before reaching here.
  // Same detection used by the single-block-form rule.
  return !Array.isArray(node.statements);
}

export const componentBlockFormRule = createRule({
  name: "component-block-form",
  severity: "warning",
  description:
    "Namespaces carrying any decorator must use block form `namespace X { }` so the decorators are correctly indexed. TypeSpec silently drops decorators on semicolon-form augmentations when a block-form declaration of the same FQN exists elsewhere.",
  messages: {
    default: paramMessage`Namespace "${"fqn"}" has decorators but uses semicolon form \`namespace X;\`. Use block form \`namespace X { }\` instead so all decorators are indexed.`,
  },
  // a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.rule_component_block_form
  create(context) {
    return {
      // a10n:index:call packages.blueprint-typespec.src.rules.component-block-form.rule.root
      root(program: Program) {
        const seen = new WeakSet<object>();
        for (const [, script] of program.sourceFiles) {
          // Only consider user code. Library namespaces may legitimately
          // mix declaration forms.
          const locCtx = getLocationContext(program, script);
          if (locCtx.type !== "project") continue;

          const visit = (node: any) => {
            if (!node) return;
            if (node.kind === NS_STATEMENT_KIND) {
              if (isLeafNamespaceNode(node) && !seen.has(node)) {
                seen.add(node);
                const decs = node.decorators ?? [];
                if (decs.length > 0 && isSemicolonFormNode(node)) {
                  const fqn = fqnFromAst(node);
                  if (fqn) {
                    context.reportDiagnostic({
                      messageId: "default",
                      format: { fqn },
                      target: node,
                    });
                  }
                }
              }
              // Descend into block-form children.
              if (Array.isArray(node.statements)) {
                for (const child of node.statements) visit(child);
              } else if (node.statements && node.statements.kind === NS_STATEMENT_KIND) {
                visit(node.statements);
              }
              return;
            }
            // Top-level script: walk its statements.
            if (Array.isArray(node.statements)) {
              for (const child of node.statements) visit(child);
            }
          };

          visit(script);
        }
      },
    };
  },
});
