// @a10n/blueprint-typespec/src/rules/dataset-slices-distinct.rule.ts
import { createRule, getSourceLocation, Operation, paramMessage } from "@typespec/compiler";

function hasDecorator(node: { decorators: readonly { definition?: { name: string } }[] }, name: string): boolean {
  return node.decorators.some((d) => d.definition?.name === name);
}

/** Collect distinct invariant FQNs from all @projects decorators on an op. */
function projectsInvariantTargets(op: Operation): Set<string> {
  const fqns = new Set<string>();
  for (const dec of op.decorators) {
    if (dec.definition?.name !== "@projects") continue;
    for (const arg of dec.args) {
      const v = arg.value as any;
      if (v?.kind === "Operation" && v.decorators?.some((d: any) => d.definition?.name === "@invariant")) {
        const invNs: string[] = [];
        let cur = v.namespace;
        while (cur && cur.name) { invNs.unshift(cur.name); cur = cur.namespace; }
        const fqn = invNs.length ? `${invNs.join(".")}.${v.name}` : v.name;
        fqns.add(fqn);
      }
    }
  }
  return fqns;
}

/** Return the normalized source file path for an operation's declaration node. */
function getOpFilePath(op: Operation): string | undefined {
  if (!op.node) return undefined;
  const loc = getSourceLocation(op.node);
  if (!loc) return undefined;
  return loc.file.path.replace(/\\/g, "/");
}

/** Serialize a sorted set of FQNs to a canonical string for equality comparison. */
function setKey(fqns: Set<string>): string {
  return [...fqns].sort().join("|");
}

export const datasetSlicesDistinctRule = createRule({
  name: "dataset-slices-distinct",
  severity: "warning",
  description:
    "No two @dataset ops in the same file may declare an identical set of @projects invariant targets. Split into distinct slices or merge the datasets.",
  messages: {
    default: paramMessage`@dataset op "${"second"}" has the same @projects invariant set as "${"first"}" in the same file — datasets in the same file must target distinct invariant slices.`,
  },
  // a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.rule_dataset_slices_distinct
  // a10n:blueprint Products.A10nBlueprintEngine.Features.DatasetAuthoring.dataset_slices_distinct
  create(context) {
    // file path → Map<canonical-set-key, first op name>
    const seenByFile = new Map<string, Map<string, string>>();

    return {
      // a10n:index:call packages.blueprint-typespec.src.rules.dataset-slices-distinct.rule.operation
      operation(op: Operation) {
        if (!hasDecorator(op, "@dataset")) return;
        if (!hasDecorator(op, "@projects")) return;

        const filePath = getOpFilePath(op);
        if (!filePath) return;

        const targets = projectsInvariantTargets(op);
        if (targets.size === 0) return;

        const key = setKey(targets);

        if (!seenByFile.has(filePath)) {
          seenByFile.set(filePath, new Map());
        }
        const fileMap = seenByFile.get(filePath)!;

        const existing = fileMap.get(key);
        if (existing !== undefined) {
          // a10n:blueprint Products.A10nBlueprintEngine.Features.DatasetAuthoring.dataset_slices_distinct
          context.reportDiagnostic({
            messageId: "default",
            format: { second: op.name, first: existing },
            target: op,
          });
        } else {
          fileMap.set(key, op.name);
        }
      },
    };
  },
});
