// @a10n/blueprint-typespec/src/rules/event-call-graph.rule.ts
import {
  createRule,
  Namespace,
  Operation,
  Program,
  Union,
  UnionVariant,
  paramMessage,
} from "@typespec/compiler";

// ── helpers ───────────────────────────────────────────────────────────────────

function hasDecorator(node: any, name: string): boolean {
  return (node?.decorators ?? []).some((d: any) => d.definition?.name === name);
}

/** Check if a union has the @events decorator (from @typespec/events). */
function isEventsUnion(union: Union): boolean {
  return (union as any).decorators?.some(
    (d: any) => d.definition?.name === "@events",
  ) ?? false;
}

/** Check if a UnionVariant's parent union has @events. */
function variantIsInEventsUnion(v: UnionVariant): boolean {
  const parent = v.union;
  if (!parent) return false;
  return isEventsUnion(parent);
}

/** Get the variant name as a string. */
function variantName(v: UnionVariant): string {
  return v.name != null ? String(v.name) : "<unnamed>";
}

// ── AST-based decorator reads (avoid stateMap Symbol mismatch in vitest) ─────
//
// When vitest loads rule source (TS), the decorator impls run from dist/ (JS).
// The Symbol keys differ, so stateMap lookups always return undefined in tests.
// We walk the decorator AST directly instead.

/** Normalize a TypeSpec ArrayValue/Tuple to a plain JS array. */
// a10n:blueprint:ignore
function normalizeItems(v: any): any[] {
  if (Array.isArray(v)) return v;
  if (v?.valueKind === "ArrayValue") return v.values ?? [];
  if (v?.kind === "Tuple") return v.values ?? [];
  return [v];
}

/**
 * Read all @publishes variant targets from an op's decorator AST.
 * Returns an array of UnionVariant objects.
 */
function getPublishesFromAst(op: Operation): UnionVariant[] {
  const results: UnionVariant[] = [];
  for (const dec of (op as any).decorators ?? []) {
    if (dec.definition?.name !== "@publishes") continue;
    for (const arg of dec.args ?? []) {
      const v = arg.value as any;
      if (v?.kind === "UnionVariant") {
        results.push(v as UnionVariant);
      }
    }
  }
  return results;
}

/** Check if an op has @subscribes in its decorator list (AST walk). */
function isSubscribesOpFromAst(op: Operation): boolean {
  return hasDecorator(op, "@subscribes");
}

/** Check if a namespace has @worker in its decorator list (AST walk). */
// a10n:blueprint:ignore
function isWorkerNamespaceFromAst(ns: Namespace): boolean {
  return hasDecorator(ns, "@worker");
}

/** Collect all names from an op's parameter types (short names only). */
function collectParamTypeNames(op: Operation): Set<string> {
  const names = new Set<string>();
  for (const [, param] of op.parameters.properties) {
    const t = param.type as any;
    if (!t) continue;
    if (t.kind === "UnionVariant" && t.name != null) {
      names.add(String(t.name));
    } else if (t.kind === "Union") {
      for (const [, v] of (t as Union).variants) {
        if (v.name != null) names.add(String(v.name));
      }
    } else if (t.kind === "Model" && t.name) {
      names.add(t.name as string);
    }
  }
  return names;
}

/** Get the @doc string from a type node's decorator list. */
function getDocString(op: Operation): string | undefined {
  for (const dec of (op as any).decorators ?? []) {
    if (dec.definition?.name !== "@doc") continue;
    const arg = dec.args?.[0]?.value;
    if (arg?.valueKind === "StringValue") return arg.value as string;
    if (arg?.kind === "String") return arg.value as string;
    if (typeof arg === "string") return arg;
  }
  return undefined;
}

/** Extract {@TypeName} references from a doc string. */
function extractDocRefs(doc: string): string[] {
  const refs: string[] = [];
  const regex = /\{@([\w.]+)\}/g;
  let match: RegExpExecArray | null;
  while ((match = regex.exec(doc)) !== null) {
    refs.push(match[1]);
  }
  return refs;
}

/** Get the short name (last segment) of a dotted ref. */
function shortName(ref: string): string {
  const parts = ref.split(".");
  return parts[parts.length - 1] ?? ref;
}

/** Collect all operations across the entire program. */
function collectAllOps(program: Program): Operation[] {
  const ops: Operation[] = [];
  function walk(ns: Namespace): void {
    for (const [, op] of ns.operations) ops.push(op);
    for (const [, child] of ns.namespaces) walk(child);
  }
  walk(program.getGlobalNamespaceType());
  return ops;
}

/** Build the set of event variant names that worker @subscribes ops listen to (via input param types). */
function buildSubscribedEventNames(program: Program): Set<string> {
  const names = new Set<string>();
  const allOps = collectAllOps(program);
  for (const op of allOps) {
    if (!isSubscribesOpFromAst(op)) continue;
    for (const name of collectParamTypeNames(op)) {
      names.add(name);
    }
  }
  return names;
}

// ── Rule ───────────────────────────────────────────────────────────────────────

export const eventCallGraphRule = createRule({
  name: "event-call-graph",
  severity: "warning",
  description:
    "Enforces @publishes/@subscribes completeness, target-kind constraints, and event graph connectivity.",
  messages: {
    // publishes-event-model: @publishes target is not a variant of @events union
    publishesEventModel: paramMessage`@publishes target "${"ref"}" is not a variant of an @events-decorated union — only event payload variants are valid @publishes targets.`,
    // publishes-op-only: @publishes found on non-operation
    publishesOpOnly: paramMessage`@publishes can only appear on operations, but was found on a ${"kind"} — move it to an op declaration.`,
    // publishes-incomplete: op doc refs an event not in @publishes
    publishesListsAll: paramMessage`op "${"opName"}" doc references event "${"ref"}" which is not declared in @publishes — add it to @publishes.`,
    // publishes-absent-pure: op has @publishes but doc has no event refs
    publishesAbsentPure: paramMessage`op "${"opName"}" carries @publishes but doc references no event emissions — remove @publishes or declare the event emission in @doc.`,
    // subscribes-incomplete: worker op doc refs an event not in input type
    subscribesListsAll: paramMessage`worker op "${"opName"}" doc references event "${"ref"}" not matched by its input type — update the input parameter to subscribe to all listed events.`,
    // publishes-no-subscriber: published event has no @subscribes worker op
    publishesHasSubscriber: paramMessage`published event "${"ref"}" has no @subscribes worker op consuming it — add a worker that subscribes to it.`,
    // worker-input-variant: worker op input type is not a variant of @events union
    workerInputVariant: paramMessage`worker op "${"opName"}" input type "${"ref"}" is not a variant of an @events-decorated union — worker inputs must be event payload models.`,
  },
  // a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.rule_event_call_graph
  // a10n:blueprint Products.A10nBlueprintEngine.Features.EventCallGraph.publishes_absent_pure
  // a10n:blueprint Products.A10nBlueprintEngine.Features.EventCallGraph.publishes_has_subscriber
  // a10n:blueprint Products.A10nBlueprintEngine.Features.EventCallGraph.publishes_lists_all
  // a10n:blueprint Products.A10nBlueprintEngine.Features.EventCallGraph.subscribes_lists_all
  // a10n:blueprint Products.A10nBlueprintEngine.Features.EventCallGraph.worker_input_variant
  create(context) {
    return {
      // a10n:index:call packages.blueprint-typespec.src.rules.event-call-graph.rule.operation
      // a10n:index:call packages.blueprint-typespec.src.rules.event-call-graph.rule.root
      operation(op: Operation) {
        const publishedVariants = getPublishesFromAst(op);

        // ── publishes-incomplete & publishes-absent-pure ─────────────────────
        // Cross-check @doc inline refs against @publishes entries.
        //
        // Note on publishes-op-only: the extern decorator declaration
        //   `extern dec publishes(target: Operation, variant: UnionVariant)`
        // means the TypeSpec compiler itself rejects @publishes on any non-Operation
        // before this rule runs. The publishesOpOnly diagnostic therefore cannot be
        // emitted here — it is enforced at the decorator-declaration level and the
        // program never reaches rule evaluation with a malformed target.
        const doc = getDocString(op);
        const hasPublishes = publishedVariants.length > 0;

        if (hasPublishes) {
          if (!doc) {
            // publishes-absent-pure: has @publishes but no @doc at all — we cannot
            // verify what events are emitted at runtime, which is a violation of the
            // invariant that every emission must be documented.
            // a10n:blueprint Products.A10nBlueprintEngine.Features.EventCallGraph.publishes_absent_pure
            context.reportDiagnostic({
              messageId: "publishesAbsentPure",
              format: { opName: op.name },
              target: op,
            });
          } else {
            const docRefs = extractDocRefs(doc);
            // Filter doc refs that look like event model references (contain a dot or start uppercase)
            const eventDocRefs = docRefs.filter((ref) => {
              if (ref.includes(".")) return true;
              const last = shortName(ref);
              return last.length > 0 && /[A-Z]/.test(last[0]);
            });

            if (eventDocRefs.length === 0) {
              // publishes-absent-pure: has @publishes but @doc has no event refs at all
              // a10n:blueprint Products.A10nBlueprintEngine.Features.EventCallGraph.publishes_absent_pure
              context.reportDiagnostic({
                messageId: "publishesAbsentPure",
                format: { opName: op.name },
                target: op,
              });
            } else {
              // publishes-incomplete: doc refs events not in @publishes
              // a10n:blueprint Products.A10nBlueprintEngine.Features.EventCallGraph.publishes_lists_all
              const publishedNames = new Set(publishedVariants.map(variantName));
              for (const ref of eventDocRefs) {
                const refShort = shortName(ref);
                if (!publishedNames.has(refShort) && !publishedNames.has(ref)) {
                  context.reportDiagnostic({
                    messageId: "publishesListsAll",
                    format: { opName: op.name, ref },
                    target: op,
                  });
                }
              }
            }
          }
        }

        // ── subscribes-incomplete & worker-input-variant ─────────────────────
        if (isSubscribesOpFromAst(op)) {
          const paramTypeNames = collectParamTypeNames(op);
          const doc2 = getDocString(op);
          if (doc2) {
            const docRefs = extractDocRefs(doc2);
            const eventDocRefs = docRefs.filter((ref) => {
              if (ref.includes(".")) return true;
              const last = shortName(ref);
              return last.length > 0 && /[A-Z]/.test(last[0]);
            });
            // a10n:blueprint Products.A10nBlueprintEngine.Features.EventCallGraph.subscribes_lists_all
            for (const ref of eventDocRefs) {
              const refShort = shortName(ref);
              if (!paramTypeNames.has(refShort) && !paramTypeNames.has(ref)) {
                context.reportDiagnostic({
                  messageId: "subscribesListsAll",
                  format: { opName: op.name, ref },
                  target: op,
                });
              }
            }
          }

          // ── worker-input-variant ────────────────────────────────────────────
          // Every input param type of a @subscribes op must be a variant of an @events union
          // a10n:blueprint Products.A10nBlueprintEngine.Features.EventCallGraph.worker_input_variant
          for (const [, param] of op.parameters.properties) {
            const t = param.type as any;
            if (!t) continue;
            if (t.kind === "UnionVariant") {
              if (!variantIsInEventsUnion(t as UnionVariant)) {
                context.reportDiagnostic({
                  messageId: "workerInputVariant",
                  format: { opName: op.name, ref: t.name != null ? String(t.name) : "<unnamed>" },
                  target: op,
                });
              }
            } else if (t.kind === "Union") {
              for (const [, v] of (t as Union).variants) {
                if (!variantIsInEventsUnion(v)) {
                  context.reportDiagnostic({
                    messageId: "workerInputVariant",
                    format: { opName: op.name, ref: v.name != null ? String(v.name) : "<unnamed>" },
                    target: op,
                  });
                }
              }
            } else if (t.kind === "Model") {
              // A standalone model (not from an @events union) is never valid
              context.reportDiagnostic({
                messageId: "workerInputVariant",
                format: { opName: op.name, ref: (t.name as string) ?? "<unnamed>" },
                target: op,
              });
            }
          }
        }
      },

      // ── publishes-no-subscriber ──────────────────────────────────────────────
      // The `root` hook fires once for the entire program after all types are visited.
      root(program: Program) {
        const subscribedNames = buildSubscribedEventNames(program);
        const allOps = collectAllOps(program);

        for (const op of allOps) {
          const published = getPublishesFromAst(op);
          for (const variant of published) {
            // Only check variants that are actually from @events unions
            if (!variantIsInEventsUnion(variant)) continue;
            const name = variantName(variant);
            if (!subscribedNames.has(name)) {
              // a10n:blueprint Products.A10nBlueprintEngine.Features.EventCallGraph.publishes_has_subscriber
              context.reportDiagnostic({
                messageId: "publishesHasSubscriber",
                format: { ref: name },
                target: op,
              });
            }
          }
        }
      },
    };
  },
});
