// @a10n/blueprint-typespec/src/rules/featureconfig-field-cap.rule.ts
import { createRule, Model, paramMessage } from "@typespec/compiler";

export const featureconfigFieldCapRule = createRule({
  name: "featureconfig-field-cap",
  severity: "warning",
  description:
    "A @featureConfig model must not declare more than 10 fields. Split the feature config or the feature itself.",
  messages: {
    default: paramMessage`@featureConfig model "${"modelName"}" has ${"count"} fields; the maximum is 10. Split the feature config into two models or split the feature.`,
  },
  // a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.rule_featureconfig_field_cap
  create(context) {
    return {
      // a10n:index:call packages.blueprint-typespec.src.rules.featureconfig-field-cap.rule.model
      model(m: Model) {
        const hasFeatureConfig = m.decorators.some(
          (d) => d.definition?.name === "@featureConfig",
        );
        if (!hasFeatureConfig) return;
        const count = m.properties.size;
        if (count > 10) {
          context.reportDiagnostic({
            messageId: "default",
            format: { modelName: m.name, count: String(count) },
            target: m,
          });
        }
      },
    };
  },
});
