// @a10n/blueprint-typespec/src/rules/graph-authoring.rule.ts
import { createRule, Model, ModelProperty, Namespace, Operation, Program, Type } from "@typespec/compiler";

function hasDecorator(node: any, name: string): boolean {
  return (node?.decorators ?? []).some((d: any) => d.definition?.name === name);
}

function nsToFqn(ns: Namespace): string {
  const parts: string[] = [];
  let cur: Namespace | undefined = ns;
  while (cur && cur.name) { parts.unshift(cur.name); cur = cur.namespace; }
  return parts.join(".");
}

function opFqn(op: Operation): string {
  return op.namespace ? `${nsToFqn(op.namespace)}.${op.name}` : op.name;
}

function getCallsFromAst(op: Operation): Operation[] {
  const result: Operation[] = [];
  for (const dec of op.decorators) {
    if (dec.definition?.name !== "@calls") continue;
    for (const arg of dec.args ?? []) {
      const v = arg.value as any;
      const items: any[] = v?.valueKind === "ArrayValue" ? v.values : (v?.kind === "Tuple" ? v.values : [v]);
      for (const item of items) {
        if (item?.kind === "Operation") result.push(item as Operation);
      }
    }
  }
  return result;
}

// ── Command<TUpdate, TGoTo> type-level analysis ───────────────────────────────
//
// TypeSpec represents `Command<PickProperties<MyState, "status">, next_node>` as a Model
// whose templateMapper has the two type args. We walk the return type to find:
//   1. The go_to arg (second arg of Command) — to check routing target
//   2. The update arg (first arg of Command) — to check state targeting
//   3. Whether the node returns Command[] (fanout) or Command (single)
//
// Command<TUpdate, TGoTo> is a template instantiation; its template is named "Command".
// We detect it by the template name.

interface CommandShape {
  updateType: Type | undefined;
  goToType: Type | undefined;
  isFanout: boolean;
}

function extractCommandShapes(returnType: Type): CommandShape[] {
  const shapes: CommandShape[] = [];

  function walk(t: Type): void {
    if (!t) return;
    // Array type: Command[] → fanout
    if (t.kind === "Model" && (t as any).indexer) {
      const valueType = (t as any).indexer?.value;
      if (valueType) walkAsCommand(valueType, true);
      return;
    }
    // Union: Command<A,X> | Command<B,Y>
    if (t.kind === "Union") {
      for (const variant of (t as any).variants?.values() ?? []) {
        walk(variant.type);
      }
      return;
    }
    walkAsCommand(t, false);
  }

  function walkAsCommand(t: Type, isFanout: boolean): void {
    if (t.kind !== "Model") return;
    const m = t as Model;
    // Check if this is a Command instantiation by looking for template name
    const templateName = (m as any).templateMapper?.target?.name ?? (m as any).name;
    if (templateName !== "Command") return;
    const args: Type[] = (m as any).templateMapper?.args ?? [];
    shapes.push({
      updateType: args[0],
      goToType: args[1],
      isFanout,
    });
  }

  walk(returnType);
  return shapes;
}

// Determine if a type is a PickProperties instantiation of a given model
function isPickPropertiesOf(t: Type | undefined, stateModel: Model): boolean {
  if (!t || t.kind !== "Model") return false;
  const m = t as Model;
  const templateName = (m as any).templateMapper?.target?.name ?? (m as any).name;
  if (templateName !== "PickProperties") return false;
  const args: Type[] = (m as any).templateMapper?.args ?? [];
  return args[0] === stateModel;
}

// Find the state model for a @graph namespace (first non-template model declared directly in ns)
function findStateModel(ns: Namespace): Model | undefined {
  for (const [, m] of ns.models) {
    if (m.name && !(m as any).templateMapper) return m;
  }
  return undefined;
}

// Check whether a type refers to a store model (inside a @component(Store) ns)
function isStoreModel(t: Type | undefined): boolean {
  if (!t || t.kind !== "Model") return false;
  const m = t as Model;
  if (!m.namespace) return false;
  let ns: Namespace | undefined = m.namespace;
  while (ns && ns.name) {
    if (hasDecorator(ns, "@component")) {
      // Check if it's a Store component
      for (const dec of (ns as any).decorators ?? []) {
        if (dec.definition?.name !== "@component") continue;
        const arg = dec.args?.[0]?.value;
        let typeVal: any;
        if (arg?.valueKind === "ObjectValue") {
          typeVal = (arg.properties as Map<string, any>)?.get("type")?.value;
        } else if (arg?.kind === "Object") {
          typeVal = (arg as any).properties?.type;
        }
        const typeName = typeVal?.valueKind === "EnumValue" ? typeVal.value?.name : typeVal?.value?.name;
        if (typeName === "Store") return true;
      }
      return false;
    }
    ns = ns.namespace;
  }
  return false;
}

// Check if a namespace is a @component with type Agent
function isAgentComponent(ns: Namespace): boolean {
  for (const dec of (ns as any).decorators ?? []) {
    if (dec.definition?.name !== "@component") continue;
    const arg = dec.args?.[0]?.value;
    let typeVal: any;
    if (arg?.valueKind === "ObjectValue") {
      typeVal = (arg.properties as Map<string, any>)?.get("type")?.value;
    } else if (arg?.kind === "Object") {
      typeVal = (arg as any).properties?.type;
    }
    const typeName = typeVal?.valueKind === "EnumValue" ? typeVal.value?.name : typeVal?.value?.name;
    return typeName === "Agent";
  }
  return false;
}

// Walk all ops in a namespace subtree (including interfaces)
function walkAllOpsInNs(ns: Namespace, visitor: (op: Operation) => void): void {
  for (const [, op] of ns.operations) visitor(op);
  for (const [, iface] of ns.interfaces) {
    for (const [, op] of iface.operations) visitor(op);
  }
  for (const [, child] of ns.namespaces) walkAllOpsInNs(child, visitor);
}

// Entry node: a @node whose decorator arg has is_entrypoint: true
function isEntryNode(op: Operation): boolean {
  return op.decorators.some((d: any) => {
    if (d.definition?.name !== "@node") return false;
    const arg = d.args?.[0]?.value;
    if (!arg) return false;
    // #{...} → ObjectValue with properties Map
    if (arg.valueKind === "ObjectValue") {
      const prop = (arg.properties as Map<string, any>)?.get("is_entrypoint");
      return prop?.value?.value === true || prop?.value === true;
    }
    return false;
  });
}

// Join node: a @node whose decorator arg has is_join: true
function isJoinNode(op: Operation): boolean {
  return op.decorators.some((d: any) => {
    if (d.definition?.name !== "@node") return false;
    const arg = d.args?.[0]?.value;
    if (!arg) return false;
    if (arg.valueKind === "ObjectValue") {
      const prop = (arg.properties as Map<string, any>)?.get("is_join");
      return prop?.value?.value === true || prop?.value === true;
    }
    return false;
  });
}

// Terminal node: a @node whose Command.go_to is the End model (by name, non-instantiated)
function isTerminalNode(op: Operation): boolean {
  const shapes = extractCommandShapes(op.returnType);
  return shapes.some((s) => {
    if (!s.goToType || s.goToType.kind !== "Model") return false;
    const m = s.goToType as Model;
    return m.name === "End" && !(m as any).templateMapper;
  });
}

// Check whether a go_to type arg refers to a node op in a different @graph namespace
function isNodeInForeignGraph(goToType: Type | undefined, currentGraph: Namespace): boolean {
  if (!goToType || goToType.kind !== "Operation") return false;
  const op = goToType as Operation;
  // A node op's namespace is the @graph namespace
  if (!op.namespace) return false;
  // Same graph → not foreign
  if (op.namespace === currentGraph) return false;
  // Is the target namespace a @graph? If so, it's a foreign graph routing target
  return hasDecorator(op.namespace, "@graph");
}

// ── Eval/predict index ────────────────────────────────────────────────────────

function buildEvalCalledPredictFqns(program: Program): Set<string> {
  const called = new Set<string>();
  function walkNs(ns: Namespace): void {
    for (const [, op] of ns.operations) {
      // @evalSuite ops call @predict ops directly — that's the quality gate link
      if (hasDecorator(op, "@evalSuite")) {
        for (const target of getCallsFromAst(op)) {
          if (hasDecorator(target, "@predict")) {
            called.add(opFqn(target));
          }
        }
      }
    }
    for (const [, child] of ns.namespaces) walkNs(child);
  }
  walkNs(program.getGlobalNamespaceType());
  return called;
}

// Collect all Command.go_to routing targets across a @graph namespace
// Returns: map from op name → count of incoming Command.go_to edges
function buildRoutingTargetCounts(ns: Namespace): Map<string, number> {
  const counts = new Map<string, number>();
  for (const [, op] of ns.operations) {
    if (!hasDecorator(op, "@node")) continue;
    const rt = op.returnType;
    if (!rt) continue;
    for (const shape of extractCommandShapes(rt)) {
      if (shape.goToType?.kind === "Operation") {
        const name = (shape.goToType as Operation).name;
        counts.set(name, (counts.get(name) ?? 0) + 1);
      }
    }
  }
  return counts;
}

// ── Rule ──────────────────────────────────────────────────────────────────────

export const graphAuthoringRule = createRule({
  name: "graph-authoring",
  severity: "warning",
  description: "Enforces structural authoring rules on @graph namespaces and their @node / @predict ops.",
  messages: {
    missingTerminal: `@graph namespace has no terminal @node op — declare a @node op whose Command.go_to is End.`,
    missingEntry: `@graph namespace has no entry @node — exactly one @node must carry \`@node(#{ is_entrypoint: true })\`.`,
    multipleEntries: `@graph namespace has multiple entry @node ops — exactly one may carry \`@node(#{ is_entrypoint: true })\`.`,
    missingState: `@graph namespace has no state model — declare a model inside the @graph namespace to represent shared graph state.`,
    nodeCallsSibling: `@node op @calls a sibling @node — siblings must be routed via Command.go_to, not @calls.`,
    callsNonEntryNode: `@calls target is a @node inside a @graph but is not the entry node — external callers may only @calls the entry node (@node(#{ is_entrypoint: true })) of a graph.`,
    nodeMissingReturn: `@node op has no outbound Command routing — non-sentinel @node ops must return Command<TUpdate, TGoTo> to declare their routing edge.`,
    predictMissingEval: `@predict op has no @evalSuite targeting it via @calls. Add an @evalSuite that carries @calls([this_predict_op, Score, dataset]).`,
    foreignRoutingTarget: `@node Command.go_to points to a node in a different @graph — routing targets must be nodes in the same @graph.`,
    updateForeignTarget: `@node Command update targets a model that is not the @graph's own state — Command updates must use PickProperties of the graph's declared state model.`,
    dynamicFanoutSharedState: `Dynamic fanout @node (all branches share the same go_to) uses a shared state delta — dynamic fanout branches must carry private (non-state) payloads.`,
    undeclaredJoin: `@node receives multiple incoming Command branches but does not declare \`@node(#{ is_join: true })\`. Add the decorator to acknowledge the join.`,
    joinNoFanout: `@node declares \`is_join: true\` but only one Command branch routes to it — a join requires at least two incoming branches.`,
    nodeUnreachable: `@node op is never targeted by a Command.go_to in this @graph — add a routing edge or remove the node.`,
    graphMissingPredict: `@graph inside an Agent component has no @predict op — every agent graph must declare at least one @predict to represent its probabilistic model boundary.`,
  },
  // a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.rule_graph_authoring
  create(context) {
    let evalCalledPredictFqns: Set<string> | null = null;

    function getEvalIdx(): Set<string> {
      if (!evalCalledPredictFqns) {
        evalCalledPredictFqns = buildEvalCalledPredictFqns(context.program);
      }
      return evalCalledPredictFqns;
    }

    return {
      // a10n:index:call packages.blueprint-typespec.src.rules.graph-authoring.rule.namespace
      // a10n:index:call packages.blueprint-typespec.src.rules.graph-authoring.rule.operation
      namespace(ns: Namespace) {
        if (!hasDecorator(ns, "@graph")) return;

        // ── graph_has_bounds: terminal (go_to is End) + entry (is_entrypoint: true) ──
        let hasTerminal = false;
        let entryCount = 0;
        for (const [, op] of ns.operations) {
          if (!hasDecorator(op, "@node")) continue;
          if (isTerminalNode(op)) hasTerminal = true;
          if (isEntryNode(op)) entryCount++;
        }
        // a10n:blueprint Products.A10nBlueprintEngine.Features.GraphAuthoring.graph_has_bounds
        if (!hasTerminal) {
          context.reportDiagnostic({ messageId: "missingTerminal", target: ns });
        }
        if (entryCount === 0) {
          context.reportDiagnostic({ messageId: "missingEntry", target: ns });
        } else if (entryCount > 1) {
          context.reportDiagnostic({ messageId: "multipleEntries", target: ns });
        }

        // ── graph_has_state ──────────────────────────────────────────────────
        // a10n:blueprint Products.A10nBlueprintEngine.Features.GraphAuthoring.graph_has_state
        const stateModel = findStateModel(ns);
        if (!stateModel) {
          context.reportDiagnostic({ messageId: "missingState", target: ns });
        }

        // ── graph_has_predict / graph_in_agent ───────────────────────────────
        // Only applies to @graph inside an Agent @component
        // a10n:blueprint Products.A10nBlueprintEngine.Features.GraphAuthoring.graph_has_predict
        // a10n:blueprint Products.A10nBlueprintEngine.Features.GraphAuthoring.graph_in_agent
        {
          let compNs: Namespace | undefined = ns.namespace;
          while (compNs && compNs.name) {
            if (hasDecorator(compNs, "@component")) break;
            compNs = compNs.namespace;
          }
          if (compNs && isAgentComponent(compNs)) {
            let hasPredictOp = false;
            walkAllOpsInNs(ns, (op) => { if (hasDecorator(op, "@predict")) hasPredictOp = true; });
            if (!hasPredictOp) {
              context.reportDiagnostic({ messageId: "graphMissingPredict", target: ns });
            }
          }
        }

        // ── join / fanout / reachability analysis ────────────────────────────
        const routingCounts = buildRoutingTargetCounts(ns);

        for (const [, op] of ns.operations) {
          if (!hasDecorator(op, "@node")) continue;
          // Entry node is exempt only from reachability (nothing routes to it — it IS the start)
          // Terminal node is exempt only from reachability (it routes to End, not to peers)
          // Neither is exempt from join checks

          const isJoinDeclared = isJoinNode(op);
          const incomingCount = routingCounts.get(op.name) ?? 0;

          // join checks apply to all non-entry nodes
          if (!isEntryNode(op)) {
            // a10n:blueprint Products.A10nBlueprintEngine.Features.GraphAuthoring.join_declared
            if (incomingCount >= 2 && !isJoinDeclared) {
              context.reportDiagnostic({ messageId: "undeclaredJoin", target: op });
            }
            // a10n:blueprint Products.A10nBlueprintEngine.Features.GraphAuthoring.join_has_fanout
            if (isJoinDeclared && incomingCount < 2) {
              context.reportDiagnostic({ messageId: "joinNoFanout", target: op });
            }
          }

          // reachability check: exempt entry and terminal nodes
          // a10n:blueprint Products.A10nBlueprintEngine.Features.OperationReachability.graph_node_reached
          if (!isEntryNode(op) && !isTerminalNode(op)) {
            if (routingCounts.size > 0 && incomingCount === 0 && hasDecorator(op, "@calls")) {
              context.reportDiagnostic({ messageId: "nodeUnreachable", target: op });
            }
          }
        }
      },

      operation(op: Operation) {
        // ── @node checks ─────────────────────────────────────────────────────
        if (hasDecorator(op, "@node")) {
          // Terminal nodes return Command<..., End> which is valid — no nodeMissingReturn
          // Entry/terminal nodes are exempt from nodeMissingReturn and nodeCallsSibling
          const isTerminal = isTerminalNode(op);
          const isEntry = isEntryNode(op);
          // a10n:blueprint Products.A10nBlueprintEngine.Features.GraphAuthoring.node_in_graph
          if (!isTerminal && !isEntry) {
            // node_declares_return: must return Command<TUpdate, TGoTo>
            // a10n:blueprint Products.A10nBlueprintEngine.Features.GraphAuthoring.node_declares_return
            if (extractCommandShapes(op.returnType).length === 0) {
              context.reportDiagnostic({ messageId: "nodeMissingReturn", target: op });
            }

            // calls_external_only: must not @calls a sibling @node
            // a10n:blueprint Products.A10nBlueprintEngine.Features.GraphAuthoring.calls_external_only
            for (const called of getCallsFromAst(op)) {
              if (hasDecorator(called, "@node") && called.namespace === op.namespace) {
                context.reportDiagnostic({ messageId: "nodeCallsSibling", target: op });
              }
            }

          }

          // Routing / update checks apply to all @node ops (including entry)
          if (!isTerminal && op.namespace && hasDecorator(op.namespace, "@graph")) {
            const graphNs = op.namespace;
            const stateModel = findStateModel(graphNs);
            const shapes = extractCommandShapes(op.returnType);

            for (const shape of shapes) {
              // routing_typed_internal: go_to must not be a node in a foreign @graph
              // a10n:blueprint Products.A10nBlueprintEngine.Features.GraphAuthoring.routing_typed_internal
              if (shape.goToType && isNodeInForeignGraph(shape.goToType, graphNs)) {
                context.reportDiagnostic({ messageId: "foreignRoutingTarget", target: op });
              }

              // update_targets_state: update must not target a store model
              // a10n:blueprint Products.A10nBlueprintEngine.Features.GraphAuthoring.update_targets_state
              // a10n:blueprint Products.A10nBlueprintEngine.Features.GraphAuthoring.update_fields_explicit
              if (stateModel && shape.updateType) {
                const upd = shape.updateType;
                const isPickPropertiesOfStore = (): boolean => {
                  if (upd.kind !== "Model") return false;
                  const m = upd as Model;
                  const templateName = (m as any).templateMapper?.target?.name ?? (m as any).name;
                  if (templateName !== "PickProperties") return false;
                  const args: Type[] = (m as any).templateMapper?.args ?? [];
                  return args.length > 0 && isStoreModel(args[0]);
                };
                if (isStoreModel(upd) || isPickPropertiesOfStore()) {
                  context.reportDiagnostic({ messageId: "updateForeignTarget", target: op });
                }
              }

              // dynamic_fanout: same go_to + shared state delta
              // a10n:blueprint Products.A10nBlueprintEngine.Features.GraphAuthoring.dynamic_fanout_private
              // a10n:blueprint Products.A10nBlueprintEngine.Features.GraphAuthoring.reducer_matches_type
              if (shape.isFanout) {
                const fanoutShapes = extractCommandShapes(op.returnType).filter((s) => s.isFanout);
                if (fanoutShapes.length > 0) {
                  const firstGoTo = fanoutShapes[0]?.goToType;
                  const allSameGoTo = fanoutShapes.every((s) => s.goToType === firstGoTo);
                  if (allSameGoTo && stateModel && shape.updateType && isPickPropertiesOf(shape.updateType, stateModel)) {
                    context.reportDiagnostic({ messageId: "dynamicFanoutSharedState", target: op });
                  }
                }
              }
            }
          }
        }

        // ── calls_external_only (external caller side) / calls_only_entry / tool_calls_subgraph ──
        // Any op that @calls a @node inside a @graph must target the entry node only.
        // Skip if the caller is itself a @node — sibling calls are caught by nodeCallsSibling.
        // a10n:blueprint Products.A10nBlueprintEngine.Features.GraphAuthoring.calls_only_entry
        // a10n:blueprint Products.A10nBlueprintEngine.Features.GraphAuthoring.tool_calls_subgraph
        if (!hasDecorator(op, "@node")) {
          for (const called of getCallsFromAst(op)) {
            if (hasDecorator(called, "@node") && called.namespace && hasDecorator(called.namespace, "@graph")) {
              if (!isEntryNode(called)) {
                context.reportDiagnostic({ messageId: "callsNonEntryNode", target: op });
              }
            }
          }
        }

        // ── predict_has_eval ─────────────────────────────────────────────────
        // FQN-only matching to prevent false negatives when two @predict ops share a short name
        // a10n:blueprint Products.A10nBlueprintEngine.Features.PredictContextBoundary.predict_has_eval
        if (hasDecorator(op, "@predict")) {
          const fqn = opFqn(op);
          const idx = getEvalIdx();
          if (!idx.has(fqn)) {
            context.reportDiagnostic({ messageId: "predictMissingEval", target: op });
          }
        }

      },
    };
  },
});
