// @a10n/blueprint-typespec/src/rules/inline-refs.rule.ts
import {
  ArrayValue,
  createRule,
  DecoratedType,
  DecoratorApplication,
  Model,
  ModelProperty,
  Namespace,
  ObjectValue,
  Operation,
  paramMessage,
  Program,
  StringLiteral,
  StringValue,
  Tuple,
  Type,
  Value,
} from "@typespec/compiler";
import { extractInlineRefs, findSuggestion, resolveRef } from "../validators.js";

function checkDecoratorsOnType(
  decorators: DecoratorApplication[],
  reportTarget: Type,
  program: Program,
  contextNamespace: Namespace | undefined,
  report: (ref: string, target: Type) => void,
): void {
  for (const dec of decorators) {
    for (const arg of dec.args) {
      checkArgEntity(arg.value, reportTarget, program, contextNamespace, report);
    }
  }
}

function checkArgEntity(
  entity: Type | Value,
  reportTarget: Type,
  program: Program,
  contextNamespace: Namespace | undefined,
  report: (ref: string, target: Type) => void,
): void {
  if ("entityKind" in entity) {
    const val = entity as Value;
    if (val.valueKind === "StringValue") {
      const strVal = (val as StringValue).value;
      for (const ref of extractInlineRefs(strVal)) {
        if (!resolveRef(program, ref, contextNamespace)) {
          report(ref, reportTarget);
        }
      }
    }
    if (val.valueKind === "ArrayValue") {
      for (const item of (val as ArrayValue).values) {
        checkArgEntity(item, reportTarget, program, contextNamespace, report);
      }
    }
    if (val.valueKind === "ObjectValue") {
      for (const [, prop] of (val as ObjectValue).properties) {
        checkArgEntity(prop.value, reportTarget, program, contextNamespace, report);
      }
    }
    return;
  }

  const t = entity as Type;
  if (t.kind === "String") {
    const strVal = (t as StringLiteral).value;
    for (const ref of extractInlineRefs(strVal)) {
      if (!resolveRef(program, ref, contextNamespace)) {
        report(ref, reportTarget);
      }
    }
    return;
  }
  if (t.kind === "Tuple") {
    for (const item of (t as Tuple).values) {
      checkArgEntity(item, reportTarget, program, contextNamespace, report);
    }
  }
}

function checkDecoratedType(
  node: DecoratedType & Type,
  program: Program,
  contextNamespace: Namespace | undefined,
  report: (ref: string, target: Type) => void,
): void {
  checkDecoratorsOnType(node.decorators, node, program, contextNamespace, report);
}

export const inlineRefsRule = createRule({
  name: "inline-refs",
  severity: "warning",
  description:
    "Validates that {@Ref} inline references inside string decorator parameters resolve to known types.",
  messages: {
    default: paramMessage`Cannot resolve inline reference {@${"ref"}} — no type named "${"ref"}" found in the program. If this ref describes a hypothetical type in example code (not a real spec type), escape it as \\\\{@${"ref"}} to suppress this warning.`,
    withSuggestion: paramMessage`Cannot resolve inline reference {@${"ref"}} — no type named "${"ref"}" found in the program. Did you mean {@${"suggestion"}}? If this ref describes a hypothetical type in example code (not a real spec type), escape it as \\\\{@${"ref"}} to suppress this warning.`,
  },
  // a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.rule_inline_refs
  // a10n:blueprint Products.A10nBlueprintEngine.Features.SpecReferences.no_std_ref
  // a10n:blueprint Products.A10nBlueprintEngine.Features.SpecReferences.no_generic_ref
  create(context) {
    function reportMissing(ref: string, target: Type): void {
      const suggestion = findSuggestion(context.program, ref);
      if (suggestion && suggestion !== ref) {
        context.reportDiagnostic({
          messageId: "withSuggestion",
          format: { ref, suggestion },
          target,
        });
      } else {
        // a10n:blueprint Products.A10nBlueprintEngine.Features.SpecReferences.no_std_ref
        // a10n:blueprint Products.A10nBlueprintEngine.Features.SpecReferences.no_generic_ref
        context.reportDiagnostic({
          messageId: "default",
          format: { ref },
          target,
        });
      }
    }

    return {
      // a10n:index:call packages.blueprint-typespec.src.rules.inline-refs.rule.model
      // a10n:index:call packages.blueprint-typespec.src.rules.inline-refs.rule.modelProperty
      // a10n:index:call packages.blueprint-typespec.src.rules.inline-refs.rule.operation
      // a10n:index:call packages.blueprint-typespec.src.rules.inline-refs.rule.namespace
      model(model: Model) {
        const ns = model.namespace;
        // a10n:index:call packages.blueprint-typespec.src.rules.inline-refs.rule.reportMissing
        checkDecoratedType(model, context.program, ns, reportMissing);
      },
      // a10n:blueprint:ignore
      modelProperty(prop: ModelProperty) {
        const ns = prop.model?.namespace;
        checkDecoratedType(prop, context.program, ns, reportMissing);
      },
      // a10n:blueprint:ignore
      operation(op: Operation) {
        const ns = op.namespace;
        checkDecoratedType(op, context.program, ns, reportMissing);
      },
      // a10n:blueprint:ignore
      namespace(ns: Namespace) {
        checkDecoratedType(ns, context.program, ns, reportMissing);
      },
    };
  },
});
