// @a10n/blueprint-typespec/src/rules/invariant-projected-once-per-op.rule.ts
import { createRule, Operation, paramMessage } from "@typespec/compiler";

function opFqn(op: Operation): string {
  const parts: string[] = [];
  let ns = op.namespace;
  while (ns && ns.name) { parts.unshift(ns.name); ns = ns.namespace; }
  return parts.length ? `${parts.join(".")}.${op.name}` : op.name;
}

function invFqn(v: any): string {
  const parts: string[] = [];
  let cur = v.namespace;
  while (cur && cur.name) { parts.unshift(cur.name); cur = cur.namespace; }
  return parts.length ? `${parts.join(".")}.${v.name}` : v.name;
}

export const invariantProjectedOncePerOpRule = createRule({
  name: "invariant-projected-once-per-op",
  severity: "warning",
  description:
    "An op must not list the same @invariant more than once across its @projects decorator(s).",
  messages: {
    default: paramMessage`op "${"opFqn"}" projects invariant "${"inv"}" more than once — list each invariant at most once.`,
  },
  // a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.rule_invariant_projected_once_per_op
  create(context) {
    return {
      // a10n:index:call packages.blueprint-typespec.src.rules.invariant-projected-once-per-op.rule.operation
      operation(op: Operation) {
        if (!op.decorators.some((d) => d.definition?.name === "@projects")) return;

        const seen = new Set<string>();
        const reported = new Set<string>();

        for (const dec of op.decorators) {
          if (dec.definition?.name !== "@projects") continue;
          for (const arg of dec.args) {
            const v = arg.value as any;
            if (v?.kind !== "Operation") continue;
            if (!v.decorators?.some((d: any) => d.definition?.name === "@invariant")) continue;
            const fqn = invFqn(v);
            // a10n:blueprint Products.A10nBlueprintEngine.Features.InvariantProjection.invariant_projected_once
            if (seen.has(fqn) && !reported.has(fqn)) {
              reported.add(fqn);
              context.reportDiagnostic({
                messageId: "default",
                format: { opFqn: opFqn(op), inv: fqn },
                target: op,
              });
            }
            seen.add(fqn);
          }
        }
      },
    };
  },
});
