// @a10n/blueprint-typespec/src/rules/namespace-path.rule.ts
import { createRule, Namespace, paramMessage } from "@typespec/compiler";

// SyntaxKind.NamespaceStatement = 8 (not exported from public API)
const NS_STATEMENT_KIND = 8;
import { getTypeFilePath, isExemptPath, toKebab } from "../validators.js";

function hasDecorator(node: any, name: string): boolean {
  return (node?.decorators ?? []).some((d: any) => d.definition?.name === name);
}

// Returns true when this namespace has @component(type: Store) on it.
function isStoreComponentNs(ns: Namespace): boolean {
  for (const dec of (ns as any).decorators ?? []) {
    if (dec.definition?.name !== "@component") continue;
    const arg = dec.args?.[0]?.value;
    if (!arg) continue;
    let typeVal: any;
    if (arg.valueKind === "ObjectValue") {
      typeVal = (arg.properties as Map<string, any>)?.get("type")?.value;
    } else if (arg.kind === "Object") {
      typeVal = (arg as any).properties?.type;
    }
    if (!typeVal) continue;
    const typeStr =
      typeVal.valueKind === "EnumValue"
        ? typeVal.value?.name
        : typeVal.value?.name ?? String(typeVal);
    if (typeStr === "Store") return true;
  }
  return false;
}

// Returns true when any model anywhere under this namespace carries a DBML
// table decorator (@table or fully-qualified @TypeSpec.Dbml.table). Matches
// the detection idiom from store-model-coverage.rule.ts.
function hasDbmlTableInTree(ns: Namespace): boolean {
  for (const [, model] of (ns as any).models ?? []) {
    for (const dec of (model as any).decorators ?? []) {
      const name = dec.definition?.name;
      if (name === "@table" || name === "@TypeSpec.Dbml.table") return true;
    }
  }
  for (const [, child] of (ns as any).namespaces ?? []) {
    if (hasDbmlTableInTree(child)) return true;
  }
  return false;
}

// Returns true when @storage is present anywhere in the namespace tree (the
// decorator typically sits on the @component namespace itself, but file-store
// authors sometimes nest it on a sub-namespace).
function hasStorageInTree(ns: Namespace): boolean {
  if (hasDecorator(ns, "@storage")) return true;
  for (const [, child] of (ns as any).namespaces ?? []) {
    if (hasStorageInTree(child)) return true;
  }
  return false;
}

/**
 * Walk up the namespace tree to find the nearest @component ancestor (or self).
 */
// a10n:blueprint:ignore
function findComponentAncestorNs(ns: Namespace): Namespace | undefined {
  let cur: Namespace | undefined = ns;
  while (cur && cur.name) {
    if (hasDecorator(cur, "@component")) return cur;
    cur = cur.namespace;
  }
  return undefined;
}

/**
 * Allowed file path patterns inside state/, derived directly from the globs in
 * checks.yaml. Each entry is a minimatch-style pattern where:
 *   *   matches one path segment (no slashes)
 *   **  matches any number of segments
 *
 * Component sub-folder permissions are type-suffix-gated — http/ is only valid
 * inside *-service/, tools/ only inside *-agent/, etc. The rule matches the
 * actual file path against this list; if none match, the file is in the wrong place.
 *
 * Source: state/.a10n/task-executor/checks.yaml globs
 */
const ALLOWED_PATH_PATTERNS: ReadonlyArray<string> = [
  // ── Components ─────────────────────────────────────────────────────────────
  // All components: root main file only — extra type-specific paths are
  // enumerated below per component kind.
  "state/components/*/main.tsp",
  // All components: test suites and evals (type-agnostic)
  "state/components/*/tests/*/*.tsp",
  "state/components/*/evals/*/*.tsp",
  // All components: flows (any component type can have flows)
  "state/components/**/flows/*.tsp",

  // Agent-specific
  "state/components/*-agent/tools/*.tsp",
  "state/components/*-agent/graph/*.tsp",
  "state/components/*-agent/interrupts/*.tsp",
  "state/components/*-agent/predicts/*.tsp",
  // Workers — any component type
  "state/components/*/workers/*.tsp",
  // datasets/ allowed for any component kind
  "state/components/*/datasets/*/*.tsp",

  // Service-specific
  "state/components/*-service/http/*.tsp",
  "state/components/*-service/mcp/*.tsp",
  "state/components/*-service/events/*.tsp",
  "state/components/*-service/models/*.tsp",

  // SDK-specific
  "state/components/*-sdk/operations/*.tsp",

  // Store-specific (any depth — stores have sub-namespaces inline)
  "state/components/*-file-store/*.tsp",
  "state/components/*-file-store/**/*.tsp",
  "state/components/*-relational-store/*.tsp",
  "state/components/*-relational-store/**/*.tsp",

  // UI-specific (flat route files or any depth nested sub-routes)
  "state/components/*-ui/routes/*.tsp",
  "state/components/*-ui/routes/**/*.tsp",

  // CLI-specific
  "state/components/*-service/commands/*.tsp",

  // ── Products ────────────────────────────────────────────────────────────────
  "state/products/*/main.tsp",
  "state/products/*/features/*.tsp",
  "state/products/*/entities/*.tsp",

  // ── Dependencies ─────────────────────────────────────────────────────────
  "state/dependencies/*/main.tsp",
  "state/dependencies/*/*/main.tsp",
  // Nested @component inside a @dependency parent — Dependencies.<Dep>.Components.<Name>.
  // The nested component lives one segment deeper, alongside the dep's own
  // main.tsp, so its file path includes a literal "components/" segment.
  "state/dependencies/*/components/*/main.tsp",
  // untracked.tsp is an intentional exception — multiple deps in one file
  "state/dependencies/untracked.tsp",
];

/**
 * Returns true if this namespace node is the "leaf" of a dotted namespace
 * declaration — the innermost segment that actually owns the file.
 *
 * When TypeSpec compiles `namespace A.B.C;` it creates three Namespace type
 * objects sharing the same source file: A, A.B, and A.B.C. Each has a
 * NamespaceStatementNode. The outer ones (A, A.B) have their `statements`
 * field pointing to another NamespaceStatementNode (the next segment), while
 * the leaf (A.B.C) has `statements` as an array (possibly empty) or undefined.
 *
 * Checking only the leaf avoids duplicate diagnostics for the same file.
 */
function isDeclaredLeaf(ns: Namespace): boolean {
  const node = ns.node as any;
  if (!node || node.kind !== NS_STATEMENT_KIND) return false;
  // If statements is a single NamespaceStatementNode (not an array), this is
  // an intermediate segment — the real declaration is deeper.
  const stmts = node.statements;
  if (stmts && !Array.isArray(stmts) && stmts.kind === NS_STATEMENT_KIND) {
    return false;
  }
  return true;
}

/**
 * Convert a namespace FQN to the expected file-path fragment.
 * Every PascalCase segment → kebab-case, joined with "/".
 *
 * Terminal T\d{3} and E\d{3} segments (test suite / eval namespace ids) are
 * stripped from the fragment because their corresponding files use numeric
 * prefixes (001_suite-name.tsp) that don't match the segment name directly.
 * The directory path (tests/pr-review/ or evals/pr-review/) is still enforced.
 *
 * Examples:
 *   Components.ReviewAgent                     → "components/review-agent"
 *   Components.ReviewAgent.Http                → "components/review-agent/http"
 *   Components.ReviewAgent.Tests.PrReview.T001 → "components/review-agent/tests/pr-review"
 *   Components.ReviewAgent.Evals.PrReview.E001 → "components/review-agent/evals/pr-review"
 *   Products.ReviewBot                         → "products/review-bot"
 *   Products.ReviewBot.Features.PrReview       → "products/review-bot/features/pr-review"
 *   Products.ReviewBot.Entities.Task           → "products/review-bot/entities/task"
 *   Dependencies.StripeApi                     → "dependencies/stripe-api"
 */
function expectedPathFragment(fqn: string): string {
  let segs = fqn.split(".");
  // Strip terminal T\d{3} (test suite) or E\d{3} (eval) — these use numeric-
  // prefixed file names that don't map directly to the segment name.
  const last = segs[segs.length - 1];
  if (last && /^[TE]\d{3}$/.test(last)) {
    segs = segs.slice(0, -1);
  }
  return segs.map(toKebab).join("/");
}

/**
 * Return the directory portion of a file path (everything before the last "/").
 * e.g. "state/components/review-agent/evals/pr-review/001_quality.tsp"
 *   → "state/components/review-agent/evals/pr-review"
 */
function dirOf(filePath: string): string {
  const norm = filePath.replace(/\\/g, "/");
  const slash = norm.lastIndexOf("/");
  return slash !== -1 ? norm.slice(0, slash) : norm;
}

/**
 * Check whether the namespace fragment matches the file path exactly — no
 * looser substring containment. The fragment must appear at a path-segment
 * boundary and reach either:
 *   - the end of the directory (for NNN_*.tsp files where the last namespace
 *     segment is a terminal T\d{3}/E\d{3} already stripped from the fragment), or
 *   - the end of the filename stem (for single-file namespaces like
 *     features/pr-review.tsp, entities/task.tsp, commands/doctor.tsp where
 *     the last segment maps to the filename, not a sub-folder).
 *
 * Concretely:
 *   fragment "components/review-agent/evals/pr-review"
 *     matches dir  "…/components/review-agent/evals/pr-review"    ✓ (NNN_ file)
 *     rejects dir  "…/components/review-agent"                    ✗ (too shallow)
 *
 *   fragment "products/review-bot/features/pr-review"
 *     matches path "…/products/review-bot/features/pr-review.tsp" ✓ (stem match)
 *     rejects path "…/products/review-bot/features/other.tsp"     ✗ (wrong file)
 */
function pathContainsFragment(filePath: string, fragment: string): boolean {
  const norm = filePath.replace(/\\/g, "/");
  const dir = dirOf(norm);
  const idx = norm.indexOf(fragment);
  if (idx === -1) return false;
  const before = idx === 0 || norm[idx - 1] === "/";
  if (!before) return false;
  const end = idx + fragment.length;
  // Case 1: fragment reaches end of filename stem — e.g. features/pr-review in pr-review.tsp
  if (norm[end] === ".") return true;
  // Case 2: fragment reaches end of directory — e.g. evals/pr-review in evals/pr-review/001.tsp
  // Check fragment matches dir exactly (modulo leading prefix already validated by `before`)
  const dirIdx = dir.indexOf(fragment);
  if (dirIdx !== -1) {
    const dirBefore = dirIdx === 0 || dir[dirIdx - 1] === "/";
    if (dirBefore && dirIdx + fragment.length === dir.length) return true;
  }
  return false;
}

/**
 * Match a file path against a glob pattern using the same semantics as
 * ALLOWED_PATH_PATTERNS: * matches one segment, ** matches any segments.
 * Operates on the state/-relative portion of the path.
 */
function matchesPattern(filePath: string, pattern: string): boolean {
  const norm = filePath.replace(/\\/g, "/");
  // Extract the state/-relative portion. In the real spec the root is "state/";
  // in the example project and tests there is no "state/" prefix, so fall back
  // to the first occurrence of a known root folder as the anchor point.
  const stateIdx = norm.indexOf("state/");
  let rel: string;
  if (stateIdx !== -1) {
    rel = norm.slice(stateIdx);
  } else {
    // Find the first occurrence of components/, products/, or dependencies/
    // and prepend "state/" so patterns still match.
    const rootIdx = Math.min(
      ...(["components/", "products/", "dependencies/"].map((r) => {
        const i = norm.indexOf(r);
        return i !== -1 ? i : Infinity;
      }))
    );
    rel = rootIdx !== Infinity ? `state/${norm.slice(rootIdx)}` : norm;
  }

  // Convert glob pattern to regex
  const escaped = pattern
    .replace(/[.+^${}()|[\]\\]/g, "\\$&") // escape regex metacharacters
    .replace(/\*\*/g, "§GLOBSTAR§")         // protect ** before single-* expansion
    .replace(/\*/g, "[^/]+")                // * → one segment
    .replace(/§GLOBSTAR§/g, ".+");          // ** → any segments

  return new RegExp(`^${escaped}$`).test(rel);
}

/**
 * Returns true if the file path is allowed by any pattern in ALLOWED_PATH_PATTERNS.
 */
function isAllowedPath(filePath: string): boolean {
  return ALLOWED_PATH_PATTERNS.some((p) => matchesPattern(filePath, p));
}

/**
 * Extract the deepest directory segment from a file path.
 * e.g. "state/components/a10n-dotdir-file-store/blueprint.tsp" → "a10n-dotdir-file-store"
 */
function getComponentDirSegment(filePath: string): string | undefined {
  const norm = filePath.replace(/\\/g, "/");
  // Find "components/" prefix and extract the component directory
  const compIdx = norm.indexOf("components/");
  if (compIdx === -1) return undefined;
  const afterComponents = norm.slice(compIdx + "components/".length);
  const slashIdx = afterComponents.indexOf("/");
  return slashIdx !== -1 ? afterComponents.slice(0, slashIdx) : afterComponents.replace(/\.tsp$/, "");
}

export const namespacePathRule = createRule({
  name: "namespace-path",
  severity: "warning",
  description:
    "The file path must match the namespace FQN converted to kebab-case path segments. Move the file or rename the namespace.",
  messages: {
    default: paramMessage`Namespace "${"fqn"}" must be declared in a file under "${"expected"}" but is in "${"actual"}".`,
    disallowedPath: paramMessage`File "${"actual"}" is not in an allowed location for ${"root"} files. Check the component type suffix and sub-folder name.`,
    suffixFileStore: paramMessage`@storage component "${"name"}" must be declared in a directory ending in \`-file-store\` (e.g. \`my-file-store/\`), but is in "${"actual"}".`,
    suffixRelationalStore: paramMessage`Relational @component(Store) "${"name"}" (DBML-tabled) must be declared in a directory ending in \`-relational-store\` (e.g. \`my-relational-store/\`), but is in "${"actual"}".`,
    nsSuffixFileStore: paramMessage`@storage component namespace "${"name"}" must end in \`FileStore\` (e.g. \`Components.MyFileStore\`).`,
    nsSuffixRelationalStore: paramMessage`Relational @component(Store) namespace "${"name"}" must end in \`RelationalStore\` (e.g. \`Components.MyRelationalStore\`).`,
    storeTypeUndeclared: paramMessage`@component(Store) "${"name"}" must declare its store type — add @storage (file store) or DBML decorators on at least one model (relational store).`,
  },
  // a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.rule_namespace_path
  // a10n:blueprint Products.A10nBlueprintEngine.Features.ToolAuthoring.tool_namespace
  // a10n:blueprint Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.two_layers_only
  // a10n:todo Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.two_layers_only — full SpecLayerKind/SpecConstructKind classification is not yet computed at compile time; ALLOWED_PATH_PATTERNS structurally partitions files into the domain (entities/, features/) and system (components/) layer roots and rejects anything that straddles or escapes both.
  // a10n:blueprint Products.A10nBlueprintEngine.Features.InterfaceNamespacePlacement.http_namespace_placement
  // a10n:blueprint Products.A10nBlueprintEngine.Features.InterfaceNamespacePlacement.mcp_namespace_placement
  // a10n:blueprint Products.A10nBlueprintEngine.Features.InterfaceNamespacePlacement.cli_namespace_placement
  // a10n:blueprint Products.A10nBlueprintEngine.Features.InterfaceNamespacePlacement.workers_namespace_placement
  // a10n:blueprint Products.A10nBlueprintEngine.Features.InterfaceNamespacePlacement.events_namespace_placement
  // a10n:blueprint Products.A10nBlueprintEngine.Features.DatasetAuthoring.dataset_namespace
  // a10n:blueprint Products.A10nBlueprintEngine.Features.DomainAuthoring.entity_one_product
  create(context) {
    // a10n:index:call Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.entity_path
    // a10n:index:call Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.feature_path
    // a10n:index:call Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.feature_single_file
    // a10n:index:call Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.eval_path
    // a10n:index:call Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.layer_from_path
    // a10n:index:call Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.entity_layer
    // a10n:index:call Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.feature_layer
    // a10n:index:call Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.system_constructs_layer
    return {
      // a10n:index:call packages.blueprint-typespec.src.rules.namespace-path.rule.namespace
      namespace(ns: Namespace) {
        if (!ns.name) return;

        const fqn = (() => {
          const parts: string[] = [];
          let cur: Namespace | undefined = ns;
          while (cur && cur.name) { parts.unshift(cur.name); cur = cur.namespace; }
          return parts.join(".");
        })();

        // Only enforce our known roots — skip TypeSpec, State, Http, third-party.
        const root = fqn.split(".")[0];
        if (!["Components", "Products", "Dependencies"].includes(root)) return;

        // Only fire on the leaf namespace of each declaration — the innermost
        // segment that actually owns the file node. Intermediate segments
        // (synthesized by TypeSpec for dotted namespace paths) are skipped to
        // avoid duplicate diagnostics for the same file.
        if (!isDeclaredLeaf(ns)) return;

        const filePath = getTypeFilePath(ns);
        if (!filePath) return;
        if (isExemptPath(filePath)) return;

        // Check 1: file is in an allowed location for this root.
        // a10n:blueprint Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.layer_from_path
        // a10n:blueprint Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.entity_layer
        // a10n:blueprint Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.feature_layer
        // a10n:blueprint Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.system_constructs_layer
        if (!isAllowedPath(filePath)) {
          context.reportDiagnostic({
            messageId: "disallowedPath",
            format: { actual: filePath, root },
            target: ns,
          });
          return;
        }

        // UI route convention: Components.ReviewUi.Routes.ReviewList → components/review-ui/routes/review-list
        // The `Routes` segment is explicit in the namespace and maps directly to the `routes/` folder.
        // No special-case handler needed — the standard FQN→kebab path check handles it correctly.

        // Check 2: the namespace FQN → kebab fragment is present in the file path.
        // a10n:blueprint Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.entity_path
        // a10n:blueprint Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.feature_path
        // a10n:blueprint Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.feature_single_file
        // a10n:blueprint Products.A10nBlueprintEngine.Features.TwoLayerArchitecture.eval_path
        const fragment = expectedPathFragment(fqn);
        if (!pathContainsFragment(filePath, fragment)) {
          context.reportDiagnostic({
            messageId: "default",
            format: { fqn, expected: fragment, actual: filePath },
            target: ns,
          });
          return;
        }


        // Store-component enforcement. The store's type comes from a
        // decorator/decorator-pattern, not the namespace name:
        //   @storage anywhere in the tree                       → File store
        //   At least one model with @TypeSpec.Dbml.table        → Relational store
        // We require @component(Store) namespaces to carry one of these. A
        // store with neither is "type-undeclared" and gets a diagnostic.
        // For each declared type we also enforce the directory and namespace
        // name suffix:
        //   File store     → dir ends in `-file-store`         + ns ends in `FileStore`
        //   Relational     → dir ends in `-relational-store`   + ns ends in `RelationalStore`
        // These all fire only when the path check above passed so we don't
        // double-report with namespace-path mismatch diagnostics.
        if (hasDecorator(ns, "@component") && isStoreComponentNs(ns)) {
          const isFileStore = hasStorageInTree(ns);
          const isRelational = hasDbmlTableInTree(ns);

          if (!isFileStore && !isRelational) {
            context.reportDiagnostic({
              messageId: "storeTypeUndeclared",
              format: { name: ns.name },
              target: ns,
            });
          } else {
            const dirSegment = getComponentDirSegment(filePath);
            // a10n:blueprint Products.A10nBlueprintEngine.Features.FileStoreAuthoring.suffix_file_store
            if (isFileStore) {
              if (dirSegment && !dirSegment.endsWith("-file-store")) {
                context.reportDiagnostic({
                  messageId: "suffixFileStore",
                  format: { name: ns.name, actual: filePath },
                  target: ns,
                });
              }
              if (!ns.name.endsWith("FileStore")) {
                context.reportDiagnostic({
                  messageId: "nsSuffixFileStore",
                  format: { name: ns.name },
                  target: ns,
                });
              }
            }
            // a10n:blueprint Products.A10nBlueprintEngine.Features.RelationalStoreAuthoring.suffix_relational_store
            if (isRelational) {
              if (dirSegment && !dirSegment.endsWith("-relational-store")) {
                context.reportDiagnostic({
                  messageId: "suffixRelationalStore",
                  format: { name: ns.name, actual: filePath },
                  target: ns,
                });
              }
              if (!ns.name.endsWith("RelationalStore")) {
                context.reportDiagnostic({
                  messageId: "nsSuffixRelationalStore",
                  format: { name: ns.name },
                  target: ns,
                });
              }
            }
          }
        }
      },
    };
  },
});
