// @a10n/blueprint-typespec/src/rules/op-invariant-cap.rule.ts
import { createRule, Operation, paramMessage } from "@typespec/compiler";

const OP_INVARIANT_CAP = 8;

function opFqn(op: Operation): string {
  const parts: string[] = [];
  let ns = op.namespace;
  while (ns && ns.name) { parts.unshift(ns.name); ns = ns.namespace; }
  return parts.length ? `${parts.join(".")}.${op.name}` : op.name;
}

/** Collect distinct invariant FQNs from all @projects decorators on an op. */
function projectsInvariantTargets(op: Operation): Set<string> {
  const fqns = new Set<string>();
  for (const dec of op.decorators) {
    if (dec.definition?.name !== "@projects") continue;
    for (const arg of dec.args) {
      const v = arg.value as any;
      if (v?.kind === "Operation" && v.decorators?.some((d: any) => d.definition?.name === "@invariant")) {
        const invNs: string[] = [];
        let cur = v.namespace;
        while (cur && cur.name) { invNs.unshift(cur.name); cur = cur.namespace; }
        const fqn = invNs.length ? `${invNs.join(".")}.${v.name}` : v.name;
        fqns.add(fqn);
      }
    }
  }
  return fqns;
}

export const opInvariantCapRule = createRule({
  name: "op-invariant-cap",
  severity: "warning",
  description:
    `An op must not @projects more than ${OP_INVARIANT_CAP} invariants. Split the op or move invariants to a peer op.`,
  messages: {
    default: paramMessage`op "${"opFqn"}" projects ${"count"} invariants; max is ${String(OP_INVARIANT_CAP)}. Split the op or move invariants to a peer op.`,
  },
  // a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.rule_op_invariant_cap
  create(context) {
    return {
      // a10n:index:call packages.blueprint-typespec.src.rules.op-invariant-cap.rule.operation
      operation(op: Operation) {
        if (!op.decorators.some((d) => d.definition?.name === "@projects")) return;
        const invs = projectsInvariantTargets(op);
        if (invs.size > OP_INVARIANT_CAP) {
          // a10n:blueprint Products.A10nBlueprintEngine.Features.InvariantProjection.op_invariant_cap
          context.reportDiagnostic({
            messageId: "default",
            format: { opFqn: opFqn(op), count: String(invs.size) },
            target: op,
          });
        }
      },
    };
  },
});
