// @a10n/blueprint-typespec/src/rules/projects-targets-invariant.rule.ts
import { createRule, Operation, paramMessage } from "@typespec/compiler";
import { isProjectingOpKind } from "../projecting-op.js";

function hasDecorator(node: { decorators: readonly { definition?: { name: string } }[] }, name: string): boolean {
  return node.decorators.some((d) => d.definition?.name === name);
}

function opFqn(op: Operation): string {
  const parts: string[] = [];
  let ns = op.namespace;
  while (ns && ns.name) { parts.unshift(ns.name); ns = ns.namespace; }
  return parts.length ? `${parts.join(".")}.${op.name}` : op.name;
}

export const projectsTargetsInvariantRule = createRule({
  name: "projects-targets-invariant",
  severity: "warning",
  description:
    "@projects on a component op must target @invariant ops only; @projects must not appear on non-projecting op kinds.",
  messages: {
    wrongTarget: paramMessage`@projects on op "${"opFqn"}" targets "${"ref"}" which is not an @invariant — ops project onto invariants only (entity-field projection lives on model fields).`,
    wrongSource: paramMessage`@projects on op "${"opFqn"}" is not allowed: it must be an op that enforces invariants — a @tool, @node, @predict, @interrupt, @operation, an op in an Http/Mcp/Grpc interface, an op in a @uiRoute or @worker namespace, or a @signal/@dataset op.`,
  },
  // a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.rule_projects_targets_invariant
  // a10n:blueprint Products.A10nBlueprintEngine.Features.SystemAuthoring.op_projects_inv
  create(context) {
    return {
      // a10n:index:call packages.blueprint-typespec.src.rules.projects-targets-invariant.rule.operation
      operation(op: Operation) {
        if (!hasDecorator(op, "@projects")) return;

        // @signal and @dataset are also valid @projects sources (as proofs)
        const isLegacyProof = hasDecorator(op, "@signal") || hasDecorator(op, "@dataset");
        if (!isProjectingOpKind(op) && !isLegacyProof) {
          // The @projects decorator itself fires projects-op-not-proof for this case.
          // Emit wrongSource as a linter-level signal so it appears in rule reports too.
          context.reportDiagnostic({
            messageId: "wrongSource",
            format: { opFqn: opFqn(op) },
            target: op,
          });
          return;
        }

        for (const dec of op.decorators) {
          if (dec.definition?.name !== "@projects") continue;
          for (const arg of dec.args) {
            const v = arg.value as any;
            if (!v) continue;
            if (v.kind === "Operation" && hasDecorator(v, "@invariant")) continue;
            // Not an invariant op
            const ref = v.name ?? String(v);
            context.reportDiagnostic({
              messageId: "wrongTarget",
              format: { opFqn: opFqn(op), ref },
              target: op,
            });
          }
        }
      },
    };
  },
});
