// @a10n/blueprint-typespec/src/rules/sdk-operations-namespace.rule.ts
import { createRule, Operation, paramMessage } from "@typespec/compiler";
import { findComponentAncestor, getComponentType, hasAncestorNamespaceNamed } from "../projecting-op.js";

function opFqn(op: Operation): string {
  const parts: string[] = [];
  let ns = op.namespace;
  while (ns && ns.name) { parts.unshift(ns.name); ns = ns.namespace; }
  return parts.length ? `${parts.join(".")}.${op.name}` : op.name;
}

export const sdkOperationsNamespaceRule = createRule({
  name: "sdk-operations-namespace",
  severity: "warning",
  description:
    "Every @operation op inside an Sdk @component must be nested under a 'namespace Operations' block.",
  messages: {
    default: paramMessage`SDK operation "${"opFqn"}" must live inside a 'namespace Operations' block — flat ops are invisible to the planner and mapping check (they don't match the Sdk.Operations.* classifier).`,
  },
  // a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.rule_sdk_operations_namespace
  // a10n:blueprint Products.A10nBlueprintEngine.Features.InterfaceNamespacePlacement.sdk_operations_namespace
  create(context) {
    return {
      // a10n:index:call packages.blueprint-typespec.src.rules.sdk-operations-namespace.rule.operation
      operation(op: Operation) {
        const comp = findComponentAncestor(op);
        if (!comp || getComponentType(comp) !== "Sdk") return;
        if (!op.decorators.some((d) => d.definition?.name === "@operation")) return;
        if (!hasAncestorNamespaceNamed(op, "Operations")) {
          context.reportDiagnostic({
            messageId: "default",
            format: { opFqn: opFqn(op) },
            target: op,
          });
        }
      },
    };
  },
});
