// @a10n/blueprint-typespec/src/rules/single-block-form.rule.ts
import {
  createRule,
  getLocationContext,
  getSourceLocation,
  paramMessage,
  Program,
} from "@typespec/compiler";

// NS_STATEMENT_KIND = 8 (inlined; SyntaxKind isn't re-exported
// from the @typespec/compiler entry, only from @typespec/compiler/ast).
const NS_STATEMENT_KIND = 8;

// A namespace AST node uses block form when its `statements` field is an
// array (possibly empty). Semicolon form leaves `statements` as undefined.
// A dotted intermediate segment (`namespace A.B { ... }` — the `A` part)
// has `statements` set to a single NamespaceStatementNode, not an array.
function isBlockFormNode(node: any): boolean {
  if (!node || node.kind !== NS_STATEMENT_KIND) return false;
  return Array.isArray(node.statements);
}

// Walk parent chain to build the FQN for a namespace AST node when the
// semantic checker can't resolve it (fallback only).
function leafFqnFromAst(node: any): string | undefined {
  if (!Array.isArray(node.statements)) return undefined;
  const ancestors: string[] = [];
  let cur: any = node.parent;
  while (cur && cur.kind === NS_STATEMENT_KIND) {
    const idText: string | undefined = cur.id?.sv;
    if (idText) ancestors.unshift(idText);
    cur = cur.parent;
  }
  const leafId: string | undefined = node.id?.sv;
  const all = leafId ? [...ancestors, leafId] : ancestors;
  return all.join(".") || undefined;
}

interface BlockOccurrence {
  node: any;
  file: string;
  fileOrder: number; // index into program.sourceFiles iteration
}

export const singleBlockFormRule = createRule({
  name: "single-block-form",
  severity: "warning",
  description:
    "A namespace FQN may use block form `namespace X { ... }` in at most one file. TypeSpec uses INSERT-OR-REPLACE symbol resolution: when multiple files declare the same namespace in block form, only the last-processed file's block is indexed and its decorators win — earlier blocks' decorators are silently dropped. Other files that augment the same namespace must use bare `namespace X;` form.",
  messages: {
    default: paramMessage`Namespace "${"fqn"}" is declared in block form \`namespace X { }\` here AND in "${"otherFile"}" — TypeSpec's symbol indexing is INSERT-OR-REPLACE, so this block competes for and (because it is processed last) overwrites the other file's symbol entry, silently dropping its decorators. Convert this declaration to bare \`namespace X;\` form so the other file's block stays canonical.`,
  },
  // a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.rule_single_block_form
  create(context) {
    return {
      // a10n:index:call packages.blueprint-typespec.src.rules.single-block-form.rule.root
      root(program: Program) {
        // FQN → list of block-form occurrences in file-iteration order.
        const occurrencesByFqn = new Map<string, BlockOccurrence[]>();
        // Track AST nodes already recorded so re-visits don't double-count.
        const seen = new WeakSet<object>();

        let fileOrder = 0;
        for (const [, script] of program.sourceFiles) {
          // Only consider user code. Library namespaces (e.g. @typespec/http)
          // legitimately split block-form declarations across files.
          const locCtx = getLocationContext(program, script);
          if (locCtx.type !== "project") {
            fileOrder++;
            continue;
          }

          const currentFileOrder = fileOrder++;
          const visit = (node: any) => {
            if (!node) return;
            if (node.kind === NS_STATEMENT_KIND) {
              if (isBlockFormNode(node) && !seen.has(node)) {
                seen.add(node);
                // Prefer semantic FQN so file-scope `namespace X;`
                // declarations contribute their prefix correctly. Fall back
                // to AST walk if the checker can't resolve.
                let fqn: string | undefined;
                try {
                  const sem: any = (program as any).checker?.getTypeForNode?.(node);
                  if (sem && sem.kind === "Namespace") {
                    const parts: string[] = [];
                    let cur: any = sem;
                    while (cur && cur.name) {
                      parts.unshift(cur.name);
                      cur = cur.namespace;
                    }
                    fqn = parts.join(".") || undefined;
                  }
                } catch {
                  fqn = undefined;
                }
                if (!fqn) fqn = leafFqnFromAst(node);
                if (fqn) {
                  const loc = getSourceLocation(node);
                  const file = loc?.file?.path ?? "<unknown>";
                  const list = occurrencesByFqn.get(fqn) ?? [];
                  list.push({ node, file, fileOrder: currentFileOrder });
                  occurrencesByFqn.set(fqn, list);
                }
              }
              // Descend into block-form children.
              if (Array.isArray(node.statements)) {
                for (const child of node.statements) visit(child);
              } else if (node.statements && node.statements.kind === NS_STATEMENT_KIND) {
                // Dotted intermediate segment — descend into its leaf.
                visit(node.statements);
              }
              return;
            }
            // Top-level script: walk its statements.
            if (Array.isArray(node.statements)) {
              for (const child of node.statements) visit(child);
            }
          };

          visit(script);
        }

        // For each FQN with more than one block-form decl, TypeSpec's
        // INSERT-OR-REPLACE symbol resolution makes the LAST-processed
        // file's block win — overwriting the earlier files' decorators.
        // The fix is for the WINNER to use bare `namespace X;`, so the
        // earlier (decorator-bearing) block stays canonical. Therefore
        // we flag the LAST occurrence (the one whose block overwrites)
        // and point it at the earlier (about-to-be-overwritten) file.
        for (const [fqn, occs] of occurrencesByFqn) {
          if (occs.length < 2) continue;
          occs.sort((a, b) => a.fileOrder - b.fileOrder);
          // The earliest-processed block is what we want to preserve as
          // canonical. Every later block-form decl of the same FQN
          // overwrites it — flag those.
          const canonical = occs[0]!;
          for (let i = 1; i < occs.length; i++) {
            const overwriter = occs[i]!;
            context.reportDiagnostic({
              messageId: "default",
              format: { fqn, otherFile: canonical.file },
              target: overwriter.node,
            });
          }
        }
      },
    };
  },
});
