// @a10n/blueprint-typespec/src/rules/spec-references.rule.ts
import { createRule, Operation, paramMessage } from "@typespec/compiler";

// actor_ref (spec: bare external-trigger nouns must use {@Actor}):
// NOT implementable deterministically. {@Actor} is a declared scalar in the State
// namespace — but detecting the ABSENCE of it when an actor is mentioned requires
// knowing which words in any language denote external actors. Any word list is
// incomplete and has false positives (e.g. "Operator" as a TypeSpec model name).
// Left to eval.
//
// no_std_ref / no_generic_ref:
// Already covered by inline-refs rule (unresolved-inline-ref fires for any
// {@ref} that doesn't resolve to a declared construct). No additional check needed.
//
// ref_every_entity / ref_most_specific / ref_all_occurrences / using_per_ns / no_unused_using:
// These require detecting ABSENCE of {@...} refs for bare names, or auditing
// `using` declarations against actual ref usage. Not implementable without
// enumerating all in-scope construct names per file. Left to eval.

// This rule currently has no deterministically implementable checks.
// It is registered to satisfy the linter rule-set contract and will emit
// diagnostics as new deterministic checks are identified.

// a10n:blueprint:ignore
function hasDecorator(node: any, name: string): boolean {
  return (node?.decorators ?? []).some((d: any) => d.definition?.name === name);
}

export const specReferencesRule = createRule({
  name: "spec-references",
  severity: "warning",
  description: "Spec prose reference hygiene — currently enforced via inline-refs rule.",
  messages: {
    placeholder: paramMessage`spec-references: ${"msg"}`,
  },
  // a10n:blueprint Components.A10nBlueprintTypespecSdk.Operations.rule_spec_references
  create(_context) {
    return {
      // a10n:index:call packages.blueprint-typespec.src.rules.spec-references.rule.operation
      operation(_op: Operation) {
        // No deterministic checks currently implemented.
        // See comments above for escalated items.
      },
    };
  },
});
